// @pnp/cli-microsoft365/dist/m365/planner/commands/task/task-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PlannerTaskSetCommand_instances, _PlannerTaskSetCommand_initTelemetry, _PlannerTaskSetCommand_initOptions, _PlannerTaskSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../../../Auth");
const request_1 = require("../../../../request");
const accessToken_1 = require("../../../../utils/accessToken");
const formatting_1 = require("../../../../utils/formatting");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const planner_1 = require("../../../../utils/planner");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
const taskPriority_1 = require("../../taskPriority");
class PlannerTaskSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _PlannerTaskSetCommand_instances.add(this);
        this.allowedAppliedCategories = ['category1', 'category2', 'category3', 'category4', 'category5', 'category6'];
        __classPrivateFieldGet(this, _PlannerTaskSetCommand_instances, "m", _PlannerTaskSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PlannerTaskSetCommand_instances, "m", _PlannerTaskSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PlannerTaskSetCommand_instances, "m", _PlannerTaskSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.TASK_SET;
    }
    get description() {
        return 'Updates a Microsoft Planner Task';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.planName) {
                args.options.planTitle = args.options.planName;
                this.warn(logger, `Option 'planName' is deprecated. Please use 'planTitle' instead`);
            }
            if (accessToken_1.accessToken.isAppOnlyAccessToken(Auth_1.default.service.accessTokens[this.resource].accessToken)) {
                this.handleError('This command does not support application permissions.');
                return;
            }
            try {
                this.bucketId = yield this.getBucketId(args.options);
                this.assignments = yield this.generateUserAssignments(args.options);
                const etag = yield this.getTaskEtag(args.options.id);
                const appliedCategories = this.generateAppliedCategories(args.options);
                const data = this.mapRequestBody(args.options, appliedCategories);
                const requestOptions = {
                    url: `${this.resource}/v1.0/planner/tasks/${args.options.id}`,
                    headers: {
                        'accept': 'application/json;odata.metadata=none',
                        'If-Match': etag,
                        'Prefer': 'return=representation'
                    },
                    responseType: 'json',
                    data: data
                };
                const newTask = yield request_1.default.patch(requestOptions);
                const result = yield this.updateTaskDetails(args.options, newTask);
                logger.log(result);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    updateTaskDetails(options, newTask) {
        if (!options.description) {
            return Promise.resolve(newTask);
        }
        const taskId = newTask.id;
        return this
            .getTaskDetailsEtag(taskId)
            .then(etag => {
            const requestOptionsTaskDetails = {
                url: `${this.resource}/v1.0/planner/tasks/${taskId}/details`,
                headers: {
                    'accept': 'application/json;odata.metadata=none',
                    'If-Match': etag,
                    'Prefer': 'return=representation'
                },
                responseType: 'json',
                data: {
                    description: options.description
                }
            };
            return request_1.default.patch(requestOptionsTaskDetails);
        })
            .then(taskDetails => {
            return Object.assign(Object.assign({}, newTask), taskDetails);
        });
    }
    getTaskDetailsEtag(taskId) {
        const requestOptions = {
            url: `${this.resource}/v1.0/planner/tasks/${encodeURIComponent(taskId)}/details`,
            headers: {
                accept: 'application/json'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(requestOptions)
            .then((response) => response['@odata.etag']);
    }
    getTaskEtag(taskId) {
        const requestOptions = {
            url: `${this.resource}/v1.0/planner/tasks/${encodeURIComponent(taskId)}`,
            headers: {
                accept: 'application/json'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(requestOptions)
            .then((response) => response['@odata.etag']);
    }
    generateAppliedCategories(options) {
        if (!options.appliedCategories) {
            return {};
        }
        const categories = {};
        options.appliedCategories.toLocaleLowerCase().split(',').forEach(x => categories[x] = true);
        return categories;
    }
    generateUserAssignments(options) {
        const assignments = {};
        if (!options.assignedToUserIds && !options.assignedToUserNames) {
            return Promise.resolve(assignments);
        }
        return this
            .getUserIds(options)
            .then((userIds) => {
            userIds.forEach(x => assignments[x] = {
                '@odata.type': '#microsoft.graph.plannerAssignment',
                orderHint: ' !'
            });
            return Promise.resolve(assignments);
        });
    }
    getUserIds(options) {
        if (options.assignedToUserIds) {
            return Promise.resolve(options.assignedToUserIds.split(',').map(o => o.trim()));
        }
        // Hitting this section means assignedToUserNames won't be undefined
        const userNames = options.assignedToUserNames;
        const userArr = userNames.split(',').map(o => o.trim());
        let userIds = [];
        const promises = userArr.map(user => {
            const requestOptions = {
                url: `${this.resource}/v1.0/users?$filter=userPrincipalName eq '${formatting_1.formatting.encodeQueryParameter(user)}'&$select=id,userPrincipalName`,
                headers: {
                    'accept ': 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            return request_1.default.get(requestOptions);
        });
        return Promise
            .all(promises)
            .then((usersRes) => {
            let userUpns = [];
            userUpns = usersRes.map(res => { var _a; return (_a = res.value[0]) === null || _a === void 0 ? void 0 : _a.userPrincipalName; });
            userIds = usersRes.map(res => { var _a; return (_a = res.value[0]) === null || _a === void 0 ? void 0 : _a.id; });
            // Find the members where no graph response was found
            const invalidUsers = userArr.filter(user => !userUpns.some((upn) => (upn === null || upn === void 0 ? void 0 : upn.toLowerCase()) === user.toLowerCase()));
            if (invalidUsers && invalidUsers.length > 0) {
                return Promise.reject(`Cannot proceed with planner task update. The following users provided are invalid : ${invalidUsers.join(',')}`);
            }
            return Promise.resolve(userIds);
        });
    }
    getBucketId(options) {
        if (options.bucketId) {
            return Promise.resolve(options.bucketId);
        }
        if (!options.bucketName) {
            return Promise.resolve(undefined);
        }
        return this
            .getPlanId(options)
            .then(planId => {
            const requestOptions = {
                url: `${this.resource}/v1.0/planner/plans/${planId}/buckets?$select=id,name`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            return request_1.default.get(requestOptions);
        })
            .then((response) => {
            const bucket = response.value.find(val => val.name === options.bucketName);
            if (!bucket) {
                return Promise.reject(`The specified bucket does not exist`);
            }
            return Promise.resolve(bucket.id);
        });
    }
    getPlanId(options) {
        if (options.planId) {
            return Promise.resolve(options.planId);
        }
        return this
            .getGroupId(options)
            .then((groupId) => planner_1.planner.getPlanByTitle(options.planTitle, groupId))
            .then(plan => plan.id);
    }
    getGroupId(options) {
        if (options.ownerGroupId) {
            return Promise.resolve(options.ownerGroupId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(options.ownerGroupName)
            .then(group => group.id);
    }
    mapRequestBody(options, appliedCategories) {
        const requestBody = {};
        if (options.title) {
            requestBody.title = options.title;
        }
        if (this.bucketId) {
            requestBody.bucketId = this.bucketId;
        }
        if (options.startDateTime) {
            requestBody.startDateTime = options.startDateTime;
        }
        if (options.dueDateTime) {
            requestBody.dueDateTime = options.dueDateTime;
        }
        if (options.percentComplete) {
            requestBody.percentComplete = options.percentComplete;
        }
        if (this.assignments && Object.keys(this.assignments).length > 0) {
            requestBody.assignments = this.assignments;
        }
        if (options.assigneePriority) {
            requestBody.assigneePriority = options.assigneePriority;
        }
        if (appliedCategories && Object.keys(appliedCategories).length > 0) {
            requestBody.appliedCategories = appliedCategories;
        }
        if (options.orderHint) {
            requestBody.orderHint = options.orderHint;
        }
        if (options.priority !== undefined) {
            requestBody.priority = taskPriority_1.taskPriority.getPriorityValue(options.priority);
        }
        return requestBody;
    }
}
_PlannerTaskSetCommand_instances = new WeakSet(), _PlannerTaskSetCommand_initTelemetry = function _PlannerTaskSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            title: typeof args.options.title !== 'undefined',
            planId: typeof args.options.planId !== 'undefined',
            planName: typeof args.options.planName !== 'undefined',
            planTitle: typeof args.options.planTitle !== 'undefined',
            ownerGroupId: typeof args.options.ownerGroupId !== 'undefined',
            ownerGroupName: typeof args.options.ownerGroupName !== 'undefined',
            bucketId: typeof args.options.bucketId !== 'undefined',
            bucketName: typeof args.options.bucketName !== 'undefined',
            startDateTime: typeof args.options.startDateTime !== 'undefined',
            dueDateTime: typeof args.options.dueDateTime !== 'undefined',
            percentComplete: typeof args.options.percentComplete !== 'undefined',
            assignedToUserIds: typeof args.options.assignedToUserIds !== 'undefined',
            assignedToUserNames: typeof args.options.assignedToUserNames !== 'undefined',
            assigneePriority: typeof args.options.assigneePriority !== 'undefined',
            description: typeof args.options.description !== 'undefined',
            appliedCategories: typeof args.options.appliedCategories !== 'undefined',
            orderHint: typeof args.options.orderHint !== 'undefined',
            priority: typeof args.options.priority !== 'undefined'
        });
    });
}, _PlannerTaskSetCommand_initOptions = function _PlannerTaskSetCommand_initOptions() {
    this.options.unshift({ option: '-i, --id <id>' }, { option: '-t, --title [title]' }, { option: '--planId [planId]' }, { option: '--planName [planName]' }, { option: '--planTitle [planTitle]' }, { option: '--ownerGroupId [ownerGroupId]' }, { option: '--ownerGroupName [ownerGroupName]' }, { option: '--bucketId [bucketId]' }, { option: '--bucketName [bucketName]' }, { option: '--startDateTime [startDateTime]' }, { option: '--dueDateTime [dueDateTime]' }, { option: '--percentComplete [percentComplete]' }, { option: '--assignedToUserIds [assignedToUserIds]' }, { option: '--assignedToUserNames [assignedToUserNames]' }, { option: '--assigneePriority [assigneePriority]' }, { option: '--description [description]' }, { option: '--appliedCategories [appliedCategories]' }, { option: '--orderHint [orderHint]' }, { option: '--priority [priority]', autocomplete: taskPriority_1.taskPriority.priorityValues });
}, _PlannerTaskSetCommand_initValidators = function _PlannerTaskSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.bucketId && args.options.bucketName) {
            return 'Specify either bucketId or bucketName but not both';
        }
        if (args.options.bucketName && !args.options.planId && !args.options.planName && !args.options.planTitle) {
            return 'Specify either planId or planTitle when using bucketName';
        }
        if (args.options.bucketName && args.options.planId && (args.options.planName || args.options.planTitle)) {
            return 'Specify either planId or planTitle when using bucketName but not both';
        }
        if ((args.options.planName || args.options.planTitle) && !args.options.ownerGroupId && !args.options.ownerGroupName) {
            return 'Specify either ownerGroupId or ownerGroupName when using planTitle';
        }
        if ((args.options.planName || args.options.planTitle) && args.options.ownerGroupId && args.options.ownerGroupName) {
            return 'Specify either ownerGroupId or ownerGroupName when using planTitle but not both';
        }
        if (args.options.ownerGroupId && !validation_1.validation.isValidGuid(args.options.ownerGroupId)) {
            return `${args.options.ownerGroupId} is not a valid GUID`;
        }
        if (args.options.startDateTime && !validation_1.validation.isValidISODateTime(args.options.startDateTime)) {
            return 'The startDateTime is not a valid ISO date string';
        }
        if (args.options.dueDateTime && !validation_1.validation.isValidISODateTime(args.options.dueDateTime)) {
            return 'The dueDateTime is not a valid ISO date string';
        }
        if (args.options.percentComplete && isNaN(args.options.percentComplete)) {
            return `percentComplete is not a number`;
        }
        if (args.options.percentComplete && (args.options.percentComplete < 0 || args.options.percentComplete > 100)) {
            return `percentComplete should be between 0 and 100`;
        }
        if (args.options.assignedToUserIds && !validation_1.validation.isValidGuidArray(args.options.assignedToUserIds.split(','))) {
            return 'assignedToUserIds contains invalid GUID';
        }
        if (args.options.assignedToUserIds && args.options.assignedToUserNames) {
            return 'Specify either assignedToUserIds or assignedToUserNames but not both';
        }
        if (args.options.appliedCategories && args.options.appliedCategories.split(',').filter(category => this.allowedAppliedCategories.indexOf(category.toLocaleLowerCase()) < 0).length !== 0) {
            return 'The appliedCategories contains invalid value. Specify either category1, category2, category3, category4, category5 and/or category6 as properties';
        }
        if (args.options.priority !== undefined) {
            if (typeof args.options.priority === "number") {
                if (isNaN(args.options.priority) || args.options.priority < 0 || args.options.priority > 10 || !Number.isInteger(args.options.priority)) {
                    return 'priority should be an integer between 0 and 10.';
                }
            }
            else if (taskPriority_1.taskPriority.priorityValues.map(l => l.toLowerCase()).indexOf(args.options.priority.toString().toLowerCase()) === -1) {
                return `${args.options.priority} is not a valid priority value. Allowed values are ${taskPriority_1.taskPriority.priorityValues.join('|')}.`;
            }
        }
        return true;
    }));
};
module.exports = new PlannerTaskSetCommand();
