// @pnp/cli-microsoft365/dist/m365/aad/commands/siteclassification/siteclassification-disable.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadSiteClassificationDisableCommand_instances, _AadSiteClassificationDisableCommand_initTelemetry, _AadSiteClassificationDisableCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadSiteClassificationDisableCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadSiteClassificationDisableCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadSiteClassificationDisableCommand_instances, "m", _AadSiteClassificationDisableCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _AadSiteClassificationDisableCommand_instances, "m", _AadSiteClassificationDisableCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.SITECLASSIFICATION_DISABLE;
    }
    get description() {
        return 'Disables site classification';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const disableSiteClassification = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    let requestOptions = {
                        url: `${this.resource}/v1.0/groupSettings`,
                        headers: {
                            accept: 'application/json;odata.metadata=none'
                        },
                        responseType: 'json'
                    };
                    const res = yield request_1.default.get(requestOptions);
                    if (res.value.length === 0) {
                        throw 'Site classification is not enabled.';
                    }
                    const unifiedGroupSetting = res.value.filter((directorySetting) => {
                        return directorySetting.displayName === 'Group.Unified';
                    });
                    if (!unifiedGroupSetting || unifiedGroupSetting.length === 0) {
                        throw 'Missing DirectorySettingTemplate for "Group.Unified"';
                    }
                    if (!unifiedGroupSetting[0] ||
                        !unifiedGroupSetting[0].id || unifiedGroupSetting[0].id.length === 0) {
                        throw 'Missing UnifiedGroupSettting id';
                    }
                    requestOptions = {
                        url: `${this.resource}/v1.0/groupSettings/` + unifiedGroupSetting[0].id,
                        headers: {
                            accept: 'application/json;odata.metadata=none',
                            'content-type': 'application/json'
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield disableSiteClassification();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to disable site classification?`
                });
                if (result.continue) {
                    yield disableSiteClassification();
                }
            }
        });
    }
}
_AadSiteClassificationDisableCommand_instances = new WeakSet(), _AadSiteClassificationDisableCommand_initTelemetry = function _AadSiteClassificationDisableCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _AadSiteClassificationDisableCommand_initOptions = function _AadSiteClassificationDisableCommand_initOptions() {
    this.options.unshift({
        option: '--confirm'
    });
};
module.exports = new AadSiteClassificationDisableCommand();
