// @pnp/cli-microsoft365/dist/m365/aad/commands/siteclassification/siteclassification-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadSiteClassificationSetCommand_instances, _AadSiteClassificationSetCommand_initTelemetry, _AadSiteClassificationSetCommand_initOptions, _AadSiteClassificationSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
const DirectorySetting_1 = require("./DirectorySetting");
class AadSiteClassificationSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadSiteClassificationSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadSiteClassificationSetCommand_instances, "m", _AadSiteClassificationSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _AadSiteClassificationSetCommand_instances, "m", _AadSiteClassificationSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _AadSiteClassificationSetCommand_instances, "m", _AadSiteClassificationSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITECLASSIFICATION_SET;
    }
    get description() {
        return 'Updates site classification configuration';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                let requestOptions = {
                    url: `${this.resource}/v1.0/groupSettings`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    responseType: 'json'
                };
                const res = yield request_1.default.get(requestOptions);
                const unifiedGroupSetting = res.value.filter((directorySetting) => {
                    return directorySetting.displayName === 'Group.Unified';
                });
                if (!unifiedGroupSetting ||
                    unifiedGroupSetting.length === 0) {
                    throw "There is no previous defined site classification which can updated.";
                }
                const updatedDirSettings = new DirectorySetting_1.UpdateDirectorySetting();
                unifiedGroupSetting[0].values.forEach((directorySetting) => {
                    switch (directorySetting.name) {
                        case "ClassificationList":
                            if (args.options.classifications) {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": args.options.classifications
                                });
                            }
                            else {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": directorySetting.value
                                });
                            }
                            break;
                        case "DefaultClassification":
                            if (args.options.defaultClassification) {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": args.options.defaultClassification
                                });
                            }
                            else {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": directorySetting.value
                                });
                            }
                            break;
                        case "UsageGuidelinesUrl":
                            if (args.options.usageGuidelinesUrl) {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": args.options.usageGuidelinesUrl
                                });
                            }
                            else {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": directorySetting.value
                                });
                            }
                            break;
                        case "GuestUsageGuidelinesUrl":
                            if (args.options.guestUsageGuidelinesUrl) {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": args.options.guestUsageGuidelinesUrl
                                });
                            }
                            else {
                                updatedDirSettings.values.push({
                                    "name": directorySetting.name,
                                    "value": directorySetting.value
                                });
                            }
                            break;
                        default:
                            updatedDirSettings.values.push({
                                "name": directorySetting.name,
                                "value": directorySetting.value
                            });
                            break;
                    }
                });
                requestOptions = {
                    url: `${this.resource}/v1.0/groupSettings/${unifiedGroupSetting[0].id}`,
                    headers: {
                        accept: 'application/json;odata.metadata=none',
                        'content-type': 'application/json'
                    },
                    responseType: 'json',
                    data: updatedDirSettings
                };
                yield request_1.default.patch(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadSiteClassificationSetCommand_instances = new WeakSet(), _AadSiteClassificationSetCommand_initTelemetry = function _AadSiteClassificationSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            classifications: typeof args.options.classifications !== 'undefined',
            defaultClassification: typeof args.options.defaultClassification !== 'undefined',
            usageGuidelinesUrl: typeof args.options.usageGuidelinesUrl !== 'undefined',
            guestUsageGuidelinesUrl: typeof args.options.guestUsageGuidelinesUrl !== 'undefined'
        });
    });
}, _AadSiteClassificationSetCommand_initOptions = function _AadSiteClassificationSetCommand_initOptions() {
    this.options.unshift({
        option: '-c, --classifications [classifications]'
    }, {
        option: '-d, --defaultClassification [defaultClassification]'
    }, {
        option: '-u, --usageGuidelinesUrl [usageGuidelinesUrl]'
    }, {
        option: '-g, --guestUsageGuidelinesUrl [guestUsageGuidelinesUrl]'
    });
}, _AadSiteClassificationSetCommand_initValidators = function _AadSiteClassificationSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!args.options.classifications &&
            !args.options.defaultClassification &&
            !args.options.usageGuidelinesUrl &&
            !args.options.guestUsageGuidelinesUrl) {
            return 'Specify at least one property to update';
        }
        return true;
    }));
};
module.exports = new AadSiteClassificationSetCommand();
