// @fluentui/accessibility/dist/dts/src/behaviors/Datepicker/datepickerBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.datepickerBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @description
 * Behavior for a datepicker component
 * @specification
 * Triggers 'open' action with 'Enter' on 'input'.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'input' slot.
 * Adds attribute 'aria-invalid' based on the property 'aria-invalid' to 'input' slot.
 */
var datepickerBehavior = function (props) { return ({
    attributes: {
        input: __assign({ 'aria-labelledby': props['aria-labelledby'], 'aria-invalid': props['aria-invalid'] }, (!props.allowManualInput && { 'aria-haspopup': true })),
    },
    keyActions: {
        input: {
            open: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }],
            },
        },
    },
}); };
exports.datepickerBehavior = datepickerBehavior;
