// @pnp/cli-microsoft365/dist/m365/spo/commands/listitem/listitem-roleassignment-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListItemRoleAssignmentAddCommand_instances, _SpoListItemRoleAssignmentAddCommand_initTelemetry, _SpoListItemRoleAssignmentAddCommand_initOptions, _SpoListItemRoleAssignmentAddCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
const SpoUserGetCommand = require("../user/user-get");
const SpoGroupGetCommand = require("../group/group-get");
const SpoRoleDefinitionListCommand = require("../roledefinition/roledefinition-list");
const validation_1 = require("../../../../utils/validation");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const Cli_1 = require("../../../../cli/Cli");
class SpoListItemRoleAssignmentAddCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListItemRoleAssignmentAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListItemRoleAssignmentAddCommand_instances, "m", _SpoListItemRoleAssignmentAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListItemRoleAssignmentAddCommand_instances, "m", _SpoListItemRoleAssignmentAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListItemRoleAssignmentAddCommand_instances, "m", _SpoListItemRoleAssignmentAddCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.LISTITEM_ROLEASSIGNMENT_ADD;
    }
    get description() {
        return 'Adds a role assignment to a listitem.';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Adding role assignment to listitem in site at ${args.options.webUrl}...`);
            }
            try {
                let requestUrl = `${args.options.webUrl}/_api/web/`;
                if (args.options.listId) {
                    requestUrl += `lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')/`;
                }
                else if (args.options.listTitle) {
                    requestUrl += `lists/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')/`;
                }
                else if (args.options.listUrl) {
                    const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                    requestUrl += `GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')/`;
                }
                requestUrl += `items(${args.options.listItemId})/`;
                const roleDefinitionId = yield this.getRoleDefinitionId(args.options);
                let principalId = 0;
                if (args.options.upn) {
                    principalId = yield this.getUserPrincipalId(args.options);
                    yield this.addRoleAssignment(requestUrl, roleDefinitionId, principalId);
                }
                else if (args.options.groupName) {
                    principalId = yield this.getGroupPrincipalId(args.options);
                    yield this.addRoleAssignment(requestUrl, roleDefinitionId, principalId);
                }
                else {
                    principalId = args.options.principalId;
                    yield this.addRoleAssignment(requestUrl, roleDefinitionId, principalId);
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    addRoleAssignment(requestUrl, roleDefinitionId, principalId) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const requestOptions = {
                    url: `${requestUrl}roleassignments/addroleassignment(principalid='${principalId}',roledefid='${roleDefinitionId}')`,
                    method: 'POST',
                    headers: {
                        'accept': 'application/json;odata=nometadata',
                        'content-type': 'application/json'
                    },
                    responseType: 'json'
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                return Promise.reject(err);
            }
        });
    }
    getRoleDefinitionId(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!options.roleDefinitionName) {
                return options.roleDefinitionId;
            }
            try {
                const roleDefinitionListCommandOptions = {
                    webUrl: options.webUrl,
                    output: 'json',
                    debug: this.debug,
                    verbose: this.verbose
                };
                const output = yield Cli_1.Cli.executeCommandWithOutput(SpoRoleDefinitionListCommand, { options: Object.assign(Object.assign({}, roleDefinitionListCommandOptions), { _: [] }) });
                const getRoleDefinitionListOutput = JSON.parse(output.stdout);
                const roleDefinitionId = getRoleDefinitionListOutput.find((role) => role.Name === options.roleDefinitionName).Id;
                return roleDefinitionId;
            }
            catch (err) {
                return Promise.reject(err);
            }
        });
    }
    getGroupPrincipalId(options) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const groupGetCommandOptions = {
                    webUrl: options.webUrl,
                    name: options.groupName,
                    output: 'json',
                    debug: this.debug,
                    verbose: this.verbose
                };
                const output = yield Cli_1.Cli.executeCommandWithOutput(SpoGroupGetCommand, { options: Object.assign(Object.assign({}, groupGetCommandOptions), { _: [] }) });
                const getGroupOutput = JSON.parse(output.stdout);
                return getGroupOutput.Id;
            }
            catch (err) {
                return Promise.reject(err);
            }
        });
    }
    getUserPrincipalId(options) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const userGetCommandOptions = {
                    webUrl: options.webUrl,
                    email: options.upn,
                    id: undefined,
                    output: 'json',
                    debug: this.debug,
                    verbose: this.verbose
                };
                const output = yield Cli_1.Cli.executeCommandWithOutput(SpoUserGetCommand, { options: Object.assign(Object.assign({}, userGetCommandOptions), { _: [] }) });
                const getUserOutput = JSON.parse(output.stdout);
                return getUserOutput.Id;
            }
            catch (err) {
                return Promise.reject(err);
            }
        });
    }
}
_SpoListItemRoleAssignmentAddCommand_instances = new WeakSet(), _SpoListItemRoleAssignmentAddCommand_initTelemetry = function _SpoListItemRoleAssignmentAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            principalId: typeof args.options.principalId !== 'undefined',
            upn: typeof args.options.upn !== 'undefined',
            groupName: typeof args.options.groupName !== 'undefined',
            roleDefinitionId: typeof args.options.roleDefinitionId !== 'undefined',
            roleDefinitionName: typeof args.options.roleDefinitionName !== 'undefined'
        });
    });
}, _SpoListItemRoleAssignmentAddCommand_initOptions = function _SpoListItemRoleAssignmentAddCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--listItemId <listItemId>'
    }, {
        option: '--listId [listId]'
    }, {
        option: '--listTitle [listTitle]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '--principalId [principalId]'
    }, {
        option: '--upn [upn]'
    }, {
        option: '--groupName [groupName]'
    }, {
        option: '--roleDefinitionId [roleDefinitionId]'
    }, {
        option: '--roleDefinitionName [roleDefinitionName]'
    });
}, _SpoListItemRoleAssignmentAddCommand_initValidators = function _SpoListItemRoleAssignmentAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId && !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} is not a valid GUID`;
        }
        if (args.options.listItemId && isNaN(args.options.listItemId)) {
            return `Specified listItemId ${args.options.listItemId} is not a number`;
        }
        if (args.options.principalId && isNaN(args.options.principalId)) {
            return `Specified principalId ${args.options.principalId} is not a number`;
        }
        if (args.options.roleDefinitionId && isNaN(args.options.roleDefinitionId)) {
            return `Specified roleDefinitionId ${args.options.roleDefinitionId} is not a number`;
        }
        const listOptions = [args.options.listId, args.options.listTitle, args.options.listUrl];
        if (listOptions.some(item => item !== undefined) && listOptions.filter(item => item !== undefined).length > 1) {
            return `Specify either list id or title or list url`;
        }
        if (listOptions.filter(item => item !== undefined).length === 0) {
            return `Specify at least list id or title or list url`;
        }
        const principalOptions = [args.options.principalId, args.options.upn, args.options.groupName];
        if (!principalOptions.some(item => item !== undefined)) {
            return `Specify either principalId, upn or groupName`;
        }
        if (principalOptions.filter(item => item !== undefined).length > 1) {
            return `Specify either principalId, upn or groupName but not multiple`;
        }
        const roleDefinitionOptions = [args.options.roleDefinitionId, args.options.roleDefinitionName];
        if (!roleDefinitionOptions.some(item => item !== undefined)) {
            return `Specify either roleDefinitionId or roleDefinitionName`;
        }
        if (roleDefinitionOptions.filter(item => item !== undefined).length > 1) {
            return `Specify either roleDefinitionId or roleDefinitionName but not multiple`;
        }
        return true;
    }));
};
module.exports = new SpoListItemRoleAssignmentAddCommand();
