// @pnp/cli-microsoft365/dist/m365/teams/commands/messagingsettings/messagingsettings-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsMessagingSettingsSetCommand_instances, _TeamsMessagingSettingsSetCommand_initTelemetry, _TeamsMessagingSettingsSetCommand_initOptions, _TeamsMessagingSettingsSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsMessagingSettingsSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsMessagingSettingsSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsMessagingSettingsSetCommand_instances, "m", _TeamsMessagingSettingsSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsMessagingSettingsSetCommand_instances, "m", _TeamsMessagingSettingsSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsMessagingSettingsSetCommand_instances, "m", _TeamsMessagingSettingsSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.MESSAGINGSETTINGS_SET;
    }
    get description() {
        return 'Updates messaging settings of a Microsoft Teams team';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {
                messagingSettings: {}
            };
            TeamsMessagingSettingsSetCommand.props.forEach((p) => {
                if (typeof args.options[p] !== 'undefined') {
                    data.messagingSettings[p] = args.options[p].toLowerCase() === 'true';
                }
            });
            const requestOptions = {
                url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                data: data,
                responseType: 'json'
            };
            try {
                yield request_1.default.patch(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsMessagingSettingsSetCommand_instances = new WeakSet(), _TeamsMessagingSettingsSetCommand_initTelemetry = function _TeamsMessagingSettingsSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        TeamsMessagingSettingsSetCommand.props.forEach((p) => {
            this.telemetryProperties[p] = typeof args.options[p] !== 'undefined';
        });
    });
}, _TeamsMessagingSettingsSetCommand_initOptions = function _TeamsMessagingSettingsSetCommand_initOptions() {
    this.options.unshift({
        option: '-i, --teamId <teamId>'
    }, {
        option: '--allowUserEditMessages [allowUserEditMessages]'
    }, {
        option: '--allowUserDeleteMessages [allowUserDeleteMessages]'
    }, {
        option: '--allowOwnerDeleteMessages [allowOwnerDeleteMessages]'
    }, {
        option: '--allowTeamMentions [allowTeamMentions]'
    }, {
        option: '--allowChannelMentions [allowChannelMentions]'
    });
}, _TeamsMessagingSettingsSetCommand_initValidators = function _TeamsMessagingSettingsSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        let hasDuplicate = false;
        let property = '';
        TeamsMessagingSettingsSetCommand.props.forEach((prop) => {
            if (args.options[prop] instanceof Array) {
                property = prop;
                hasDuplicate = true;
            }
        });
        if (hasDuplicate) {
            return `Duplicate option ${property} specified. Specify only one`;
        }
        let isValid = true;
        let value = '';
        TeamsMessagingSettingsSetCommand.props.every((p) => {
            property = p;
            value = args.options[p];
            isValid = typeof value === 'undefined' || validation_1.validation.isValidBoolean(value);
            return isValid;
        });
        if (!isValid) {
            return `Value ${value} for option ${property} is not a valid boolean`;
        }
        return true;
    }));
};
TeamsMessagingSettingsSetCommand.props = [
    'allowUserEditMessages',
    'allowUserDeleteMessages',
    'allowOwnerDeleteMessages',
    'allowTeamMentions',
    'allowChannelMentions'
];
module.exports = new TeamsMessagingSettingsSetCommand();
