// @pnp/cli-microsoft365/dist/m365/spo/commands/theme/theme-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoThemeSetCommand_instances, _SpoThemeSetCommand_initTelemetry, _SpoThemeSetCommand_initOptions, _SpoThemeSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../../../config");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoThemeSetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoThemeSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoThemeSetCommand_instances, "m", _SpoThemeSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoThemeSetCommand_instances, "m", _SpoThemeSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoThemeSetCommand_instances, "m", _SpoThemeSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.THEME_SET;
    }
    get description() {
        return 'Add or update a theme';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const spoAdminUrl = yield spo_1.spo.getSpoAdminUrl(logger, this.debug);
                const res = yield spo_1.spo.getRequestDigest(spoAdminUrl);
                const palette = JSON.parse(args.options.theme);
                if (this.debug) {
                    logger.logToStderr('');
                    logger.logToStderr('Palette');
                    logger.logToStderr(JSON.stringify(palette));
                }
                const isInverted = args.options.isInverted ? true : false;
                const requestOptions = {
                    url: `${spoAdminUrl}/_vti_bin/client.svc/ProcessQuery`,
                    headers: {
                        'X-RequestDigest': res.FormDigestValue
                    },
                    data: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config_1.default.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><ObjectPath Id="10" ObjectPathId="9" /><Method Name="UpdateTenantTheme" Id="11" ObjectPathId="9"><Parameters><Parameter Type="String">${formatting_1.formatting.escapeXml(args.options.name)}</Parameter><Parameter Type="String">{"isInverted":${isInverted},"name":"${formatting_1.formatting.escapeXml(args.options.name)}","palette":${JSON.stringify(palette)}}</Parameter></Parameters></Method></Actions><ObjectPaths><Constructor Id="9" TypeId="{268004ae-ef6b-4e9b-8425-127220d84719}"/></ObjectPaths></Request>`
                };
                const processQuery = yield request_1.default.post(requestOptions);
                const json = JSON.parse(processQuery);
                const contents = json.find(x => { return x['ErrorInfo']; });
                if (contents && contents.ErrorInfo) {
                    throw contents.ErrorInfo.ErrorMessage || 'ClientSvc unknown error';
                }
                return Promise.resolve();
            }
            catch (err) {
                this.handleRejectedPromise(err);
            }
        });
    }
}
_SpoThemeSetCommand_instances = new WeakSet(), _SpoThemeSetCommand_initTelemetry = function _SpoThemeSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            inverted: (!(!args.options.isInverted)).toString()
        });
    });
}, _SpoThemeSetCommand_initOptions = function _SpoThemeSetCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    }, {
        option: '-t, --theme <theme>'
    }, {
        option: '--isInverted'
    });
}, _SpoThemeSetCommand_initValidators = function _SpoThemeSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidTheme(args.options.theme)) {
            return 'The specified theme is not valid';
        }
        return true;
    }));
};
module.exports = new SpoThemeSetCommand();
