// @pnp/cli-microsoft365/dist/m365/spo/commands/theme/theme-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoThemeRemoveCommand_instances, _SpoThemeRemoveCommand_initTelemetry, _SpoThemeRemoveCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoThemeRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoThemeRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoThemeRemoveCommand_instances, "m", _SpoThemeRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoThemeRemoveCommand_instances, "m", _SpoThemeRemoveCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.THEME_REMOVE;
    }
    get description() {
        return 'Removes existing theme';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeTheme = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const spoAdminUrl = yield spo_1.spo.getSpoAdminUrl(logger, this.debug);
                    if (this.verbose) {
                        logger.logToStderr(`Removing theme from tenant...`);
                    }
                    const requestOptions = {
                        url: `${spoAdminUrl}/_api/thememanager/DeleteTenantTheme`,
                        headers: {
                            'accept': 'application/json;odata=nometadata'
                        },
                        data: {
                            name: args.options.name
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeTheme();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the theme`
                });
                if (result.continue) {
                    yield removeTheme();
                }
            }
        });
    }
}
_SpoThemeRemoveCommand_instances = new WeakSet(), _SpoThemeRemoveCommand_initTelemetry = function _SpoThemeRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoThemeRemoveCommand_initOptions = function _SpoThemeRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    }, {
        option: '--confirm'
    });
};
module.exports = new SpoThemeRemoveCommand();
