// @pnp/cli-microsoft365/dist/m365/spo/commands/feature/feature-disable.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoFeatureDisableCommand_instances, _SpoFeatureDisableCommand_initTelemetry, _SpoFeatureDisableCommand_initOptions, _SpoFeatureDisableCommand_initValidators, _SpoFeatureDisableCommand_initTypes;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoFeatureDisableCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoFeatureDisableCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoFeatureDisableCommand_instances, "m", _SpoFeatureDisableCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoFeatureDisableCommand_instances, "m", _SpoFeatureDisableCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoFeatureDisableCommand_instances, "m", _SpoFeatureDisableCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoFeatureDisableCommand_instances, "m", _SpoFeatureDisableCommand_initTypes).call(this);
    }
    get name() {
        return commands_1.default.FEATURE_DISABLE;
    }
    get description() {
        return 'Disables feature for the specified site or web';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            let scope = args.options.scope;
            let force = args.options.force;
            if (!scope) {
                scope = "web";
            }
            if (!force) {
                force = false;
            }
            if (this.verbose) {
                logger.logToStderr(`Disabling feature '${args.options.featureId}' on scope '${scope}' for url '${args.options.url}' (force='${force}')...`);
            }
            const url = `${args.options.url}/_api/${scope}/features/remove(featureId=guid'${args.options.featureId}',force=${force})`;
            const requestOptions = {
                url: url,
                headers: {
                    accept: 'application/json;odata=nometadata'
                }
            };
            try {
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoFeatureDisableCommand_instances = new WeakSet(), _SpoFeatureDisableCommand_initTelemetry = function _SpoFeatureDisableCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            scope: args.options.scope || 'web',
            force: args.options.force || false
        });
    });
}, _SpoFeatureDisableCommand_initOptions = function _SpoFeatureDisableCommand_initOptions() {
    this.options.unshift({
        option: '-u, --url <url>'
    }, {
        option: '-f, --featureId <id>'
    }, {
        option: '-s, --scope [scope]',
        autocomplete: ['Site', 'Web']
    }, {
        option: '--force'
    });
}, _SpoFeatureDisableCommand_initValidators = function _SpoFeatureDisableCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.scope) {
            if (['site', 'web'].indexOf(args.options.scope.toLowerCase()) < 0) {
                return `${args.options.scope} is not a valid Feature scope. Allowed values are Site|Web`;
            }
        }
        return true;
    }));
}, _SpoFeatureDisableCommand_initTypes = function _SpoFeatureDisableCommand_initTypes() {
    this.types.string.push('scope', 's');
};
module.exports = new SpoFeatureDisableCommand();
