// @pnp/cli-microsoft365/dist/m365/yammer/commands/network/network-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YammerNetworkListCommand_instances, _YammerNetworkListCommand_initTelemetry, _YammerNetworkListCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const YammerCommand_1 = require("../../../base/YammerCommand");
const commands_1 = require("../../commands");
class YammerNetworkListCommand extends YammerCommand_1.default {
    constructor() {
        super();
        _YammerNetworkListCommand_instances.add(this);
        __classPrivateFieldGet(this, _YammerNetworkListCommand_instances, "m", _YammerNetworkListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _YammerNetworkListCommand_instances, "m", _YammerNetworkListCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.NETWORK_LIST;
    }
    get description() {
        return 'Returns a list of networks to which the current user has access';
    }
    defaultProperties() {
        return ['id', 'name', 'email', 'community', 'permalink', 'web_url'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/v1/networks/current.json`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json;odata=nometadata'
                },
                responseType: 'json',
                data: {
                    includeSuspended: args.options.includeSuspended !== undefined && args.options.includeSuspended !== false
                }
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_YammerNetworkListCommand_instances = new WeakSet(), _YammerNetworkListCommand_initTelemetry = function _YammerNetworkListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            includeSuspended: args.options.includeSuspended
        });
    });
}, _YammerNetworkListCommand_initOptions = function _YammerNetworkListCommand_initOptions() {
    this.options.unshift({
        option: '--includeSuspended'
    });
};
module.exports = new YammerNetworkListCommand();
