// @pnp/cli-microsoft365/dist/m365/spo/commands/spo-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSetCommand_instances, _SpoSetCommand_initOptions, _SpoSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../../Auth");
const validation_1 = require("../../../utils/validation");
const SpoCommand_1 = require("../../base/SpoCommand");
const commands_1 = require("../commands");
class SpoSetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSetCommand_instances, "m", _SpoSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSetCommand_instances, "m", _SpoSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SET;
    }
    get description() {
        return 'Sets the URL of the root SharePoint site collection for use in SPO commands';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            Auth_1.default.service.spoUrl = args.options.url;
            try {
                yield Auth_1.default.storeConnectionInfo();
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSetCommand_instances = new WeakSet(), _SpoSetCommand_initOptions = function _SpoSetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --url <url>'
    });
}, _SpoSetCommand_initValidators = function _SpoSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.url); }));
};
module.exports = new SpoSetCommand();
