// @pnp/cli-microsoft365/dist/utils/odata.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.odata = void 0;
const request_1 = require("../request");
exports.odata = {
    getAllItems(url, metadata) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            let items = [];
            const requestOptions = {
                url: url,
                headers: {
                    accept: `application/json;odata.metadata=${metadata !== null && metadata !== void 0 ? metadata : 'none'}`
                },
                responseType: 'json'
            };
            const res = yield request_1.default.get(requestOptions);
            items = res.value;
            const nextLink = (_a = res['@odata.nextLink']) !== null && _a !== void 0 ? _a : res.nextLink;
            if (nextLink) {
                const nextPageItems = yield exports.odata.getAllItems(nextLink, metadata);
                items = items.concat(nextPageItems);
            }
            return items;
        });
    }
};
