// @pnp/cli-microsoft365/dist/utils/powerPlatform.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.powerPlatform = void 0;
const request_1 = require("../request");
const powerPlatformResource = 'https://api.bap.microsoft.com';
exports.powerPlatform = {
    getDynamicsInstanceApiUrl(environment, asAdmin) {
        return __awaiter(this, void 0, void 0, function* () {
            let url = '';
            if (asAdmin) {
                url = `${powerPlatformResource}/providers/Microsoft.BusinessAppPlatform/scopes/admin/environments/${encodeURIComponent(environment)}`;
            }
            else {
                url = `${powerPlatformResource}/providers/Microsoft.BusinessAppPlatform/environments/${encodeURIComponent(environment)}`;
            }
            const requestOptions = {
                url: `${url}?api-version=2020-10-01&$select=properties.linkedEnvironmentMetadata.instanceApiUrl`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            try {
                const response = yield request_1.default.get(requestOptions);
                return Promise.resolve(response.properties.linkedEnvironmentMetadata.instanceApiUrl);
            }
            catch (ex) {
                throw Error(`The environment '${environment}' could not be retrieved. See the inner exception for more details: ${ex.message}`);
            }
        });
    }
};
