// @pnp/cli-microsoft365/dist/m365/teams/commands/tab/tab-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsTabGetCommand_instances, _TeamsTabGetCommand_initTelemetry, _TeamsTabGetCommand_initOptions, _TeamsTabGetCommand_initValidators, _TeamsTabGetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsTabGetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsTabGetCommand_instances.add(this);
        this.teamId = "";
        this.channelId = "";
        __classPrivateFieldGet(this, _TeamsTabGetCommand_instances, "m", _TeamsTabGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsTabGetCommand_instances, "m", _TeamsTabGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsTabGetCommand_instances, "m", _TeamsTabGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _TeamsTabGetCommand_instances, "m", _TeamsTabGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.TAB_GET;
    }
    get description() {
        return 'Gets information about the specified Microsoft Teams tab';
    }
    getTeamId(args) {
        if (args.options.teamId) {
            return Promise.resolve(args.options.teamId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.teamName)
            .then(group => {
            if (group.resourceProvisioningOptions.indexOf('Team') === -1) {
                return Promise.reject(`The specified team does not exist in the Microsoft Teams`);
            }
            return group.id;
        });
    }
    getChannelId(args) {
        if (args.options.channelId) {
            return Promise.resolve(args.options.channelId);
        }
        const channelRequestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(this.teamId)}/channels?$filter=displayName eq '${encodeURIComponent(args.options.channelName)}'`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(channelRequestOptions)
            .then(response => {
            const channelItem = response.value[0];
            if (!channelItem) {
                return Promise.reject(`The specified channel does not exist in the Microsoft Teams team`);
            }
            return Promise.resolve(channelItem.id);
        });
    }
    getTabId(args) {
        if (args.options.tabId) {
            return Promise.resolve(args.options.tabId);
        }
        const tabRequestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(this.teamId)}/channels/${encodeURIComponent(this.channelId)}/tabs?$filter=displayName eq '${encodeURIComponent(args.options.tabName)}'`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(tabRequestOptions)
            .then(response => {
            const tabItem = response.value[0];
            if (!tabItem) {
                return Promise.reject(`The specified tab does not exist in the Microsoft Teams team channel`);
            }
            return Promise.resolve(tabItem.id);
        });
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.teamId = yield this.getTeamId(args);
                this.channelId = yield this.getChannelId(args);
                const tabId = yield this.getTabId(args);
                const endpoint = `${this.resource}/v1.0/teams/${encodeURIComponent(this.teamId)}/channels/${encodeURIComponent(this.channelId)}/tabs/${encodeURIComponent(tabId)}`;
                const requestOptions = {
                    url: endpoint,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    responseType: 'json'
                };
                const res = yield request_1.default.get(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsTabGetCommand_instances = new WeakSet(), _TeamsTabGetCommand_initTelemetry = function _TeamsTabGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            teamId: typeof args.options.teamId !== 'undefined',
            teamName: typeof args.options.teamName !== 'undefined',
            channelId: typeof args.options.channelId !== 'undefined',
            channelName: typeof args.options.channelName !== 'undefined',
            tabId: typeof args.options.tabId !== 'undefined',
            tabName: typeof args.options.tabName !== 'undefined'
        });
    });
}, _TeamsTabGetCommand_initOptions = function _TeamsTabGetCommand_initOptions() {
    this.options.unshift({
        option: '--teamId [teamId]'
    }, {
        option: '--teamName [teamName]'
    }, {
        option: '--channelId [channelId]'
    }, {
        option: '--channelName [channelName]'
    }, {
        option: '-i, --tabId [tabId]'
    }, {
        option: '-n, --tabName [tabName]'
    });
}, _TeamsTabGetCommand_initValidators = function _TeamsTabGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.teamId && !validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (args.options.channelId && !validation_1.validation.isValidTeamsChannelId(args.options.channelId)) {
            return `${args.options.channelId} is not a valid Teams ChannelId`;
        }
        if (args.options.tabId && !validation_1.validation.isValidGuid(args.options.tabId)) {
            return `${args.options.tabId} is not a valid GUID`;
        }
        return true;
    }));
}, _TeamsTabGetCommand_initOptionSets = function _TeamsTabGetCommand_initOptionSets() {
    this.optionSets.push(['teamId', 'teamName'], ['channelId', 'channelName'], ['tabId', 'tabName']);
};
module.exports = new TeamsTabGetCommand();
