// @pnp/cli-microsoft365/dist/m365/teams/commands/tab/tab-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsTabRemoveCommand_instances, _TeamsTabRemoveCommand_initTelemetry, _TeamsTabRemoveCommand_initOptions, _TeamsTabRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsTabRemoveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsTabRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsTabRemoveCommand_instances, "m", _TeamsTabRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsTabRemoveCommand_instances, "m", _TeamsTabRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsTabRemoveCommand_instances, "m", _TeamsTabRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.TAB_REMOVE;
    }
    get description() {
        return "Removes a tab from the specified channel";
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeTab = () => __awaiter(this, void 0, void 0, function* () {
                const requestOptions = {
                    url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}/channels/${args.options.channelId}/tabs/${encodeURIComponent(args.options.tabId)}`,
                    headers: {
                        accept: "application/json;odata.metadata=none"
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeTab();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the tab with id ${args.options.tabId} from channel ${args.options.channelId} in team ${args.options.teamId}?`
                });
                if (result.continue) {
                    yield removeTab();
                }
            }
        });
    }
}
_TeamsTabRemoveCommand_instances = new WeakSet(), _TeamsTabRemoveCommand_initTelemetry = function _TeamsTabRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!!args.options.confirm).toString()
        });
    });
}, _TeamsTabRemoveCommand_initOptions = function _TeamsTabRemoveCommand_initOptions() {
    this.options.unshift({
        option: "-i, --teamId <teamId>"
    }, {
        option: "-c, --channelId <channelId>"
    }, {
        option: "-t, --tabId <tabId>"
    }, {
        option: "--confirm"
    });
}, _TeamsTabRemoveCommand_initValidators = function _TeamsTabRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (!validation_1.validation.isValidTeamsChannelId(args.options.channelId)) {
            return `${args.options.channelId} is not a valid Teams ChannelId`;
        }
        if (!validation_1.validation.isValidGuid(args.options.tabId)) {
            return `${args.options.tabId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new TeamsTabRemoveCommand();
