// @pnp/cli-microsoft365/dist/m365/teams/commands/tab/tab-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsTabListCommand_instances, _TeamsTabListCommand_initOptions, _TeamsTabListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const odata_1 = require("../../../../utils/odata");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsTabListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsTabListCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsTabListCommand_instances, "m", _TeamsTabListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsTabListCommand_instances, "m", _TeamsTabListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.TAB_LIST;
    }
    get description() {
        return 'Lists tabs in the specified Microsoft Teams channel';
    }
    defaultProperties() {
        return ['id', 'displayName', 'teamsAppTabId'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const endpoint = `${this.resource}/v1.0/teams/${args.options.teamId}/channels/${encodeURIComponent(args.options.channelId)}/tabs?$expand=teamsApp`;
            try {
                const items = yield odata_1.odata.getAllItems(endpoint);
                items.forEach(i => {
                    i.teamsAppTabId = i.teamsApp.id;
                });
                logger.log(items);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsTabListCommand_instances = new WeakSet(), _TeamsTabListCommand_initOptions = function _TeamsTabListCommand_initOptions() {
    this.options.unshift({
        option: '-i, --teamId <teamId>'
    }, {
        option: '-c, --channelId <channelId>'
    });
}, _TeamsTabListCommand_initValidators = function _TeamsTabListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (!validation_1.validation.isValidTeamsChannelId(args.options.channelId)) {
            return `${args.options.channelId} is not a valid Teams ChannelId`;
        }
        return true;
    }));
};
module.exports = new TeamsTabListCommand();
