// @pnp/cli-microsoft365/dist/m365/pp/commands/gateway/gateway-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PpGatewayGetCommand_instances, _PpGatewayGetCommand_initOptions, _PpGatewayGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const PowerBICommand_1 = require("../../../base/PowerBICommand");
const commands_1 = require("../../commands");
class PpGatewayGetCommand extends PowerBICommand_1.default {
    constructor() {
        super();
        _PpGatewayGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _PpGatewayGetCommand_instances, "m", _PpGatewayGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PpGatewayGetCommand_instances, "m", _PpGatewayGetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.GATEWAY_GET;
    }
    get description() {
        return "Get information about the specified gateway.";
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const gateway = yield this.getGateway(args.options.id);
                logger.log(gateway);
            }
            catch (error) {
                this.handleRejectedODataJsonPromise(error);
            }
        });
    }
    getGateway(gatewayId) {
        const requestOptions = {
            url: `${this.resource}/v1.0/myorg/gateways/${encodeURIComponent(gatewayId)}`,
            headers: {
                accept: "application/json;odata.metadata=none"
            },
            responseType: "json"
        };
        return request_1.default.get(requestOptions);
    }
}
_PpGatewayGetCommand_instances = new WeakSet(), _PpGatewayGetCommand_initOptions = function _PpGatewayGetCommand_initOptions() {
    this.options.unshift({
        option: "-i, --id <id>"
    });
}, _PpGatewayGetCommand_initValidators = function _PpGatewayGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new PpGatewayGetCommand();
