// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-rights-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignRightsListCommand_instances, _SpoSiteDesignRightsListCommand_initOptions, _SpoSiteDesignRightsListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignRightsListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignRightsListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRightsListCommand_instances, "m", _SpoSiteDesignRightsListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRightsListCommand_instances, "m", _SpoSiteDesignRightsListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_RIGHTS_LIST;
    }
    get description() {
        return 'Gets a list of principals that have access to a site design';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                const requestDigest = yield spo_1.spo.getRequestDigest(spoUrl);
                const requestOptions = {
                    url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GetSiteDesignRights`,
                    headers: {
                        'X-RequestDigest': requestDigest.FormDigestValue,
                        'content-type': 'application/json;charset=utf-8',
                        accept: 'application/json;odata=nometadata'
                    },
                    data: { id: args.options.id },
                    responseType: 'json'
                };
                const res = yield request_1.default.post(requestOptions);
                logger.log(res.value.map(p => {
                    p.Rights = p.Rights === "1" ? "View" : p.Rights;
                    return p;
                }));
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteDesignRightsListCommand_instances = new WeakSet(), _SpoSiteDesignRightsListCommand_initOptions = function _SpoSiteDesignRightsListCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    });
}, _SpoSiteDesignRightsListCommand_initValidators = function _SpoSiteDesignRightsListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new SpoSiteDesignRightsListCommand();
