// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-task-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignTaskListCommand_instances, _SpoSiteDesignTaskListCommand_initOptions, _SpoSiteDesignTaskListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignTaskListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignTaskListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignTaskListCommand_instances, "m", _SpoSiteDesignTaskListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignTaskListCommand_instances, "m", _SpoSiteDesignTaskListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_TASK_LIST;
    }
    get description() {
        return 'Lists site designs scheduled for execution on the specified site';
    }
    defaultProperties() {
        return ['ID', 'SiteDesignID', 'LogonName'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${args.options.webUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GetSiteDesignTasks`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.post(requestOptions);
                logger.log(res.value);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteDesignTaskListCommand_instances = new WeakSet(), _SpoSiteDesignTaskListCommand_initOptions = function _SpoSiteDesignTaskListCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    });
}, _SpoSiteDesignTaskListCommand_initValidators = function _SpoSiteDesignTaskListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.webUrl); }));
};
module.exports = new SpoSiteDesignTaskListCommand();
