// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-run-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignRunListCommand_instances, _SpoSiteDesignRunListCommand_initTelemetry, _SpoSiteDesignRunListCommand_initOptions, _SpoSiteDesignRunListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignRunListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignRunListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRunListCommand_instances, "m", _SpoSiteDesignRunListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRunListCommand_instances, "m", _SpoSiteDesignRunListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRunListCommand_instances, "m", _SpoSiteDesignRunListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_RUN_LIST;
    }
    get description() {
        return 'Lists information about site designs applied to the specified site';
    }
    defaultProperties() {
        return ['ID', 'SiteDesignID', 'SiteDesignTitle', 'StartTime'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {};
            if (args.options.siteDesignId) {
                data.siteDesignId = args.options.siteDesignId;
            }
            const requestOptions = {
                url: `${args.options.webUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GetSiteDesignRun`,
                headers: {
                    accept: 'application/json;odata=nometadata',
                    'content-type': 'application/json;odata=nometadata'
                },
                data: data,
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.post(requestOptions);
                if (args.options.output !== 'json') {
                    res.value.forEach(d => {
                        d.StartTime = new Date(parseInt(d.StartTime)).toLocaleString();
                    });
                }
                logger.log(res.value);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteDesignRunListCommand_instances = new WeakSet(), _SpoSiteDesignRunListCommand_initTelemetry = function _SpoSiteDesignRunListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            siteDesignId: typeof args.options.siteDesignId !== 'undefined'
        });
    });
}, _SpoSiteDesignRunListCommand_initOptions = function _SpoSiteDesignRunListCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-i, --siteDesignId [siteDesignId]'
    });
}, _SpoSiteDesignRunListCommand_initValidators = function _SpoSiteDesignRunListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.siteDesignId) {
            if (!validation_1.validation.isValidGuid(args.options.siteDesignId)) {
                return `${args.options.siteDesignId} is not a valid GUID`;
            }
        }
        return true;
    }));
};
module.exports = new SpoSiteDesignRunListCommand();
