// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-task-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignTaskRemoveCommand_instances, _SpoSiteDesignTaskRemoveCommand_initTelemetry, _SpoSiteDesignTaskRemoveCommand_initOptions, _SpoSiteDesignTaskRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignTaskRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignTaskRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignTaskRemoveCommand_instances, "m", _SpoSiteDesignTaskRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignTaskRemoveCommand_instances, "m", _SpoSiteDesignTaskRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignTaskRemoveCommand_instances, "m", _SpoSiteDesignTaskRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_TASK_REMOVE;
    }
    get description() {
        return 'Removes the specified site design scheduled for execution';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeSiteDesignTask = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                    const requestOptions = {
                        url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.RemoveSiteDesignTask`,
                        headers: {
                            accept: 'application/json;odata=nometadata'
                        },
                        data: {
                            taskId: args.options.taskId
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeSiteDesignTask();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the site design task ${args.options.taskId}?`
                });
                if (result.continue) {
                    yield removeSiteDesignTask();
                }
            }
        });
    }
}
_SpoSiteDesignTaskRemoveCommand_instances = new WeakSet(), _SpoSiteDesignTaskRemoveCommand_initTelemetry = function _SpoSiteDesignTaskRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: args.options.confirm || false
        });
    });
}, _SpoSiteDesignTaskRemoveCommand_initOptions = function _SpoSiteDesignTaskRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --taskId <taskId>'
    }, {
        option: '--confirm'
    });
}, _SpoSiteDesignTaskRemoveCommand_initValidators = function _SpoSiteDesignTaskRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.taskId)) {
            return `${args.options.taskId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new SpoSiteDesignTaskRemoveCommand();
