// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignRemoveCommand_instances, _SpoSiteDesignRemoveCommand_initTelemetry, _SpoSiteDesignRemoveCommand_initOptions, _SpoSiteDesignRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRemoveCommand_instances, "m", _SpoSiteDesignRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRemoveCommand_instances, "m", _SpoSiteDesignRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRemoveCommand_instances, "m", _SpoSiteDesignRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_REMOVE;
    }
    get description() {
        return 'Removes the specified site design';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeSiteDesign = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                    const requestDigest = yield spo_1.spo.getRequestDigest(spoUrl);
                    const requestOptions = {
                        url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.DeleteSiteDesign`,
                        headers: {
                            'X-RequestDigest': requestDigest.FormDigestValue,
                            'content-type': 'application/json;charset=utf-8',
                            accept: 'application/json;odata=nometadata'
                        },
                        data: { id: args.options.id },
                        responseType: 'json'
                    };
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeSiteDesign();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the site design ${args.options.id}?`
                });
                if (result.continue) {
                    yield removeSiteDesign();
                }
            }
        });
    }
}
_SpoSiteDesignRemoveCommand_instances = new WeakSet(), _SpoSiteDesignRemoveCommand_initTelemetry = function _SpoSiteDesignRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: args.options.confirm || false
        });
    });
}, _SpoSiteDesignRemoveCommand_initOptions = function _SpoSiteDesignRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '--confirm'
    });
}, _SpoSiteDesignRemoveCommand_initValidators = function _SpoSiteDesignRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new SpoSiteDesignRemoveCommand();
