// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-rights-grant.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignRightsGrantCommand_instances, _SpoSiteDesignRightsGrantCommand_initOptions, _SpoSiteDesignRightsGrantCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignRightsGrantCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignRightsGrantCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRightsGrantCommand_instances, "m", _SpoSiteDesignRightsGrantCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignRightsGrantCommand_instances, "m", _SpoSiteDesignRightsGrantCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_RIGHTS_GRANT;
    }
    get description() {
        return 'Grants access to a site design for one or more principals';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                const requestDigest = yield spo_1.spo.getRequestDigest(spoUrl);
                const grantedRights = '1';
                const requestOptions = {
                    url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GrantSiteDesignRights`,
                    headers: {
                        'X-RequestDigest': requestDigest.FormDigestValue,
                        'content-type': 'application/json;charset=utf-8',
                        accept: 'application/json;odata=nometadata'
                    },
                    data: {
                        id: args.options.id,
                        principalNames: args.options.principals.split(',').map(p => p.trim()),
                        grantedRights: grantedRights
                    },
                    responseType: 'json'
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteDesignRightsGrantCommand_instances = new WeakSet(), _SpoSiteDesignRightsGrantCommand_initOptions = function _SpoSiteDesignRightsGrantCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '-p, --principals <principals>'
    }, {
        option: '-r, --rights <rights>',
        autocomplete: ['View']
    });
}, _SpoSiteDesignRightsGrantCommand_initValidators = function _SpoSiteDesignRightsGrantCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        if (args.options.rights !== 'View') {
            return `${args.options.rights} is not a valid rights value. Allowed values View`;
        }
        return true;
    }));
};
module.exports = new SpoSiteDesignRightsGrantCommand();
