// @pnp/cli-microsoft365/dist/m365/spo/commands/sitedesign/sitedesign-apply.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteDesignApplyCommand_instances, _SpoSiteDesignApplyCommand_initTelemetry, _SpoSiteDesignApplyCommand_initOptions, _SpoSiteDesignApplyCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteDesignApplyCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteDesignApplyCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteDesignApplyCommand_instances, "m", _SpoSiteDesignApplyCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignApplyCommand_instances, "m", _SpoSiteDesignApplyCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteDesignApplyCommand_instances, "m", _SpoSiteDesignApplyCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITEDESIGN_APPLY;
    }
    get description() {
        return 'Applies a site design to an existing site collection';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                const requestBody = {
                    siteDesignId: args.options.id,
                    webUrl: args.options.webUrl
                };
                const requestOptions = {
                    url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.${args.options.asTask ? 'AddSiteDesignTask' : 'ApplySiteDesign'}`,
                    headers: {
                        'content-type': 'application/json;charset=utf-8',
                        accept: 'application/json;odata=nometadata'
                    },
                    data: requestBody,
                    responseType: 'json'
                };
                const res = yield request_1.default.post(requestOptions);
                if (res.value) {
                    logger.log(res.value);
                }
                else {
                    logger.log(res);
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteDesignApplyCommand_instances = new WeakSet(), _SpoSiteDesignApplyCommand_initTelemetry = function _SpoSiteDesignApplyCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            asTask: args.options.asTask || false
        });
    });
}, _SpoSiteDesignApplyCommand_initOptions = function _SpoSiteDesignApplyCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--asTask'
    });
}, _SpoSiteDesignApplyCommand_initValidators = function _SpoSiteDesignApplyCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return validation_1.validation.isValidSharePointUrl(args.options.webUrl);
    }));
};
module.exports = new SpoSiteDesignApplyCommand();
