// @pnp/cli-microsoft365/dist/m365/spo/commands/cdn/cdn-origin-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoCdnOriginAddCommand_instances, _SpoCdnOriginAddCommand_initTelemetry, _SpoCdnOriginAddCommand_initOptions, _SpoCdnOriginAddCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../../../config");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const spo_1 = require("../../../../utils/spo");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoCdnOriginAddCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoCdnOriginAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoCdnOriginAddCommand_instances, "m", _SpoCdnOriginAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoCdnOriginAddCommand_instances, "m", _SpoCdnOriginAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoCdnOriginAddCommand_instances, "m", _SpoCdnOriginAddCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.CDN_ORIGIN_ADD;
    }
    get description() {
        return 'Adds CDN origin to the current SharePoint Online tenant';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const cdnTypeString = args.options.type || 'Public';
            const cdnType = cdnTypeString === 'Private' ? 1 : 0;
            let spoAdminUrl = '';
            let tenantId = '';
            try {
                tenantId = yield spo_1.spo.getTenantId(logger, this.debug);
                spoAdminUrl = yield spo_1.spo.getSpoAdminUrl(logger, this.debug);
                const reqDigest = yield spo_1.spo.getRequestDigest(spoAdminUrl);
                if (this.verbose) {
                    logger.logToStderr(`Adding origin ${args.options.origin} to the ${(cdnType === 1 ? 'Private' : 'Public')} CDN. Please wait, this might take a moment...`);
                }
                const requestOptions = {
                    url: `${spoAdminUrl}/_vti_bin/client.svc/ProcessQuery`,
                    headers: {
                        'X-RequestDigest': reqDigest.FormDigestValue
                    },
                    data: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config_1.default.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><Method Name="AddTenantCdnOrigin" Id="27" ObjectPathId="23"><Parameters><Parameter Type="Enum">${cdnType}</Parameter><Parameter Type="String">${formatting_1.formatting.escapeXml(args.options.origin)}</Parameter></Parameters></Method></Actions><ObjectPaths><Identity Id="23" Name="${tenantId}" /></ObjectPaths></Request>`
                };
                const res = yield request_1.default.post(requestOptions);
                const json = JSON.parse(res);
                const response = json[0];
                if (response.ErrorInfo) {
                    throw response.ErrorInfo.ErrorMessage;
                }
            }
            catch (err) {
                this.handleRejectedPromise(err);
            }
        });
    }
}
_SpoCdnOriginAddCommand_instances = new WeakSet(), _SpoCdnOriginAddCommand_initTelemetry = function _SpoCdnOriginAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            cdnType: args.options.type || 'Public'
        });
    });
}, _SpoCdnOriginAddCommand_initOptions = function _SpoCdnOriginAddCommand_initOptions() {
    this.options.unshift({
        option: '-t, --type [type]',
        autocomplete: ['Public', 'Private']
    }, {
        option: '-r, --origin <origin>'
    });
}, _SpoCdnOriginAddCommand_initValidators = function _SpoCdnOriginAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.type) {
            if (args.options.type !== 'Public' &&
                args.options.type !== 'Private') {
                return `${args.options.type} is not a valid CDN type. Allowed values are Public|Private`;
            }
        }
        return true;
    }));
};
module.exports = new SpoCdnOriginAddCommand();
