// @pnp/cli-microsoft365/dist/m365/teams/commands/team/team-archive.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsTeamArchiveCommand_instances, _TeamsTeamArchiveCommand_initTelemetry, _TeamsTeamArchiveCommand_initOptions, _TeamsTeamArchiveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsTeamArchiveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsTeamArchiveCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsTeamArchiveCommand_instances, "m", _TeamsTeamArchiveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsTeamArchiveCommand_instances, "m", _TeamsTeamArchiveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsTeamArchiveCommand_instances, "m", _TeamsTeamArchiveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.TEAM_ARCHIVE;
    }
    get description() {
        return 'Archives specified Microsoft Teams team';
    }
    getTeamId(args) {
        if (args.options.id) {
            return Promise.resolve(args.options.id);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.name)
            .then(group => {
            if (group.resourceProvisioningOptions.indexOf('Team') === -1) {
                return Promise.reject(`The specified team does not exist in the Microsoft Teams`);
            }
            return group.id;
        });
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.teamId) {
                args.options.id = args.options.teamId;
                this.warn(logger, `Option 'teamId' is deprecated. Please use 'id' instead.`);
            }
            const siteReadOnlyForMembers = args.options.shouldSetSpoSiteReadOnlyForMembers === true;
            try {
                const teamId = yield this.getTeamId(args);
                const requestOptions = {
                    url: `${this.resource}/v1.0/teams/${encodeURIComponent(teamId)}/archive`,
                    headers: {
                        'content-type': 'application/json;odata=nometadata',
                        'accept': 'application/json;odata.metadata=none'
                    },
                    responseType: 'json',
                    data: {
                        shouldSetSpoSiteReadOnlyForMembers: siteReadOnlyForMembers
                    }
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsTeamArchiveCommand_instances = new WeakSet(), _TeamsTeamArchiveCommand_initTelemetry = function _TeamsTeamArchiveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: typeof args.options.id !== 'undefined',
            name: typeof args.options.name !== 'undefined',
            shouldSetSpoSiteReadOnlyForMembers: args.options.shouldSetSpoSiteReadOnlyForMembers === true,
            teamId: typeof args.options.teamId !== 'undefined'
        });
    });
}, _TeamsTeamArchiveCommand_initOptions = function _TeamsTeamArchiveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id [id]'
    }, {
        option: '-n, --name [name]'
    }, {
        option: '--teamId [teamId]'
    }, {
        option: '--shouldSetSpoSiteReadOnlyForMembers'
    });
}, _TeamsTeamArchiveCommand_initValidators = function _TeamsTeamArchiveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!args.options.id && !args.options.name && !args.options.teamId) {
            return 'Specify either id or name';
        }
        if (args.options.name && (args.options.id || args.options.teamId)) {
            return 'Specify either id or name but not both';
        }
        if (args.options.teamId && !validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (args.options.id && !validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new TeamsTeamArchiveCommand();
