// @pnp/cli-microsoft365/dist/m365/pp/commands/card/card-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PpCardGetCommand_instances, _PpCardGetCommand_initTelemetry, _PpCardGetCommand_initOptions, _PpCardGetCommand_initOptionSets, _PpCardGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const powerPlatform_1 = require("../../../../utils/powerPlatform");
const PowerPlatformCommand_1 = require("../../../base/PowerPlatformCommand");
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
class PpCardGetCommand extends PowerPlatformCommand_1.default {
    constructor() {
        super();
        _PpCardGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _PpCardGetCommand_instances, "m", _PpCardGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PpCardGetCommand_instances, "m", _PpCardGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PpCardGetCommand_instances, "m", _PpCardGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _PpCardGetCommand_instances, "m", _PpCardGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.CARD_GET;
    }
    get description() {
        return 'Get specific Microsoft Power Platform card in the specified Power Platform environment.';
    }
    defaultProperties() {
        return ['name', 'cardid', 'publishdate', 'createdon', 'modifiedon'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving a card '${args.options.id || args.options.name}'...`);
            }
            try {
                const dynamicsApiUrl = yield powerPlatform_1.powerPlatform.getDynamicsInstanceApiUrl(args.options.environment, args.options.asAdmin);
                const res = yield this.getCard(dynamicsApiUrl, args.options);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getCard(dynamicsApiUrl, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            if (options.id) {
                requestOptions.url = `${dynamicsApiUrl}/api/data/v9.1/cards(${options.id})`;
                const result = yield request_1.default.get(requestOptions);
                return result;
            }
            requestOptions.url = `${dynamicsApiUrl}/api/data/v9.1/cards?$filter=name eq '${options.name}'`;
            const result = yield request_1.default.get(requestOptions);
            if (result.value.length === 0) {
                throw `The specified card '${options.name}' does not exist.`;
            }
            if (result.value.length > 1) {
                throw `Multiple cards with name '${options.name}' found: ${result.value.map(x => x.cardid).join(',')}`;
            }
            return result.value[0];
        });
    }
}
_PpCardGetCommand_instances = new WeakSet(), _PpCardGetCommand_initTelemetry = function _PpCardGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: typeof args.options.id !== 'undefined',
            name: typeof args.options.name !== 'undefined',
            asAdmin: !!args.options.asAdmin
        });
    });
}, _PpCardGetCommand_initOptions = function _PpCardGetCommand_initOptions() {
    this.options.unshift({
        option: '-e, --environment <environment>'
    }, {
        option: '-i, --id [id]'
    }, {
        option: '-n, --name [name]'
    }, {
        option: '-a, --asAdmin'
    });
}, _PpCardGetCommand_initOptionSets = function _PpCardGetCommand_initOptionSets() {
    this.optionSets.push(['id', 'name']);
}, _PpCardGetCommand_initValidators = function _PpCardGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.id && !validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new PpCardGetCommand();
