// @pnp/cli-microsoft365/dist/m365/yammer/commands/message/message-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YammerMessageAddCommand_instances, _YammerMessageAddCommand_initTelemetry, _YammerMessageAddCommand_initOptions, _YammerMessageAddCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const YammerCommand_1 = require("../../../base/YammerCommand");
const commands_1 = require("../../commands");
class YammerMessageAddCommand extends YammerCommand_1.default {
    constructor() {
        super();
        _YammerMessageAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _YammerMessageAddCommand_instances, "m", _YammerMessageAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _YammerMessageAddCommand_instances, "m", _YammerMessageAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _YammerMessageAddCommand_instances, "m", _YammerMessageAddCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.MESSAGE_ADD;
    }
    get description() {
        return 'Posts a Yammer network message on behalf of the current user';
    }
    defaultProperties() {
        return ['id'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/v1/messages.json`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json;odata=nometadata'
                },
                responseType: 'json',
                data: {
                    body: args.options.body,
                    replied_to_id: args.options.repliedToId,
                    direct_to_user_ids: args.options.directToUserIds,
                    group_id: args.options.groupId,
                    network_id: args.options.networkId
                }
            };
            try {
                const res = yield request_1.default.post(requestOptions);
                let result = null;
                if (res.messages && res.messages.length === 1) {
                    result = res.messages[0];
                }
                logger.log(result);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_YammerMessageAddCommand_instances = new WeakSet(), _YammerMessageAddCommand_initTelemetry = function _YammerMessageAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            repliedToId: args.options.repliedToId !== undefined,
            directToUserIds: args.options.directToUserIds !== undefined,
            groupId: args.options.groupId !== undefined,
            networkId: args.options.networkId !== undefined
        });
    });
}, _YammerMessageAddCommand_initOptions = function _YammerMessageAddCommand_initOptions() {
    this.options.unshift({
        option: '-b, --body <body>'
    }, {
        option: '-r, --repliedToId [repliedToId]'
    }, {
        option: '-d, --directToUserIds [directToUserIds]'
    }, {
        option: '--groupId [groupId]'
    }, {
        option: '--networkId [networkId]'
    });
}, _YammerMessageAddCommand_initValidators = function _YammerMessageAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.groupId && typeof args.options.groupId !== 'number') {
            return `${args.options.groupId} is not a number`;
        }
        if (args.options.networkId && typeof args.options.networkId !== 'number') {
            return `${args.options.networkId} is not a number`;
        }
        if (args.options.repliedToId && typeof args.options.repliedToId !== 'number') {
            return `${args.options.repliedToId} is not a number`;
        }
        if (args.options.groupId === undefined &&
            args.options.directToUserIds === undefined &&
            args.options.repliedToId === undefined) {
            return "You must either specify groupId, repliedToId or directToUserIds";
        }
        return true;
    }));
};
module.exports = new YammerMessageAddCommand();
