// @pnp/cli-microsoft365/dist/m365/yammer/commands/message/message-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YammerMessageGetCommand_instances, _YammerMessageGetCommand_initOptions, _YammerMessageGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const YammerCommand_1 = require("../../../base/YammerCommand");
const commands_1 = require("../../commands");
class YammerMessageGetCommand extends YammerCommand_1.default {
    constructor() {
        super();
        _YammerMessageGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _YammerMessageGetCommand_instances, "m", _YammerMessageGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _YammerMessageGetCommand_instances, "m", _YammerMessageGetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.MESSAGE_GET;
    }
    get description() {
        return 'Returns a Yammer message';
    }
    defaultProperties() {
        return ['id', 'sender_id', 'replied_to_id', 'thread_id', 'group_id', 'created_at', 'direct_message', 'system_message', 'privacy', 'message_type', 'content_excerpt'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/v1/messages/${args.options.id}.json`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_YammerMessageGetCommand_instances = new WeakSet(), _YammerMessageGetCommand_initOptions = function _YammerMessageGetCommand_initOptions() {
    this.options.unshift({
        option: '--id <id>'
    });
}, _YammerMessageGetCommand_initValidators = function _YammerMessageGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (typeof args.options.id !== 'number') {
            return `${args.options.id} is not a number`;
        }
        return true;
    }));
};
module.exports = new YammerMessageGetCommand();
