// @pnp/cli-microsoft365/dist/m365/spo/commands/tenant/tenant-appcatalogurl-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoTenantAppCatalogUrlGetCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.TENANT_APPCATALOGURL_GET;
    }
    get description() {
        return 'Gets the URL of the tenant app catalog';
    }
    commandAction(logger) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                const requestOptions = {
                    url: `${spoUrl}/_api/SP_TenantSettings_Current`,
                    headers: {
                        accept: 'application/json;odata=nometadata'
                    }
                };
                const res = yield request_1.default.get(requestOptions);
                const json = JSON.parse(res);
                if (json.CorporateCatalogUrl) {
                    logger.log(json.CorporateCatalogUrl);
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr("Tenant app catalog is not configured.");
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
module.exports = new SpoTenantAppCatalogUrlGetCommand();
