// @pnp/cli-microsoft365/dist/m365/todo/commands/task/task-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TodoTaskRemoveCommand_instances, _TodoTaskRemoveCommand_initTelemetry, _TodoTaskRemoveCommand_initOptions, _TodoTaskRemoveCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TodoTaskRemoveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TodoTaskRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _TodoTaskRemoveCommand_instances, "m", _TodoTaskRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TodoTaskRemoveCommand_instances, "m", _TodoTaskRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TodoTaskRemoveCommand_instances, "m", _TodoTaskRemoveCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.TASK_REMOVE;
    }
    get description() {
        return 'Removes the specified Microsoft To Do task';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const getToDoListId = () => __awaiter(this, void 0, void 0, function* () {
                if (args.options.listName) {
                    // Search list by its name
                    const requestOptions = {
                        url: `${this.resource}/v1.0/me/todo/lists?$filter=displayName eq '${escape(args.options.listName)}'`,
                        headers: {
                            accept: "application/json;odata.metadata=none"
                        },
                        responseType: 'json'
                    };
                    const response = yield request_1.default.get(requestOptions);
                    return response.value && response.value.length === 1 ? response.value[0].id : undefined;
                }
                return Promise.resolve(args.options.listId);
            });
            const removeToDoTask = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const toDoListId = yield getToDoListId();
                    if (!toDoListId) {
                        throw `The list ${args.options.listName} cannot be found`;
                    }
                    const requestOptions = {
                        url: `${this.resource}/v1.0/me/todo/lists/${toDoListId}/tasks/${args.options.id}`,
                        headers: {
                            accept: "application/json;odata.metadata=none"
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeToDoTask();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the task ${args.options.id} from list ${args.options.listId || args.options.listName}?`
                });
                if (result.continue) {
                    yield removeToDoTask();
                }
            }
        });
    }
}
_TodoTaskRemoveCommand_instances = new WeakSet(), _TodoTaskRemoveCommand_initTelemetry = function _TodoTaskRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listName: typeof args.options.listName !== 'undefined',
            listId: typeof args.options.listId !== 'undefined',
            confirm: typeof args.options.confirm !== 'undefined'
        });
    });
}, _TodoTaskRemoveCommand_initOptions = function _TodoTaskRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '--listName [listName]'
    }, {
        option: '--listId [listId]'
    }, {
        option: '--confirm'
    });
}, _TodoTaskRemoveCommand_initOptionSets = function _TodoTaskRemoveCommand_initOptionSets() {
    this.optionSets.push(['listName', 'listId']);
};
module.exports = new TodoTaskRemoveCommand();
