// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-doctor/rules/FN021004_PKG_spfx_devdeps_installed_as_devdeps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN021004_PKG_spfx_devdeps_installed_as_devdeps = void 0;
const JsonRule_1 = require("../../JsonRule");
const spfxDeps = require("../spfx-deps");
class FN021004_PKG_spfx_devdeps_installed_as_devdeps extends JsonRule_1.JsonRule {
    get id() {
        return 'FN021004';
    }
    get title() {
        return '';
    }
    get description() {
        return '';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './package.json';
    }
    get resolutionType() {
        return 'cmd';
    }
    visit(project, findings) {
        if (!project.version ||
            !project.packageJson ||
            !project.packageJson.dependencies) {
            return;
        }
        const projectDeps = Object.keys(project.packageJson.dependencies);
        projectDeps.forEach(dep => {
            if (!spfxDeps.devDeps.includes(dep)) {
                return;
            }
            const node = this.getAstNodeFromFile(project.packageJson, `dependencies.${dep}`);
            this.addFindingWithCustomInfo(`${dep} installed as a dependency`, `Package ${dep} is installed as a dependency. Install it as a devDependency instead`, [{
                    file: this.file,
                    resolution: `installDev ${dep}@${project.version}`,
                    position: this.getPositionFromNode(node)
                }], findings);
        });
    }
}
exports.FN021004_PKG_spfx_devdeps_installed_as_devdeps = FN021004_PKG_spfx_devdeps_installed_as_devdeps;
