// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-doctor/rules/FN021009_PKG_no_duplicate_oui_deps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN021009_PKG_no_duplicate_oui_deps = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN021009_PKG_no_duplicate_oui_deps extends JsonRule_1.JsonRule {
    get id() {
        return 'FN021009';
    }
    get title() {
        return '@fluentui/react and office-ui-fabric-react installed in the project';
    }
    get description() {
        return '@fluentui/react and office-ui-fabric-react installed in the project. Consider uninstalling @fluentui/react';
    }
    get severity() {
        return 'Optional';
    }
    get file() {
        return './package.json';
    }
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'cmd';
    }
    visit(project, findings) {
        if (!project.packageJson) {
            return;
        }
        const projectDeps = [];
        const projectDevDeps = [];
        if (project.packageJson.dependencies) {
            projectDeps.push(...Object.keys(project.packageJson.dependencies));
        }
        if (project.packageJson.devDependencies) {
            projectDevDeps.push(...Object.keys(project.packageJson.devDependencies));
        }
        const allDeps = projectDeps.concat(projectDevDeps);
        if (allDeps.length === 0) {
            return;
        }
        if (!allDeps.includes('office-ui-fabric-react') ||
            !allDeps.includes('@fluentui/react')) {
            return;
        }
        const dependencyType = projectDeps.includes('@fluentui/react') ? 'dependencies' : 'devDependencies';
        const node = this.getAstNodeFromFile(project.packageJson, `${dependencyType}.@fluentui/react`);
        this.addFindingWithCustomInfo(this.title, this.description, [{
                file: this.file,
                resolution: `${dependencyType === 'dependencies' ? 'uninstall' : 'uninstallDev'} @fluentui/react`,
                position: this.getPositionFromNode(node)
            }], findings);
    }
}
exports.FN021009_PKG_no_duplicate_oui_deps = FN021009_PKG_no_duplicate_oui_deps;
