// @pnp/cli-microsoft365/dist/m365/planner/commands/plan/plan-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PlannerPlanListCommand_instances, _PlannerPlanListCommand_initTelemetry, _PlannerPlanListCommand_initOptions, _PlannerPlanListCommand_initValidators, _PlannerPlanListCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../../../Auth");
const accessToken_1 = require("../../../../utils/accessToken");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const planner_1 = require("../../../../utils/planner");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class PlannerPlanListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _PlannerPlanListCommand_instances.add(this);
        __classPrivateFieldGet(this, _PlannerPlanListCommand_instances, "m", _PlannerPlanListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PlannerPlanListCommand_instances, "m", _PlannerPlanListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PlannerPlanListCommand_instances, "m", _PlannerPlanListCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _PlannerPlanListCommand_instances, "m", _PlannerPlanListCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.PLAN_LIST;
    }
    get description() {
        return 'Returns a list of plans associated with a specified group';
    }
    defaultProperties() {
        return ['id', 'title', 'createdDateTime', 'owner'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (accessToken_1.accessToken.isAppOnlyAccessToken(Auth_1.default.service.accessTokens[this.resource].accessToken)) {
                this.handleError('This command does not support application permissions.');
                return;
            }
            try {
                const groupId = yield this.getGroupId(args);
                const result = yield planner_1.planner.getPlansByGroupId(groupId);
                if (result && result.length > 0) {
                    logger.log(result);
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getGroupId(args) {
        if (args.options.ownerGroupId) {
            return Promise.resolve(args.options.ownerGroupId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.ownerGroupName)
            .then(group => group.id);
    }
}
_PlannerPlanListCommand_instances = new WeakSet(), _PlannerPlanListCommand_initTelemetry = function _PlannerPlanListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            ownerGroupId: typeof args.options.ownerGroupId !== 'undefined',
            ownerGroupName: typeof args.options.ownerGroupName !== 'undefined'
        });
    });
}, _PlannerPlanListCommand_initOptions = function _PlannerPlanListCommand_initOptions() {
    this.options.unshift({
        option: "--ownerGroupId [ownerGroupId]"
    }, {
        option: "--ownerGroupName [ownerGroupName]"
    });
}, _PlannerPlanListCommand_initValidators = function _PlannerPlanListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.ownerGroupId && !validation_1.validation.isValidGuid(args.options.ownerGroupId)) {
            return `${args.options.ownerGroupId} is not a valid GUID`;
        }
        return true;
    }));
}, _PlannerPlanListCommand_initOptionSets = function _PlannerPlanListCommand_initOptionSets() {
    this.optionSets.push(['ownerGroupId', 'ownerGroupName']);
};
module.exports = new PlannerPlanListCommand();
