// @pnp/cli-microsoft365/dist/m365/planner/commands/tenant/tenant-settings-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const PlannerCommand_1 = require("../../../base/PlannerCommand");
const commands_1 = require("../../commands");
class PlannerTenantSettingsListCommand extends PlannerCommand_1.default {
    get name() {
        return commands_1.default.TENANT_SETTINGS_LIST;
    }
    get description() {
        return 'Lists the Microsoft Planner configuration of the tenant';
    }
    defaultProperties() {
        return ['isPlannerAllowed', 'allowCalendarSharing', 'allowTenantMoveWithDataLoss', 'allowTenantMoveWithDataMigration', 'allowRosterCreation', 'allowPlannerMobilePushNotifications'];
    }
    commandAction(logger) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/taskAPI/tenantAdminSettings/Settings`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            try {
                const result = yield request_1.default.get(requestOptions);
                logger.log(result);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
module.exports = new PlannerTenantSettingsListCommand();
