// @pnp/cli-microsoft365/dist/m365/base/AzmgmtItemsListCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzmgmtItemsListCommand = void 0;
const url = require("url");
const request_1 = require("../../request");
const AzmgmtCommand_1 = require("./AzmgmtCommand");
class AzmgmtItemsListCommand extends AzmgmtCommand_1.default {
    constructor() {
        super();
        this.items = [];
    }
    getAllItems(_url, logger, firstRun) {
        return new Promise((resolve, reject) => {
            const requestOptions = {
                url: _url,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            request_1.default
                .get(requestOptions)
                .then((res) => {
                if (firstRun) {
                    this.items = [];
                }
                this.items = this.items.concat(res.value);
                if (res.nextLink) {
                    // when retrieving Flows as admin, the API returns nextLink
                    // pointing to https://emea.api.flow.microsoft.com:11777
                    // which leads to authentication exceptions because it's not an AAD
                    // resource for which we can retrieve an access token, so we need to
                    // rewrite it back to the API management URL
                    const nextLinkUrl = new url.URL(res.nextLink);
                    if (nextLinkUrl.host !== 'management.azure.com') {
                        nextLinkUrl.host = 'management.azure.com';
                        nextLinkUrl.port = '';
                    }
                    const nextLink = nextLinkUrl.href;
                    this
                        .getAllItems(nextLink, logger, false)
                        .then(() => {
                        resolve();
                    }, (err) => {
                        reject(err);
                    });
                }
                else {
                    resolve();
                }
            }, (err) => {
                reject(err);
            });
        });
    }
}
exports.AzmgmtItemsListCommand = AzmgmtItemsListCommand;
