// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/DependencyRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DependencyRule = void 0;
const semver_1 = require("semver");
const JsonRule_1 = require("../../JsonRule");
class DependencyRule extends JsonRule_1.JsonRule {
    constructor(packageName, packageVersion, isDevDep = false, isOptional = false, add = true) {
        super();
        this.packageName = packageName;
        this.packageVersion = packageVersion;
        this.isDevDep = isDevDep;
        this.isOptional = isOptional;
        this.add = add;
    }
    get title() {
        return this.packageName;
    }
    get description() {
        return `${(this.add ? 'Upgrade' : 'Remove')} SharePoint Framework ${(this.isDevDep ? 'dev ' : '')}dependency package ${this.packageName}`;
    }
    get resolution() {
        return this.add ?
            `${(this.isDevDep ? 'installDev' : 'install')} ${this.packageName}@${this.packageVersion}` :
            `${(this.isDevDep ? 'uninstallDev' : 'uninstall')} ${this.packageName}`;
    }
    get resolutionType() {
        return 'cmd';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './package.json';
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    customCondition(project) {
        return false;
    }
    visit(project, findings) {
        if (!project.packageJson) {
            return;
        }
        const projectDependencies = this.isDevDep ? project.packageJson.devDependencies : project.packageJson.dependencies;
        const versionEntry = projectDependencies ? projectDependencies[this.packageName] : '';
        const packageVersion = (0, semver_1.valid)(versionEntry);
        const versionRange = (0, semver_1.validRange)(versionEntry);
        if (this.add) {
            let jsonProperty = this.isDevDep ? 'devDependencies' : 'dependencies';
            if (versionEntry) {
                jsonProperty += `.${this.packageName}`;
                if (packageVersion) {
                    if ((0, semver_1.lt)(packageVersion, this.packageVersion)) {
                        const node = this.getAstNodeFromFile(project.packageJson, jsonProperty);
                        this.addFindingWithPosition(findings, node);
                    }
                }
                else {
                    if (versionRange) {
                        const node = this.getAstNodeFromFile(project.packageJson, jsonProperty);
                        this.addFindingWithPosition(findings, node);
                    }
                }
            }
            else {
                if (!this.isOptional || this.customCondition(project)) {
                    const node = this.getAstNodeFromFile(project.packageJson, jsonProperty);
                    this.addFindingWithCustomInfo(this.packageName, this.description.replace('Upgrade', 'Install'), [{
                            file: this.file,
                            resolution: this.resolution,
                            position: this.getPositionFromNode(node)
                        }], findings);
                }
            }
        }
        else {
            const jsonProperty = `${(this.isDevDep ? 'devDependencies' : 'dependencies')}.${this.packageName}`;
            if (versionEntry) {
                const node = this.getAstNodeFromFile(project.packageJson, jsonProperty);
                this.addFindingWithPosition(findings, node);
            }
        }
    }
}
exports.DependencyRule = DependencyRule;
