// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN004001_CFG_CA_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN004001_CFG_CA_schema = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN004001_CFG_CA_schema extends JsonRule_1.JsonRule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN004001';
    }
    get title() {
        return 'copy-assets.json schema';
    }
    get description() {
        return `Update copy-assets.json schema URL`;
    }
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/copy-assets.json';
    }
    visit(project, findings) {
        if (!project.copyAssetsJson) {
            return;
        }
        if (project.copyAssetsJson.$schema !== this.schema) {
            const node = this.getAstNodeFromFile(project.copyAssetsJson, '$schema');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN004001_CFG_CA_schema = FN004001_CFG_CA_schema;
