// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN003004_CFG_entries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN003004_CFG_entries = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN003004_CFG_entries extends JsonRule_1.JsonRule {
    get id() {
        return 'FN003004';
    }
    get title() {
        return `config.json entries`;
    }
    get description() {
        return `Remove the "entries" property in ${this.file}`;
    }
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/config.json';
    }
    visit(project, findings) {
        if (!project.configJson) {
            return;
        }
        const entries = project.configJson.entries;
        if (entries !== undefined) {
            const node = this.getAstNodeFromFile(project.configJson, 'entries');
            this.addFindingWithOccurrences([{
                    file: this.file,
                    resolution: JSON.stringify({ entries: entries }, null, 2),
                    position: this.getPositionFromNode(node)
                }], findings);
        }
    }
}
exports.FN003004_CFG_entries = FN003004_CFG_entries;
