// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN008003_CFG_TSL_preferConst.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN008003_CFG_TSL_preferConst = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN008003_CFG_TSL_preferConst extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN008003';
    }
    get title() {
        return 'tslint.json prefer-const';
    }
    get description() {
        return `Remove prefer-const from tslint.json`;
    }
    get resolution() {
        return `{
  "lintConfig": {
    "rules": {
      "prefer-const": true
    }
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/tslint.json';
    }
    visit(project, findings) {
        if (!project.tsLintJson) {
            return;
        }
        if (project.tsLintJson.lintConfig && project.tsLintJson.lintConfig.rules &&
            project.tsLintJson.lintConfig.rules["prefer-const"] === true) {
            const node = this.getAstNodeFromFile(project.tsLintJson, 'lintConfig.rules.prefer-const');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN008003_CFG_TSL_preferConst = FN008003_CFG_TSL_preferConst;
