// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN006005_CFG_PS_metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN006005_CFG_PS_metadata = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN006005_CFG_PS_metadata extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN006005';
    }
    get title() {
        return 'package-solution.json metadata';
    }
    get description() {
        return `In package-solution.json add metadata section`;
    }
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/package-solution.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.packageSolutionJson ||
            !project.packageSolutionJson.solution) {
            return;
        }
        if (!project.packageSolutionJson.solution.metadata) {
            const solutionDescription = `${(_a = project.packageJson) === null || _a === void 0 ? void 0 : _a.name} description`;
            const resolution = {
                solution: {
                    metadata: {
                        shortDescription: {
                            default: solutionDescription
                        },
                        longDescription: {
                            default: solutionDescription
                        },
                        screenshotPaths: [],
                        videoUrl: '',
                        categories: []
                    }
                }
            };
            const node = this.getAstNodeFromFile(project.packageSolutionJson, 'solution');
            this.addFindingWithOccurrences([{
                    file: this.file,
                    resolution: JSON.stringify(resolution, null, 2),
                    position: this.getPositionFromNode(node)
                }], findings);
        }
    }
}
exports.FN006005_CFG_PS_metadata = FN006005_CFG_PS_metadata;
