// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN007003_CFG_S_api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN007003_CFG_S_api = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN007003_CFG_S_api extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN007003';
    }
    get title() {
        return 'serve.json api';
    }
    get description() {
        return `From serve.json remove the api property`;
    }
    get resolution() {
        return ``;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/serve.json';
    }
    visit(project, findings) {
        if (!project.serveJson) {
            return;
        }
        if (project.serveJson.api) {
            const node = this.getAstNodeFromFile(project.serveJson, 'api');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN007003_CFG_S_api = FN007003_CFG_S_api;
