// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN008001_CFG_TSL_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN008001_CFG_TSL_schema = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN008001_CFG_TSL_schema extends JsonRule_1.JsonRule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN008001';
    }
    get title() {
        return 'tslint.json schema';
    }
    get description() {
        return `Update tslint.json schema URL`;
    }
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/tslint.json';
    }
    visit(project, findings) {
        if (!project.tsLintJson) {
            return;
        }
        if (project.tsLintJson.$schema !== this.schema) {
            const node = this.getAstNodeFromFile(project.tsLintJson, '$schema');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN008001_CFG_TSL_schema = FN008001_CFG_TSL_schema;
