// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN007001_CFG_S_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN007001_CFG_S_schema = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN007001_CFG_S_schema extends JsonRule_1.JsonRule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN007001';
    }
    get title() {
        return 'serve.json schema';
    }
    get description() {
        return `Update serve.json schema URL`;
    }
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/serve.json';
    }
    visit(project, findings) {
        if (!project.serveJson) {
            return;
        }
        if (project.serveJson.$schema !== this.schema) {
            const node = this.getAstNodeFromFile(project.serveJson, '$schema');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN007001_CFG_S_schema = FN007001_CFG_S_schema;
