// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN003002_CFG_version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN003002_CFG_version = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN003002_CFG_version extends JsonRule_1.JsonRule {
    constructor(version) {
        super();
        this.version = version;
    }
    get id() {
        return 'FN003002';
    }
    get title() {
        return `config.json version`;
    }
    get description() {
        return `Update config.json version number`;
    }
    get resolution() {
        return `{
  "version": "${this.version}"
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './config/config.json';
    }
    visit(project, findings) {
        if (!project.configJson) {
            return;
        }
        if (project.configJson.version !== this.version) {
            const node = this.getAstNodeFromFile(project.configJson, 'version');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN003002_CFG_version = FN003002_CFG_version;
