// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010006_YORC_framework.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010006_YORC_framework = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010006_YORC_framework extends JsonRule_1.JsonRule {
    constructor(framework, add) {
        super();
        this.framework = framework;
        this.add = add;
    }
    get id() {
        return 'FN010006';
    }
    get title() {
        return '.yo-rc.json framework';
    }
    get description() {
        return `${this.add ? 'Update' : 'Remove'} framework in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "framework": "${this.framework}"
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a, _b;
        if (!project.yoRcJson) {
            return;
        }
        if (this.add) {
            if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.framework) !== this.framework) {
                const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.framework');
                this.addFindingWithPosition(findings, node);
            }
        }
        else {
            if ((_b = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _b === void 0 ? void 0 : _b.framework) {
                const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.framework');
                this.addFindingWithPosition(findings, node);
            }
        }
    }
}
exports.FN010006_YORC_framework = FN010006_YORC_framework;
