// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010009_YORC_sdkVersions_microsoft_graph_client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010009_YORC_sdkVersions_microsoft_graph_client = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010009_YORC_sdkVersions_microsoft_graph_client extends JsonRule_1.JsonRule {
    constructor(version) {
        super();
        this.version = version;
    }
    get id() {
        return 'FN010009';
    }
    get title() {
        return '.yo-rc.json @microsoft/microsoft-graph-client SDK version';
    }
    get description() {
        return `Update @microsoft/microsoft-graph-client SDK version in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "sdkVersions": {
      "@microsoft/microsoft-graph-client": "${this.version}"
    }
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a, _b, _c;
        if (!project.yoRcJson) {
            return;
        }
        if (((_b = (_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.sdkVersions) === null || _b === void 0 ? void 0 : _b['@microsoft/microsoft-graph-client']) !== this.version) {
            let nodePath = '@microsoft/generator-sharepoint';
            if ((_c = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _c === void 0 ? void 0 : _c.sdkVersions) {
                nodePath += '.sdkVersions';
                if (project.yoRcJson["@microsoft/generator-sharepoint"].sdkVersions['@microsoft/microsoft-graph-client']) {
                    nodePath += '.@microsoft/microsoft-graph-client';
                }
            }
            const node = this.getAstNodeFromFile(project.yoRcJson, nodePath);
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN010009_YORC_sdkVersions_microsoft_graph_client = FN010009_YORC_sdkVersions_microsoft_graph_client;
