// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN014003_CODE_launch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN014003_CODE_launch = void 0;
const Rule_1 = require("../../Rule");
class FN014003_CODE_launch extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN014003';
    }
    get title() {
        return '.vscode/launch.json';
    }
    get description() {
        return `In the .vscode folder, add the launch.json file`;
    }
    get resolution() {
        return `{
  /**
    Install Chrome Debugger Extension for Visual Studio Code
    to debug your components with the Chrome browser:
    https://aka.ms/spfx-debugger-extensions
    */
  "version": "0.2.0",
  "configurations": [{
      "name": "Local workbench",
      "type": "chrome",
      "request": "launch",
      "url": "https://localhost:4321/temp/workbench.html",
      "webRoot": "\${workspaceRoot}",
      "sourceMaps": true,
      "sourceMapPathOverrides": {
        "webpack:///../../../src/*": "\${webRoot}/src/*",
        "webpack:///../../../../src/*": "\${webRoot}/src/*",
        "webpack:///../../../../../src/*": "\${webRoot}/src/*"
      },
      "runtimeArgs": [
        "--remote-debugging-port=9222"
      ]
    },
    {
      "name": "Hosted workbench",
      "type": "chrome",
      "request": "launch",
      "url": "https://enter-your-SharePoint-site/_layouts/workbench.aspx",
      "webRoot": "\${workspaceRoot}",
      "sourceMaps": true,
      "sourceMapPathOverrides": {
        "webpack:///../../../src/*": "\${webRoot}/src/*",
        "webpack:///../../../../src/*": "\${webRoot}/src/*",
        "webpack:///../../../../../src/*": "\${webRoot}/src/*"
      },
      "runtimeArgs": [
        "--remote-debugging-port=9222",
        "-incognito"
      ]
    }
  ]
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return '.vscode/launch.json';
    }
    visit(project, findings) {
        if (!project.vsCode || !project.vsCode.launchJson) {
            this.addFinding(findings);
        }
    }
}
exports.FN014003_CODE_launch = FN014003_CODE_launch;
