// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN011010_MAN_webpart_version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN011010_MAN_webpart_version = void 0;
const ManifestRule_1 = require("./ManifestRule");
class FN011010_MAN_webpart_version extends ManifestRule_1.ManifestRule {
    get id() {
        return 'FN011010';
    }
    get title() {
        return 'Web part manifest version';
    }
    get description() {
        return 'Update version in manifest to use automated component versioning';
    }
    get resolution() {
        return `{
  "version": "*",
}`;
    }
    get severity() {
        return 'Optional';
    }
    visit(project, findings) {
        if (!project.manifests ||
            project.manifests.length === 0) {
            return;
        }
        const occurrences = [];
        project.manifests.forEach(manifest => {
            if (manifest.componentType === 'WebPart' &&
                manifest.version !== '*') {
                const node = this.getAstNodeFromFile(manifest, 'version');
                this.addOccurrence(this.resolution, manifest.path, project.path, node, occurrences);
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN011010_MAN_webpart_version = FN011010_MAN_webpart_version;
