// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN012009_TSC_lib_es2015_collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN012009_TSC_lib_es2015_collection = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN012009_TSC_lib_es2015_collection extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN012009';
    }
    get title() {
        return 'tsconfig.json es2015.collection lib';
    }
    get description() {
        return `Add es2015.collection lib in tsconfig.json`;
    }
    get resolution() {
        return `{
  "compilerOptions": {
    "lib": [
      "es2015.collection"
    ]
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tsconfig.json';
    }
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (!project.tsConfigJson.compilerOptions.lib ||
            project.tsConfigJson.compilerOptions.lib.indexOf('es2015.collection') < 0) {
            const node = this.getAstNodeFromFile(project.tsConfigJson, 'compilerOptions.lib');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN012009_TSC_lib_es2015_collection = FN012009_TSC_lib_es2015_collection;
