// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN011002_MAN_applicationCustomizer_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN011002_MAN_applicationCustomizer_schema = void 0;
const ManifestRule_1 = require("./ManifestRule");
class FN011002_MAN_applicationCustomizer_schema extends ManifestRule_1.ManifestRule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN011002';
    }
    get title() {
        return 'Application customizer manifest schema';
    }
    get description() {
        return 'Update schema in manifest';
    }
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    get severity() {
        return 'Required';
    }
    visit(project, findings) {
        if (!project.manifests ||
            project.manifests.length === 0) {
            return;
        }
        const occurrences = [];
        project.manifests.forEach(manifest => {
            if (manifest.componentType === 'Extension' &&
                manifest.extensionType === 'ApplicationCustomizer' &&
                manifest.$schema !== this.schema) {
                const node = this.getAstNodeFromFile(manifest, '$schema');
                this.addOccurrence(this.resolution, manifest.path, project.path, node, occurrences);
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN011002_MAN_applicationCustomizer_schema = FN011002_MAN_applicationCustomizer_schema;
