// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN012014_TSC_inlineSources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN012014_TSC_inlineSources = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN012014_TSC_inlineSources extends JsonRule_1.JsonRule {
    constructor(inlineSources) {
        super();
        this.inlineSources = inlineSources;
    }
    get id() {
        return 'FN012014';
    }
    get title() {
        return 'tsconfig.json compiler options inlineSources';
    }
    get description() {
        return `Update tsconfig.json inlineSources value`;
    }
    get resolution() {
        return `{
  "compilerOptions": {
    "inlineSources": ${this.inlineSources}
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tsconfig.json';
    }
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.inlineSources !== this.inlineSources) {
            const node = this.getAstNodeFromFile(project.tsConfigJson, 'compilerOptions.inlineSources');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN012014_TSC_inlineSources = FN012014_TSC_inlineSources;
