// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010007_YORC_isDomainIsolated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010007_YORC_isDomainIsolated = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010007_YORC_isDomainIsolated extends JsonRule_1.JsonRule {
    constructor(value) {
        super();
        this.value = value;
    }
    get id() {
        return 'FN010007';
    }
    get title() {
        return '.yo-rc.json isDomainIsolated';
    }
    get description() {
        return `Update isDomainIsolated in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "isDomainIsolated": ${this.value.toString()}
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.yoRcJson) {
            return;
        }
        if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.isDomainIsolated) !== this.value) {
            const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.isDomainIsolated');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN010007_YORC_isDomainIsolated = FN010007_YORC_isDomainIsolated;
