// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN018001_TEAMS_folder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN018001_TEAMS_folder = void 0;
const fs = require("fs");
const path = require("path");
const Rule_1 = require("../../Rule");
class FN018001_TEAMS_folder extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN018001';
    }
    get title() {
        return 'Web part Microsoft Teams tab resources folder';
    }
    get description() {
        return 'Create folder for Microsoft Teams tab resources';
    }
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'cmd';
    }
    get file() {
        return '';
    }
    get severity() {
        return 'Optional';
    }
    visit(project, findings) {
        if (!project.manifests ||
            project.manifests.length < 1) {
            return;
        }
        const webPartManifests = project.manifests.filter(m => m.componentType === 'WebPart');
        if (webPartManifests.length < 1) {
            return;
        }
        const teamsFolderName = 'teams';
        const teamsFolderPath = path.join(project.path, teamsFolderName);
        if (!fs.existsSync(teamsFolderPath)) {
            this.addFindingWithCustomInfo(this.title, this.description, [{
                    file: path.relative(project.path, teamsFolderPath),
                    resolution: `create_dir_cmdPathParam${project.path}NameParam${teamsFolderName}ItemTypeParam`
                }], findings);
        }
    }
}
exports.FN018001_TEAMS_folder = FN018001_TEAMS_folder;
