// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010004_YORC_componentType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010004_YORC_componentType = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010004_YORC_componentType extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN010004';
    }
    get title() {
        return '.yo-rc.json componentType';
    }
    get description() {
        return `Update componentType in .yo-rc.json`;
    }
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.yoRcJson) {
            return;
        }
        let componentType;
        if (project.manifests) {
            for (let i = 0; i < project.manifests.length; i++) {
                const manifest = project.manifests[i];
                if (manifest.componentType === 'WebPart') {
                    componentType = 'webpart';
                    break;
                }
                if (manifest.componentType === 'Extension') {
                    componentType = 'extension';
                    break;
                }
            }
        }
        if (!componentType) {
            componentType = 'webpart';
        }
        if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.componentType) !== componentType) {
            const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.componentType');
            this.addFindingWithOccurrences([{
                    file: this.file,
                    resolution: JSON.stringify({
                        "@microsoft/generator-sharepoint": {
                            "componentType": componentType
                        }
                    }, null, 2),
                    position: this.getPositionFromNode(node)
                }], findings);
        }
    }
}
exports.FN010004_YORC_componentType = FN010004_YORC_componentType;
