// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010001_YORC_version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010001_YORC_version = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010001_YORC_version extends JsonRule_1.JsonRule {
    constructor(version) {
        super();
        this.version = version;
    }
    get id() {
        return 'FN010001';
    }
    get title() {
        return '.yo-rc.json version';
    }
    get description() {
        return `Update version in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "version": "${this.version}"
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.yoRcJson) {
            return;
        }
        if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.version) !== this.version) {
            const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.version');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN010001_YORC_version = FN010001_YORC_version;
