// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN019001_TSL_rulesDirectory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN019001_TSL_rulesDirectory = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN019001_TSL_rulesDirectory extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN019001';
    }
    get title() {
        return 'tslint.json rulesDirectory';
    }
    get description() {
        return `Remove rulesDirectory from tslint.json`;
    }
    get resolution() {
        return `{
  "rulesDirectory": []
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tslint.json';
    }
    visit(project, findings) {
        if (!project.tsLintJsonRoot) {
            return;
        }
        if (project.tsLintJsonRoot.rulesDirectory) {
            const node = this.getAstNodeFromFile(project.tsLintJsonRoot, 'rulesDirectory');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN019001_TSL_rulesDirectory = FN019001_TSL_rulesDirectory;
