// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN013001_GULP_msGridSassSuppression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN013001_GULP_msGridSassSuppression = void 0;
const Rule_1 = require("../../Rule");
class FN013001_GULP_msGridSassSuppression extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN013001';
    }
    get title() {
        return 'gulpfile.js ms-Grid sass suppression';
    }
    get description() {
        return `Add suppression for ms-Grid sass warning`;
    }
    get resolution() {
        return "build.addSuppression(`Warning - [sass] The local CSS class 'ms-Grid' is not camelCase and will not be type-safe.`);";
    }
    get resolutionType() {
        return 'js';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './gulpfile.js';
    }
    visit(project, findings) {
        if (!project.gulpfileJs) {
            return;
        }
        if (project.gulpfileJs.source.indexOf(this.resolution) < 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN013001_GULP_msGridSassSuppression = FN013001_GULP_msGridSassSuppression;
