// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN012010_TSC_experimentalDecorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN012010_TSC_experimentalDecorators = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN012010_TSC_experimentalDecorators extends JsonRule_1.JsonRule {
    constructor() {
        super();
    }
    get id() {
        return 'FN012010';
    }
    get title() {
        return 'tsconfig.json compiler options experimental decorators';
    }
    get description() {
        return `Enable tsconfig.json experimental decorators`;
    }
    get resolution() {
        return `{
  "compilerOptions": {
    "experimentalDecorators": true
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tsconfig.json';
    }
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (!project.tsConfigJson.compilerOptions.experimentalDecorators) {
            const node = this.getAstNodeFromFile(project.tsConfigJson, 'compilerOptions.experimentalDecorators');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN012010_TSC_experimentalDecorators = FN012010_TSC_experimentalDecorators;
