// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN014008_CODE_launch_hostedWorkbench_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN014008_CODE_launch_hostedWorkbench_type = void 0;
const path = require("path");
const JsonRule_1 = require("../../JsonRule");
class FN014008_CODE_launch_hostedWorkbench_type extends JsonRule_1.JsonRule {
    constructor(type) {
        super();
        this.type = type;
    }
    get id() {
        return 'FN014008';
    }
    get title() {
        return 'Hosted workbench type in .vscode/launch.json';
    }
    get description() {
        return `In the .vscode/launch.json file, update the type property for the hosted workbench launch configuration`;
    }
    get resolution() {
        return `{
  "configurations": [
    {
      "type": "${this.type}"
    }
  ]
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return '.vscode/launch.json';
    }
    visit(project, findings) {
        if (!project.vsCode ||
            !project.vsCode.launchJson ||
            !project.vsCode.launchJson.configurations) {
            return;
        }
        const occurrences = [];
        project.vsCode.launchJson.configurations.forEach((configuration, i) => {
            if (configuration.name === 'Hosted workbench' &&
                configuration.type !== this.type) {
                const node = this.getAstNodeFromFile(project.vsCode.launchJson, `configurations[${i}].url`);
                occurrences.push({
                    file: path.relative(project.path, this.file),
                    resolution: this.resolution,
                    position: this.getPositionFromNode(node)
                });
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN014008_CODE_launch_hostedWorkbench_type = FN014008_CODE_launch_hostedWorkbench_type;
