// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN012001_TSC_module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN012001_TSC_module = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN012001_TSC_module extends JsonRule_1.JsonRule {
    constructor(module) {
        super();
        this.module = module;
    }
    get id() {
        return 'FN012001';
    }
    get title() {
        return 'tsconfig.json module';
    }
    get description() {
        return `Update module type in tsconfig.json`;
    }
    get resolution() {
        return `{
  "compilerOptions": {
    "module": "${this.module}"
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tsconfig.json';
    }
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.module !== this.module) {
            const node = this.getAstNodeFromFile(project.tsConfigJson, 'compilerOptions.module');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN012001_TSC_module = FN012001_TSC_module;
