// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010003_YORC_packageManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010003_YORC_packageManager = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010003_YORC_packageManager extends JsonRule_1.JsonRule {
    constructor(packageManager) {
        super();
        this.packageManager = packageManager;
    }
    get id() {
        return 'FN010003';
    }
    get title() {
        return '.yo-rc.json packageManager';
    }
    get description() {
        return `Update packageManager in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "packageManager": "${this.packageManager}"
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.yoRcJson) {
            return;
        }
        if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.packageManager) !== this.packageManager) {
            const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.packageManager');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN010003_YORC_packageManager = FN010003_YORC_packageManager;
