// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN012016_TSC_noUnusedLocals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN012016_TSC_noUnusedLocals = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN012016_TSC_noUnusedLocals extends JsonRule_1.JsonRule {
    constructor(noUnusedLocals) {
        super();
        this.noUnusedLocals = noUnusedLocals;
    }
    get id() {
        return 'FN012016';
    }
    get title() {
        return 'tsconfig.json compiler options noUnusedLocals';
    }
    get description() {
        return `Update tsconfig.json noUnusedLocals value`;
    }
    get resolution() {
        return `{
  "compilerOptions": {
    "noUnusedLocals": ${this.noUnusedLocals}
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './tsconfig.json';
    }
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.noUnusedLocals !== this.noUnusedLocals) {
            const node = this.getAstNodeFromFile(project.tsConfigJson, 'compilerOptions.noUnusedLocals');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN012016_TSC_noUnusedLocals = FN012016_TSC_noUnusedLocals;
