// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/FN010005_YORC_environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN010005_YORC_environment = void 0;
const JsonRule_1 = require("../../JsonRule");
class FN010005_YORC_environment extends JsonRule_1.JsonRule {
    constructor(environment) {
        super();
        this.environment = environment;
    }
    get id() {
        return 'FN010005';
    }
    get title() {
        return '.yo-rc.json environment';
    }
    get description() {
        return `Update environment in .yo-rc.json`;
    }
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "environment": "${this.environment}"
  }
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Recommended';
    }
    get file() {
        return './.yo-rc.json';
    }
    visit(project, findings) {
        var _a;
        if (!project.yoRcJson) {
            return;
        }
        if (((_a = project.yoRcJson["@microsoft/generator-sharepoint"]) === null || _a === void 0 ? void 0 : _a.environment) !== this.environment) {
            const node = this.getAstNodeFromFile(project.yoRcJson, '@microsoft/generator-sharepoint.environment');
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.FN010005_YORC_environment = FN010005_YORC_environment;
