// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/PackageRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageRule = void 0;
const JsonRule_1 = require("../../JsonRule");
class PackageRule extends JsonRule_1.JsonRule {
    constructor(propertyName, add, propertyValue) {
        super();
        this.propertyName = propertyName;
        this.add = add;
        this.propertyValue = propertyValue;
    }
    get title() {
        return this.propertyName;
    }
    get description() {
        return `${this.add ? 'Add' : 'Remove'} package.json property`;
    }
    get resolution() {
        return `{
  "${this.propertyName}": "${this.propertyValue}"
}`;
    }
    get resolutionType() {
        return 'json';
    }
    get severity() {
        return 'Required';
    }
    get file() {
        return './package.json';
    }
    visit(project, findings) {
        if (!project.packageJson) {
            return;
        }
        if (!project.packageJson ||
            (this.add && !project.packageJson[this.propertyName]) ||
            (!this.add && project.packageJson[this.propertyName])) {
            const node = this.getAstNodeFromFile(project.packageJson, this.propertyName);
            this.addFindingWithPosition(findings, node);
        }
    }
}
exports.PackageRule = PackageRule;
