// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-upgrade/rules/TsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsRule = void 0;
const path = require("path");
const Rule_1 = require("../../Rule");
class TsRule extends Rule_1.Rule {
    get resolution() {
        return '';
    }
    get resolutionType() {
        return 'ts';
    }
    get file() {
        return '';
    }
    addOccurrence(resolution, filePath, projectPath, node, occurrences) {
        const lineChar = node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
        occurrences.push({
            file: path.relative(projectPath, filePath),
            position: {
                line: lineChar.line + 1,
                character: lineChar.character + 1
            },
            resolution: resolution
        });
    }
    static getParentOfType(node, typeComparer) {
        const parent = node.parent;
        if (!parent) {
            return undefined;
        }
        if (typeComparer(parent)) {
            return parent;
        }
        return TsRule.getParentOfType(parent, typeComparer);
    }
}
exports.TsRule = TsRule;
