// @pnp/cli-microsoft365/dist/m365/spo/commands/roledefinition/roledefinition-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoRoleDefinitionRemoveCommand_instances, _SpoRoleDefinitionRemoveCommand_initTelemetry, _SpoRoleDefinitionRemoveCommand_initOptions, _SpoRoleDefinitionRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoRoleDefinitionRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoRoleDefinitionRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoRoleDefinitionRemoveCommand_instances, "m", _SpoRoleDefinitionRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoRoleDefinitionRemoveCommand_instances, "m", _SpoRoleDefinitionRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoRoleDefinitionRemoveCommand_instances, "m", _SpoRoleDefinitionRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.ROLEDEFINITION_REMOVE;
    }
    get description() {
        return 'Removes the role definition from the specified site';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeRoleDefinition = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing role definition from site ${args.options.webUrl}...`);
                }
                const requestOptions = {
                    url: `${args.options.webUrl}/_api/web/roledefinitions(${args.options.id})`,
                    method: 'delete',
                    headers: {
                        'X-HTTP-Method': 'DELETE',
                        'If-Match': '*',
                        'accept': 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeRoleDefinition();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the role definition with id ${args.options.id} from site ${args.options.webUrl}?`
                });
                if (result.continue) {
                    yield removeRoleDefinition();
                }
            }
        });
    }
}
_SpoRoleDefinitionRemoveCommand_instances = new WeakSet(), _SpoRoleDefinitionRemoveCommand_initTelemetry = function _SpoRoleDefinitionRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoRoleDefinitionRemoveCommand_initOptions = function _SpoRoleDefinitionRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-i, --id <id>'
    }, {
        option: '--confirm'
    });
}, _SpoRoleDefinitionRemoveCommand_initValidators = function _SpoRoleDefinitionRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const id = parseInt(args.options.id);
        if (isNaN(id)) {
            return `${args.options.id} is not a valid role definition ID`;
        }
        return validation_1.validation.isValidSharePointUrl(args.options.webUrl);
    }));
};
module.exports = new SpoRoleDefinitionRemoveCommand();
