// @pnp/cli-microsoft365/dist/m365/teams/commands/app/app-uninstall.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsAppUninstallCommand_instances, _TeamsAppUninstallCommand_initTelemetry, _TeamsAppUninstallCommand_initOptions, _TeamsAppUninstallCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsAppUninstallCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsAppUninstallCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsAppUninstallCommand_instances, "m", _TeamsAppUninstallCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsAppUninstallCommand_instances, "m", _TeamsAppUninstallCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsAppUninstallCommand_instances, "m", _TeamsAppUninstallCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.APP_UNINSTALL;
    }
    get description() {
        return 'Uninstalls an app from a Microsoft Team team';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const uninstallApp = () => __awaiter(this, void 0, void 0, function* () {
                const requestOptions = {
                    url: `${this.resource}/v1.0/teams/${args.options.teamId}/installedApps/${args.options.appId}`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    }
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield uninstallApp();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to uninstall the app with id ${args.options.appId} from the Microsoft Teams team ${args.options.teamId}?`
                });
                if (result.continue) {
                    yield uninstallApp();
                }
            }
        });
    }
}
_TeamsAppUninstallCommand_instances = new WeakSet(), _TeamsAppUninstallCommand_initTelemetry = function _TeamsAppUninstallCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: args.options.confirm || false
        });
    });
}, _TeamsAppUninstallCommand_initOptions = function _TeamsAppUninstallCommand_initOptions() {
    this.options.unshift({
        option: '--appId <appId>'
    }, {
        option: '--teamId <teamId>'
    }, {
        option: '--confirm'
    });
}, _TeamsAppUninstallCommand_initValidators = function _TeamsAppUninstallCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new TeamsAppUninstallCommand();
