// @pnp/cli-microsoft365/dist/m365/spo/commands/customaction/customaction-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoCustomActionGetCommand_instances, _SpoCustomActionGetCommand_initTelemetry, _SpoCustomActionGetCommand_initOptions, _SpoCustomActionGetCommand_initValidators, _SpoCustomActionGetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoCustomActionGetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoCustomActionGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoCustomActionGetCommand_instances, "m", _SpoCustomActionGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoCustomActionGetCommand_instances, "m", _SpoCustomActionGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoCustomActionGetCommand_instances, "m", _SpoCustomActionGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoCustomActionGetCommand_instances, "m", _SpoCustomActionGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.CUSTOMACTION_GET;
    }
    get description() {
        return 'Gets details for the specified custom action';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                let customAction;
                if (args.options.scope && args.options.scope.toLowerCase() !== "all") {
                    customAction = yield this.getCustomAction(args.options);
                }
                else {
                    customAction = yield this.searchAllScopes(args.options);
                }
                if (customAction["odata.null"] === true) {
                    if (this.verbose) {
                        logger.logToStderr(`Custom action with id ${args.options.id} not found`);
                    }
                }
                else {
                    logger.log({
                        ClientSideComponentId: customAction.ClientSideComponentId,
                        ClientSideComponentProperties: customAction.ClientSideComponentProperties,
                        CommandUIExtension: customAction.CommandUIExtension,
                        Description: customAction.Description,
                        Group: customAction.Group,
                        Id: customAction.Id,
                        ImageUrl: customAction.ImageUrl,
                        Location: customAction.Location,
                        Name: customAction.Name,
                        RegistrationId: customAction.RegistrationId,
                        RegistrationType: customAction.RegistrationType,
                        Rights: JSON.stringify(customAction.Rights),
                        Scope: this.humanizeScope(customAction.Scope),
                        ScriptBlock: customAction.ScriptBlock,
                        ScriptSrc: customAction.ScriptSrc,
                        Sequence: customAction.Sequence,
                        Title: customAction.Title,
                        Url: customAction.Url,
                        VersionOfUserCustomAction: customAction.VersionOfUserCustomAction
                    });
                }
            }
            catch (err) {
                this.handleRejectedPromise(err);
            }
        });
    }
    getCustomAction(options) {
        const filter = options.id ?
            `('${encodeURIComponent(options.id)}')` :
            `?$filter=Title eq '${encodeURIComponent(options.title)}'`;
        const requestOptions = {
            url: `${options.url}/_api/${options.scope}/UserCustomActions${filter}`,
            headers: {
                accept: 'application/json;odata=nometadata'
            },
            responseType: 'json'
        };
        if (options.id) {
            return request_1.default
                .get(requestOptions)
                .then((res) => {
                return Promise.resolve(res);
            });
        }
        return request_1.default
            .get(requestOptions)
            .then((res) => {
            if (res.value.length === 1) {
                return Promise.resolve(res.value[0]);
            }
            if (res.value.length === 0) {
                return Promise.reject(`No user custom action with title '${options.title}' found`);
            }
            return Promise.reject(`Multiple user custom actions with title '${options.title}' found. Please disambiguate using IDs: ${res.value.map(a => a.Id).join(', ')}`);
        });
    }
    /**
     * Get request with `web` scope is send first.
     * If custom action not found then
     * another get request is send with `site` scope.
     */
    searchAllScopes(options) {
        return new Promise((resolve, reject) => {
            options.scope = "Web";
            this
                .getCustomAction(options)
                .then((webResult) => {
                if (webResult["odata.null"] !== true) {
                    return resolve(webResult);
                }
                options.scope = "Site";
                this
                    .getCustomAction(options)
                    .then((siteResult) => {
                    return resolve(siteResult);
                }, (err) => {
                    reject(err);
                });
            }, (err) => {
                reject(err);
            });
        });
    }
    humanizeScope(scope) {
        switch (scope) {
            case 2:
                return "Site";
            case 3:
                return "Web";
        }
        return `${scope}`;
    }
}
_SpoCustomActionGetCommand_instances = new WeakSet(), _SpoCustomActionGetCommand_initTelemetry = function _SpoCustomActionGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            scope: args.options.scope || 'All'
        });
    });
}, _SpoCustomActionGetCommand_initOptions = function _SpoCustomActionGetCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id [id]'
    }, {
        option: '-t, --title [title]'
    }, {
        option: '-u, --url <url>'
    }, {
        option: '-s, --scope [scope]',
        autocomplete: ['Site', 'Web', 'All']
    });
}, _SpoCustomActionGetCommand_initValidators = function _SpoCustomActionGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.id && validation_1.validation.isValidGuid(args.options.id) === false) {
            return `${args.options.id} is not valid. Custom action id (Guid) expected.`;
        }
        if (validation_1.validation.isValidSharePointUrl(args.options.url) !== true) {
            return 'Missing required option url';
        }
        if (args.options.scope) {
            if (args.options.scope !== 'Site' &&
                args.options.scope !== 'Web' &&
                args.options.scope !== 'All') {
                return `${args.options.scope} is not a valid custom action scope. Allowed values are Site|Web|All`;
            }
        }
        return true;
    }));
}, _SpoCustomActionGetCommand_initOptionSets = function _SpoCustomActionGetCommand_initOptionSets() {
    this.optionSets.push(['id', 'title']);
};
module.exports = new SpoCustomActionGetCommand();
