// @pnp/cli-microsoft365/dist/m365/spo/commands/field/field-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoFieldGetCommand_instances, _SpoFieldGetCommand_initTelemetry, _SpoFieldGetCommand_initOptions, _SpoFieldGetCommand_initValidators, _SpoFieldGetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoFieldGetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoFieldGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoFieldGetCommand_instances, "m", _SpoFieldGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoFieldGetCommand_instances, "m", _SpoFieldGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoFieldGetCommand_instances, "m", _SpoFieldGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoFieldGetCommand_instances, "m", _SpoFieldGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.FIELD_GET;
    }
    get description() {
        return 'Retrieves information about the specified list- or site column';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.fieldTitle) {
                args.options.title = args.options.fieldTitle;
                this.warn(logger, `Option 'fieldTitle' is deprecated. Please use 'title' instead.`);
            }
            let listRestUrl = '';
            if (args.options.listId) {
                listRestUrl = `lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')/`;
            }
            else if (args.options.listTitle) {
                listRestUrl = `lists/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')/`;
            }
            else if (args.options.listUrl) {
                const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                listRestUrl = `GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')/`;
            }
            let fieldRestUrl = '';
            if (args.options.id) {
                fieldRestUrl = `/getbyid('${formatting_1.formatting.encodeQueryParameter(args.options.id)}')`;
            }
            else {
                fieldRestUrl = `/getbyinternalnameortitle('${formatting_1.formatting.encodeQueryParameter(args.options.title)}')`;
            }
            const requestOptions = {
                url: `${args.options.webUrl}/_api/web/${listRestUrl}fields${fieldRestUrl}`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoFieldGetCommand_instances = new WeakSet(), _SpoFieldGetCommand_initTelemetry = function _SpoFieldGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            id: typeof args.options.id !== 'undefined',
            title: typeof args.options.title !== 'undefined'
        });
    });
}, _SpoFieldGetCommand_initOptions = function _SpoFieldGetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-l, --listTitle [listTitle]'
    }, {
        option: '--listId [listId]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '-i, --id [id]'
    }, {
        option: '--fieldTitle [fieldTitle]'
    }, {
        option: '-t, --title [title]'
    });
}, _SpoFieldGetCommand_initValidators = function _SpoFieldGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.id && !validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        if (args.options.listId && !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} is not a valid GUID`;
        }
        return true;
    }));
}, _SpoFieldGetCommand_initOptionSets = function _SpoFieldGetCommand_initOptionSets() {
    this.optionSets.push(['id', 'title', 'fieldTitle']);
};
module.exports = new SpoFieldGetCommand();
