// @pnp/cli-microsoft365/dist/m365/planner/commands/bucket/bucket-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PlannerBucketListCommand_instances, _PlannerBucketListCommand_initTelemetry, _PlannerBucketListCommand_initOptions, _PlannerBucketListCommand_initValidators, _PlannerBucketListCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../../../Auth");
const accessToken_1 = require("../../../../utils/accessToken");
const odata_1 = require("../../../../utils/odata");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const planner_1 = require("../../../../utils/planner");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class PlannerBucketListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _PlannerBucketListCommand_instances.add(this);
        __classPrivateFieldGet(this, _PlannerBucketListCommand_instances, "m", _PlannerBucketListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PlannerBucketListCommand_instances, "m", _PlannerBucketListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PlannerBucketListCommand_instances, "m", _PlannerBucketListCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _PlannerBucketListCommand_instances, "m", _PlannerBucketListCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.BUCKET_LIST;
    }
    get description() {
        return 'Lists the Microsoft Planner buckets in a plan';
    }
    defaultProperties() {
        return ['id', 'name', 'planId', 'orderHint'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.planName) {
                args.options.planTitle = args.options.planName;
                this.warn(logger, `Option 'planName' is deprecated. Please use 'planTitle' instead`);
            }
            if (accessToken_1.accessToken.isAppOnlyAccessToken(Auth_1.default.service.accessTokens[this.resource].accessToken)) {
                this.handleError('This command does not support application permissions.');
                return;
            }
            try {
                const planId = yield this.getPlanId(args);
                const buckets = yield odata_1.odata.getAllItems(`${this.resource}/v1.0/planner/plans/${planId}/buckets`);
                logger.log(buckets);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getPlanId(args) {
        if (args.options.planId) {
            return Promise.resolve(args.options.planId);
        }
        return this
            .getGroupId(args)
            .then(groupId => planner_1.planner.getPlanByTitle(args.options.planTitle, groupId))
            .then(plan => plan.id);
    }
    getGroupId(args) {
        if (args.options.ownerGroupId) {
            return Promise.resolve(args.options.ownerGroupId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.ownerGroupName)
            .then(group => group.id);
    }
}
_PlannerBucketListCommand_instances = new WeakSet(), _PlannerBucketListCommand_initTelemetry = function _PlannerBucketListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            planId: typeof args.options.planId !== 'undefined',
            planName: typeof args.options.planName !== 'undefined',
            planTitle: typeof args.options.planTitle !== 'undefined',
            ownerGroupId: typeof args.options.ownerGroupId !== 'undefined',
            ownerGroupName: typeof args.options.ownerGroupName !== 'undefined'
        });
    });
}, _PlannerBucketListCommand_initOptions = function _PlannerBucketListCommand_initOptions() {
    this.options.unshift({
        option: '--planId [planId]'
    }, {
        option: '--planName [planName]'
    }, {
        option: "--planTitle [planTitle]"
    }, {
        option: '--ownerGroupId [ownerGroupId]'
    }, {
        option: '--ownerGroupName [ownerGroupName]'
    });
}, _PlannerBucketListCommand_initValidators = function _PlannerBucketListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if ((args.options.planName || args.options.planTitle) && !args.options.ownerGroupId && !args.options.ownerGroupName) {
            return 'Specify either ownerGroupId or ownerGroupName when using planTitle';
        }
        if ((args.options.planName || args.options.planTitle) && args.options.ownerGroupId && args.options.ownerGroupName) {
            return 'Specify either ownerGroupId or ownerGroupName when using planTitle but not both';
        }
        if (args.options.ownerGroupId && !validation_1.validation.isValidGuid(args.options.ownerGroupId)) {
            return `${args.options.ownerGroupId} is not a valid GUID`;
        }
        return true;
    }));
}, _PlannerBucketListCommand_initOptionSets = function _PlannerBucketListCommand_initOptionSets() {
    this.optionSets.push(['planId', 'planTitle']);
};
module.exports = new PlannerBucketListCommand();
