// @pnp/cli-microsoft365/dist/m365/teams/commands/report/report-pstncalls.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsReportPstncallsCommand_instances, _TeamsReportPstncallsCommand_initTelemetry, _TeamsReportPstncallsCommand_initOptions, _TeamsReportPstncallsCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsReportPstncallsCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsReportPstncallsCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsReportPstncallsCommand_instances, "m", _TeamsReportPstncallsCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsReportPstncallsCommand_instances, "m", _TeamsReportPstncallsCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsReportPstncallsCommand_instances, "m", _TeamsReportPstncallsCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.REPORT_PSTNCALLS;
    }
    get description() {
        return 'Get details about PSTN calls made within a given time period';
    }
    defaultProperties() {
        return ['id', 'calleeNumber', 'callerNumber', 'startDateTime'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const toDateTimeParameter = encodeURIComponent(args.options.toDateTime ? args.options.toDateTime : new Date().toISOString());
            const requestOptions = {
                url: `${this.resource}/v1.0/communications/callRecords/getPstnCalls(fromDateTime=${encodeURIComponent(args.options.fromDateTime)},toDateTime=${toDateTimeParameter})`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsReportPstncallsCommand_instances = new WeakSet(), _TeamsReportPstncallsCommand_initTelemetry = function _TeamsReportPstncallsCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            toDateTime: typeof args.options.toDateTime !== 'undefined'
        });
    });
}, _TeamsReportPstncallsCommand_initOptions = function _TeamsReportPstncallsCommand_initOptions() {
    this.options.unshift({
        option: '--fromDateTime <fromDateTime>'
    }, {
        option: '--toDateTime [toDateTime]'
    });
}, _TeamsReportPstncallsCommand_initValidators = function _TeamsReportPstncallsCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidISODateDashOnly(args.options.fromDateTime)) {
            return 'The fromDateTime is not a valid ISO date string (YYYY-MM-DD)';
        }
        if (args.options.toDateTime &&
            !validation_1.validation.isValidISODateDashOnly(args.options.toDateTime)) {
            return 'The toDateTime is not a valid ISO date string (YYYY-MM-DD)';
        }
        if (Math.ceil((new Date(args.options.toDateTime || new Date().toISOString()).getTime() - new Date(args.options.fromDateTime).getTime()) / (1000 * 3600 * 24)) > 90) {
            return 'The maximum number of days between fromDateTime and toDateTime cannot exceed 90';
        }
        return true;
    }));
};
module.exports = new TeamsReportPstncallsCommand();
