// @pnp/cli-microsoft365/dist/m365/spo/commands/navigation/navigation-node-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoNavigationNodeRemoveCommand_instances, _SpoNavigationNodeRemoveCommand_initTelemetry, _SpoNavigationNodeRemoveCommand_initOptions, _SpoNavigationNodeRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoNavigationNodeRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoNavigationNodeRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoNavigationNodeRemoveCommand_instances, "m", _SpoNavigationNodeRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoNavigationNodeRemoveCommand_instances, "m", _SpoNavigationNodeRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoNavigationNodeRemoveCommand_instances, "m", _SpoNavigationNodeRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.NAVIGATION_NODE_REMOVE;
    }
    get description() {
        return 'Removes the specified navigation node';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeNode = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const res = yield spo_1.spo.getRequestDigest(args.options.webUrl);
                    if (this.verbose) {
                        logger.logToStderr(`Removing navigation node...`);
                    }
                    const requestOptions = {
                        url: `${args.options.webUrl}/_api/web/navigation/${args.options.location.toLowerCase()}/getbyid(${args.options.id})`,
                        headers: {
                            accept: 'application/json;odata=nometadata',
                            'X-RequestDigest': res.FormDigestValue
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeNode();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the node ${args.options.id} from the navigation?`
                });
                if (result.continue) {
                    yield removeNode();
                }
            }
        });
    }
}
_SpoNavigationNodeRemoveCommand_instances = new WeakSet(), _SpoNavigationNodeRemoveCommand_initTelemetry = function _SpoNavigationNodeRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            location: args.options.location,
            confirm: typeof args.options.confirm !== 'undefined'
        });
    });
}, _SpoNavigationNodeRemoveCommand_initOptions = function _SpoNavigationNodeRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-l, --location <location>',
        autocomplete: ['QuickLaunch', 'TopNavigationBar']
    }, {
        option: '-i, --id <id>'
    }, {
        option: '--confirm'
    });
}, _SpoNavigationNodeRemoveCommand_initValidators = function _SpoNavigationNodeRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.location !== 'QuickLaunch' &&
            args.options.location !== 'TopNavigationBar') {
            return `${args.options.location} is not a valid value for the location option. Allowed values are QuickLaunch|TopNavigationBar`;
        }
        const id = parseInt(args.options.id);
        if (isNaN(id)) {
            return `${args.options.id} is not a number`;
        }
        return true;
    }));
};
module.exports = new SpoNavigationNodeRemoveCommand();
