// @pnp/cli-microsoft365/dist/m365/todo/commands/list/list-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TodoListAddCommand_instances, _TodoListAddCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TodoListAddCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TodoListAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _TodoListAddCommand_instances, "m", _TodoListAddCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.LIST_ADD;
    }
    get description() {
        return 'Adds a new Microsoft To Do task list';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {
                displayName: args.options.name
            };
            const requestOptions = {
                url: `${this.resource}/v1.0/me/todo/lists`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json'
                },
                data,
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.post(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TodoListAddCommand_instances = new WeakSet(), _TodoListAddCommand_initOptions = function _TodoListAddCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    });
};
module.exports = new TodoListAddCommand();
