// @pnp/cli-microsoft365/dist/m365/teams/commands/funsettings/funsettings-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsFunSettingsSetCommand_instances, _TeamsFunSettingsSetCommand_initTelemetry, _TeamsFunSettingsSetCommand_initOptions, _TeamsFunSettingsSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsFunSettingsSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsFunSettingsSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsFunSettingsSetCommand_instances, "m", _TeamsFunSettingsSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsFunSettingsSetCommand_instances, "m", _TeamsFunSettingsSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsFunSettingsSetCommand_instances, "m", _TeamsFunSettingsSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.FUNSETTINGS_SET;
    }
    get description() {
        return 'Updates fun settings of a Microsoft Teams team';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const data = {
                    funSettings: {}
                };
                TeamsFunSettingsSetCommand.booleanProps.forEach(p => {
                    if (typeof args.options[p] !== 'undefined') {
                        data.funSettings[p] = args.options[p] === 'true';
                    }
                });
                if (args.options.giphyContentRating) {
                    data.funSettings.giphyContentRating = args.options.giphyContentRating;
                }
                const requestOptions = {
                    url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    data: data,
                    responseType: 'json'
                };
                yield request_1.default.patch(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsFunSettingsSetCommand_instances = new WeakSet(), _TeamsFunSettingsSetCommand_initTelemetry = function _TeamsFunSettingsSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            giphyContentRating: args.options.giphyContentRating
        });
        TeamsFunSettingsSetCommand.booleanProps.forEach(p => {
            this.telemetryProperties[p] = args.options[p];
        });
    });
}, _TeamsFunSettingsSetCommand_initOptions = function _TeamsFunSettingsSetCommand_initOptions() {
    this.options.unshift({
        option: '-i, --teamId <teamId>'
    }, {
        option: '--allowGiphy [allowGiphy]'
    }, {
        option: '--giphyContentRating [giphyContentRating]'
    }, {
        option: '--allowStickersAndMemes [allowStickersAndMemes]'
    }, {
        option: '--allowCustomMemes [allowCustomMemes]'
    });
}, _TeamsFunSettingsSetCommand_initValidators = function _TeamsFunSettingsSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        let isValid = true;
        let value, property = '';
        TeamsFunSettingsSetCommand.booleanProps.every(p => {
            property = p;
            value = args.options[p];
            isValid = typeof value === 'undefined' ||
                value === 'true' ||
                value === 'false';
            return isValid;
        });
        if (!isValid) {
            return `Value ${value} for option ${property} is not a valid boolean`;
        }
        if (args.options.giphyContentRating) {
            const giphyContentRating = args.options.giphyContentRating.toLowerCase();
            if (giphyContentRating !== 'strict' && giphyContentRating !== 'moderate') {
                return `giphyContentRating value ${value} is not valid.  Please specify Strict or Moderate.`;
            }
        }
        return true;
    }));
};
TeamsFunSettingsSetCommand.booleanProps = [
    'allowGiphy',
    'allowStickersAndMemes',
    'allowCustomMemes'
];
module.exports = new TeamsFunSettingsSetCommand();
