// @pnp/cli-microsoft365/dist/m365/pa/commands/app/app-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PaAppRemoveCommand_instances, _PaAppRemoveCommand_initTelemetry, _PaAppRemoveCommand_initOptions, _PaAppRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const Command_1 = require("../../../../Command");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const PowerAppsCommand_1 = require("../../../base/PowerAppsCommand");
const commands_1 = require("../../commands");
class PaAppRemoveCommand extends PowerAppsCommand_1.default {
    constructor() {
        super();
        _PaAppRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _PaAppRemoveCommand_instances, "m", _PaAppRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PaAppRemoveCommand_instances, "m", _PaAppRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _PaAppRemoveCommand_instances, "m", _PaAppRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.APP_REMOVE;
    }
    get description() {
        return 'Removes the specified Microsoft Power App';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Removing Microsoft Power App ${args.options.name}...`);
            }
            const removePaApp = () => __awaiter(this, void 0, void 0, function* () {
                const requestOptions = {
                    url: `${this.resource}/providers/Microsoft.PowerApps/apps/${encodeURIComponent(args.options.name)}?api-version=2017-08-01`,
                    resolveWithFullResponse: true,
                    headers: {
                        accept: 'application/json'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    if (err.response && err.response.status === 403) {
                        throw new Command_1.CommandError(`App '${args.options.name}' does not exist`);
                    }
                    else {
                        this.handleRejectedODataJsonPromise(err);
                    }
                }
            });
            if (args.options.confirm) {
                yield removePaApp();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the Microsoft Power App ${args.options.name}?`
                });
                if (result.continue) {
                    yield removePaApp();
                }
            }
        });
    }
}
_PaAppRemoveCommand_instances = new WeakSet(), _PaAppRemoveCommand_initTelemetry = function _PaAppRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: typeof args.options.confirm !== 'undefined'
        });
    });
}, _PaAppRemoveCommand_initOptions = function _PaAppRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    }, {
        option: '--confirm'
    });
}, _PaAppRemoveCommand_initValidators = function _PaAppRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.name)) {
            return `${args.options.name} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new PaAppRemoveCommand();
