// @pnp/cli-microsoft365/dist/m365/aad/commands/user/user-password-validate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadUserPasswordValidateCommand_instances, _AadUserPasswordValidateCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadUserPasswordValidateCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadUserPasswordValidateCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadUserPasswordValidateCommand_instances, "m", _AadUserPasswordValidateCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.USER_PASSWORD_VALIDATE;
    }
    get description() {
        return "Check a user's password against the organization's password validation policy";
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const requestOptions = {
                    url: `${this.resource}/beta/users/validatePassword`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    data: {
                        password: args.options.password
                    },
                    responseType: 'json'
                };
                const res = yield request_1.default.post(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadUserPasswordValidateCommand_instances = new WeakSet(), _AadUserPasswordValidateCommand_initOptions = function _AadUserPasswordValidateCommand_initOptions() {
    this.options.unshift({
        option: '-p, --password <password>'
    });
};
module.exports = new AadUserPasswordValidateCommand();
