// @pnp/cli-microsoft365/dist/m365/aad/commands/o365group/o365group-recyclebinitem-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadO365GroupRecycleBinItemListCommand_instances, _AadO365GroupRecycleBinItemListCommand_initTelemetry, _AadO365GroupRecycleBinItemListCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const odata_1 = require("../../../../utils/odata");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadO365GroupRecycleBinItemListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadO365GroupRecycleBinItemListCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadO365GroupRecycleBinItemListCommand_instances, "m", _AadO365GroupRecycleBinItemListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _AadO365GroupRecycleBinItemListCommand_instances, "m", _AadO365GroupRecycleBinItemListCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.O365GROUP_RECYCLEBINITEM_LIST;
    }
    get description() {
        return 'Lists Microsoft 365 Groups deleted in the current tenant';
    }
    defaultProperties() {
        return ['id', 'displayName', 'mailNickname'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const filter = `?$filter=groupTypes/any(c:c+eq+'Unified')`;
                const displayNameFilter = args.options.displayName ? ` and startswith(DisplayName,'${encodeURIComponent(args.options.displayName).replace(/'/g, `''`)}')` : '';
                const mailNicknameFilter = args.options.mailNickname ? ` and startswith(MailNickname,'${encodeURIComponent(args.options.mailNickname).replace(/'/g, `''`)}')` : '';
                const topCount = '&$top=100';
                const endpoint = `${this.resource}/v1.0/directory/deletedItems/Microsoft.Graph.Group${filter}${displayNameFilter}${mailNicknameFilter}${topCount}`;
                const recycleBinItems = yield odata_1.odata.getAllItems(endpoint);
                logger.log(recycleBinItems);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadO365GroupRecycleBinItemListCommand_instances = new WeakSet(), _AadO365GroupRecycleBinItemListCommand_initTelemetry = function _AadO365GroupRecycleBinItemListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            displayName: typeof args.options.displayName !== 'undefined',
            mailNickname: typeof args.options.mailNickname !== 'undefined'
        });
    });
}, _AadO365GroupRecycleBinItemListCommand_initOptions = function _AadO365GroupRecycleBinItemListCommand_initOptions() {
    this.options.unshift({
        option: '-d, --displayName [displayName]'
    }, {
        option: '-m, --mailNickname [mailNickname]'
    });
};
module.exports = new AadO365GroupRecycleBinItemListCommand();
