// @pnp/cli-microsoft365/dist/m365/aad/commands/o365group/o365group-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadO365GroupRemoveCommand_instances, _AadO365GroupRemoveCommand_initTelemetry, _AadO365GroupRemoveCommand_initOptions, _AadO365GroupRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadO365GroupRemoveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadO365GroupRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadO365GroupRemoveCommand_instances, "m", _AadO365GroupRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _AadO365GroupRemoveCommand_instances, "m", _AadO365GroupRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _AadO365GroupRemoveCommand_instances, "m", _AadO365GroupRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.O365GROUP_REMOVE;
    }
    get description() {
        return 'Removes an Microsoft 365 Group';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeGroup = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing Microsoft 365 Group: ${args.options.id}...`);
                }
                try {
                    const requestOptions = {
                        url: `${this.resource}/v1.0/groups/${args.options.id}`,
                        headers: {
                            'accept': 'application/json;odata.metadata=none'
                        }
                    };
                    yield request_1.default.delete(requestOptions);
                    if (args.options.skipRecycleBin) {
                        const requestOptions2 = {
                            url: `${this.resource}/v1.0/directory/deletedItems/${args.options.id}`,
                            headers: {
                                'accept': 'application/json;odata.metadata=none'
                            }
                        };
                        yield request_1.default.delete(requestOptions2);
                    }
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeGroup();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the group ${args.options.id}?`
                });
                if (result.continue) {
                    yield removeGroup();
                }
            }
        });
    }
}
_AadO365GroupRemoveCommand_instances = new WeakSet(), _AadO365GroupRemoveCommand_initTelemetry = function _AadO365GroupRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString(),
            skipRecycleBin: (!(!args.options.skipRecycleBin)).toString()
        });
    });
}, _AadO365GroupRemoveCommand_initOptions = function _AadO365GroupRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '--confirm'
    }, {
        option: '--skipRecycleBin'
    });
}, _AadO365GroupRemoveCommand_initValidators = function _AadO365GroupRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new AadO365GroupRemoveCommand();
