// @pnp/cli-microsoft365/dist/m365/aad/commands/o365group/o365group-renew.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadO365GroupRenewCommand_instances, _AadO365GroupRenewCommand_initOptions, _AadO365GroupRenewCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadO365GroupRenewCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadO365GroupRenewCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadO365GroupRenewCommand_instances, "m", _AadO365GroupRenewCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _AadO365GroupRenewCommand_instances, "m", _AadO365GroupRenewCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.O365GROUP_RENEW;
    }
    get description() {
        return `Renews Microsoft 365 group's expiration`;
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Renewing Microsoft 365 group's expiration: ${args.options.id}...`);
            }
            try {
                const requestOptions = {
                    url: `${this.resource}/v1.0/groups/${args.options.id}/renew/`,
                    headers: {
                        'accept': 'application/json;odata.metadata=none'
                    }
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadO365GroupRenewCommand_instances = new WeakSet(), _AadO365GroupRenewCommand_initOptions = function _AadO365GroupRenewCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    });
}, _AadO365GroupRenewCommand_initValidators = function _AadO365GroupRenewCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new AadO365GroupRenewCommand();
