// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-webhook-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListWebhookListCommand_instances, _SpoListWebhookListCommand_initTelemetry, _SpoListWebhookListCommand_initOptions, _SpoListWebhookListCommand_initValidators, _SpoListWebhookListCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListWebhookListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListWebhookListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListWebhookListCommand_instances, "m", _SpoListWebhookListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListWebhookListCommand_instances, "m", _SpoListWebhookListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListWebhookListCommand_instances, "m", _SpoListWebhookListCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListWebhookListCommand_instances, "m", _SpoListWebhookListCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.LIST_WEBHOOK_LIST;
    }
    get description() {
        return 'Lists all webhooks for the specified list';
    }
    defaultProperties() {
        return ['id', 'clientState', 'expirationDateTime', 'resource'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.title && this.verbose) {
                logger.logToStderr(chalk.yellow(`Option 'title' is deprecated. Please use 'listTitle' instead`));
            }
            if (args.options.id && this.verbose) {
                logger.logToStderr(chalk.yellow(`Option 'id' is deprecated. Please use 'listId' instead`));
            }
            if (this.verbose) {
                logger.logToStderr(`Retrieving webhook information for list ${args.options.id || args.options.listId || args.options.listUrl} in site at ${args.options.webUrl}...`);
            }
            let requestUrl = `${args.options.webUrl}/_api/web`;
            if (args.options.id) {
                requestUrl += `/lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.id)}')/Subscriptions`;
            }
            else if (args.options.listId) {
                requestUrl += `/lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')/Subscriptions`;
            }
            else if (args.options.listTitle) {
                requestUrl += `/lists/GetByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')/Subscriptions`;
            }
            else if (args.options.title) {
                requestUrl += `/lists/GetByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.title)}')/Subscriptions`;
            }
            else if (args.options.listUrl) {
                const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                requestUrl += `/GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')/Subscriptions`;
            }
            const requestOptions = {
                url: requestUrl,
                method: 'GET',
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                if (res.value && res.value.length > 0) {
                    res.value.forEach(w => {
                        w.clientState = w.clientState || '';
                    });
                    logger.log(res.value);
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr('No webhooks found');
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoListWebhookListCommand_instances = new WeakSet(), _SpoListWebhookListCommand_initTelemetry = function _SpoListWebhookListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: typeof args.options.id !== 'undefined',
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            title: typeof args.options.title !== 'undefined'
        });
    });
}, _SpoListWebhookListCommand_initOptions = function _SpoListWebhookListCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-i, --listId [listId]'
    }, {
        option: '-t, --listTitle [listTitle]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '--id [id]'
    }, {
        option: '--title [title]'
    });
}, _SpoListWebhookListCommand_initValidators = function _SpoListWebhookListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.id) {
            if (!validation_1.validation.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
        }
        if (args.options.listId) {
            if (!validation_1.validation.isValidGuid(args.options.listId)) {
                return `${args.options.listId} is not a valid GUID`;
            }
        }
        return true;
    }));
}, _SpoListWebhookListCommand_initOptionSets = function _SpoListWebhookListCommand_initOptionSets() {
    this.optionSets.push(['id', 'title', 'listId', 'listTitle', 'listUrl']);
};
module.exports = new SpoListWebhookListCommand();
