// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-contenttype-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListContentTypeAddCommand_instances, _SpoListContentTypeAddCommand_initTelemetry, _SpoListContentTypeAddCommand_initOptions, _SpoListContentTypeAddCommand_initValidators, _SpoListContentTypeAddCommand_initTypes, _SpoListContentTypeAddCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListContentTypeAddCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListContentTypeAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListContentTypeAddCommand_instances, "m", _SpoListContentTypeAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeAddCommand_instances, "m", _SpoListContentTypeAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeAddCommand_instances, "m", _SpoListContentTypeAddCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeAddCommand_instances, "m", _SpoListContentTypeAddCommand_initTypes).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeAddCommand_instances, "m", _SpoListContentTypeAddCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.LIST_CONTENTTYPE_ADD;
    }
    get description() {
        return 'Adds content type to list';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                const list = (args.options.listId ? args.options.listId : args.options.listTitle ? args.options.listTitle : args.options.listUrl);
                logger.logToStderr(`Adding content type ${args.options.contentTypeId} to list ${list} in site at ${args.options.webUrl}...`);
            }
            let requestUrl = `${args.options.webUrl}/_api/web/`;
            if (args.options.listId) {
                requestUrl += `lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')`;
            }
            else if (args.options.listTitle) {
                requestUrl += `lists/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')`;
            }
            else if (args.options.listUrl) {
                const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                requestUrl += `GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')`;
            }
            const requestBody = {
                contentTypeId: args.options.contentTypeId
            };
            const requestOptions = {
                url: `${requestUrl}/ContentTypes/AddAvailableContentType`,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                data: requestBody,
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.post(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoListContentTypeAddCommand_instances = new WeakSet(), _SpoListContentTypeAddCommand_initTelemetry = function _SpoListContentTypeAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined'
        });
    });
}, _SpoListContentTypeAddCommand_initOptions = function _SpoListContentTypeAddCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-l, --listId [listId]'
    }, {
        option: '-t, --listTitle [listTitle]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '-c, --contentTypeId <contentTypeId>'
    });
}, _SpoListContentTypeAddCommand_initValidators = function _SpoListContentTypeAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId && !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} is not a valid GUID`;
        }
        return true;
    }));
}, _SpoListContentTypeAddCommand_initTypes = function _SpoListContentTypeAddCommand_initTypes() {
    this.types.string.push('contentTypeId', 'c');
}, _SpoListContentTypeAddCommand_initOptionSets = function _SpoListContentTypeAddCommand_initOptionSets() {
    this.optionSets.push(['listId', 'listTitle', 'listUrl']);
};
module.exports = new SpoListContentTypeAddCommand();
