// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-sitescript-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListSiteScriptGetCommand_instances, _SpoListSiteScriptGetCommand_initTelemetry, _SpoListSiteScriptGetCommand_initOptions, _SpoListSiteScriptGetCommand_initValidators, _SpoListSiteScriptGetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListSiteScriptGetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListSiteScriptGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListSiteScriptGetCommand_instances, "m", _SpoListSiteScriptGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListSiteScriptGetCommand_instances, "m", _SpoListSiteScriptGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListSiteScriptGetCommand_instances, "m", _SpoListSiteScriptGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListSiteScriptGetCommand_instances, "m", _SpoListSiteScriptGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.LIST_SITESCRIPT_GET;
    }
    get description() {
        return 'Extracts a site script from a SharePoint list';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (this.verbose) {
                    const list = (args.options.listId ? args.options.listId : args.options.listTitle);
                    logger.logToStderr(`Extracting Site Script from list ${list} in site at ${args.options.webUrl}...`);
                }
                let requestUrl = '';
                if (args.options.listId) {
                    if (this.debug) {
                        logger.logToStderr(`Retrieving List Url from Id '${args.options.listId}'...`);
                    }
                    requestUrl = `${args.options.webUrl}/_api/web/lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')?$expand=RootFolder`;
                }
                else {
                    if (this.debug) {
                        logger.logToStderr(`Retrieving List Url from Title '${args.options.listTitle}'...`);
                    }
                    requestUrl = `${args.options.webUrl}/_api/web/lists/GetByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')?$expand=RootFolder`;
                }
                let requestOptions = {
                    url: requestUrl,
                    headers: {
                        'accept': 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                const listInstance = yield request_1.default.get(requestOptions);
                const listAbsoluteUrl = urlUtil_1.urlUtil.getAbsoluteUrl(args.options.webUrl, listInstance.RootFolder.ServerRelativeUrl);
                requestUrl = `${args.options.webUrl}/_api/Microsoft_SharePoint_Utilities_WebTemplateExtensions_SiteScriptUtility_GetSiteScriptFromList`;
                requestOptions = {
                    url: requestUrl,
                    headers: {
                        'accept': 'application/json;odata=nometadata',
                        'content-type': 'application/json;odata=nometadata'
                    },
                    responseType: 'json',
                    data: {
                        listUrl: listAbsoluteUrl
                    }
                };
                const res = yield request_1.default.post(requestOptions);
                const siteScript = res.value;
                if (!siteScript) {
                    throw `An error has occurred, the site script could not be extracted from list '${args.options.listId || args.options.listTitle}'`;
                }
                logger.log(siteScript);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoListSiteScriptGetCommand_instances = new WeakSet(), _SpoListSiteScriptGetCommand_initTelemetry = function _SpoListSiteScriptGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: (!(!args.options.listId)).toString(),
            listTitle: (!(!args.options.listTitle)).toString()
        });
    });
}, _SpoListSiteScriptGetCommand_initOptions = function _SpoListSiteScriptGetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-l, --listId [listId]'
    }, {
        option: '-t, --listTitle [listTitle]'
    });
}, _SpoListSiteScriptGetCommand_initValidators = function _SpoListSiteScriptGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId) {
            if (!validation_1.validation.isValidGuid(args.options.listId)) {
                return `${args.options.listId} is not a valid GUID`;
            }
        }
        return true;
    }));
}, _SpoListSiteScriptGetCommand_initOptionSets = function _SpoListSiteScriptGetCommand_initOptionSets() {
    this.optionSets.push(['listId', 'listTitle']);
};
module.exports = new SpoListSiteScriptGetCommand();
