// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-view-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListViewSetCommand_instances, _SpoListViewSetCommand_initTelemetry, _SpoListViewSetCommand_initOptions, _SpoListViewSetCommand_initValidators, _SpoListViewSetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListViewSetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListViewSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListViewSetCommand_instances, "m", _SpoListViewSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListViewSetCommand_instances, "m", _SpoListViewSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListViewSetCommand_instances, "m", _SpoListViewSetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListViewSetCommand_instances, "m", _SpoListViewSetCommand_initOptionSets).call(this);
    }
    allowUnknownOptions() {
        return true;
    }
    get name() {
        return commands_1.default.LIST_VIEW_SET;
    }
    get description() {
        return 'Updates existing list view';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const baseRestUrl = `${args.options.webUrl}/_api/web/lists`;
            const listRestUrl = args.options.listId ?
                `(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')`
                : `/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')`;
            const viewRestUrl = `/views/${(args.options.viewId ? `getById('${formatting_1.formatting.encodeQueryParameter(args.options.viewId)}')` : `getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.viewTitle)}')`)}`;
            try {
                const res = yield spo_1.spo.getRequestDigest(args.options.webUrl);
                const requestOptions = {
                    url: `${baseRestUrl}${listRestUrl}${viewRestUrl}`,
                    headers: {
                        'X-RequestDigest': res.FormDigestValue,
                        'content-type': 'application/json;odata=nometadata',
                        accept: 'application/json;odata=nometadata'
                    },
                    responseType: 'json',
                    data: this.getPayload(args.options)
                };
                yield request_1.default.patch(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getPayload(options) {
        const payload = {};
        const excludeOptions = [
            'webUrl',
            'listId',
            'listTitle',
            'viewId',
            'viewTitle',
            'debug',
            'verbose',
            'output'
        ];
        Object.keys(options).forEach(key => {
            if (excludeOptions.indexOf(key) === -1) {
                payload[key] = options[key];
            }
        });
        return payload;
    }
}
_SpoListViewSetCommand_instances = new WeakSet(), _SpoListViewSetCommand_initTelemetry = function _SpoListViewSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            viewId: typeof args.options.viewId !== 'undefined',
            viewTitle: typeof args.options.viewTitle !== 'undefined'
        });
    });
}, _SpoListViewSetCommand_initOptions = function _SpoListViewSetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--listId [listId]'
    }, {
        option: '--listTitle [listTitle]'
    }, {
        option: '--viewId [viewId]'
    }, {
        option: '--viewTitle [viewTitle]'
    });
}, _SpoListViewSetCommand_initValidators = function _SpoListViewSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId &&
            !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} in option listId is not a valid GUID`;
        }
        if (args.options.viewId &&
            !validation_1.validation.isValidGuid(args.options.viewId)) {
            return `${args.options.viewId} in option viewId is not a valid GUID`;
        }
        return true;
    }));
}, _SpoListViewSetCommand_initOptionSets = function _SpoListViewSetCommand_initOptionSets() {
    this.optionSets.push(['listId', 'listTitle'], ['viewId', 'viewTitle']);
};
module.exports = new SpoListViewSetCommand();
