// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-contenttype-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListContentTypeRemoveCommand_instances, _SpoListContentTypeRemoveCommand_initTelemetry, _SpoListContentTypeRemoveCommand_initOptions, _SpoListContentTypeRemoveCommand_initValidators, _SpoListContentTypeRemoveCommand_initTypes, _SpoListContentTypeRemoveCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListContentTypeRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListContentTypeRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListContentTypeRemoveCommand_instances, "m", _SpoListContentTypeRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeRemoveCommand_instances, "m", _SpoListContentTypeRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeRemoveCommand_instances, "m", _SpoListContentTypeRemoveCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeRemoveCommand_instances, "m", _SpoListContentTypeRemoveCommand_initTypes).call(this);
        __classPrivateFieldGet(this, _SpoListContentTypeRemoveCommand_instances, "m", _SpoListContentTypeRemoveCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.LIST_CONTENTTYPE_REMOVE;
    }
    get description() {
        return 'Removes content type from list';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeContentTypeFromList = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    const list = (args.options.listId ? args.options.listId : args.options.listTitle ? args.options.listTitle : args.options.listUrl);
                    logger.logToStderr(`Removing content type ${args.options.contentTypeId} from list ${list} in site at ${args.options.webUrl}...`);
                }
                let requestUrl = `${args.options.webUrl}/_api/web/`;
                if (args.options.listId) {
                    requestUrl += `lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')`;
                }
                else if (args.options.listTitle) {
                    requestUrl += `lists/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')`;
                }
                else if (args.options.listUrl) {
                    const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                    requestUrl += `GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')`;
                }
                const requestOptions = {
                    url: `${requestUrl}/ContentTypes('${encodeURIComponent(args.options.contentTypeId)}')`,
                    headers: {
                        'X-HTTP-Method': 'DELETE',
                        'If-Match': '*',
                        'accept': 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeContentTypeFromList();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the content type ${args.options.contentTypeId} from the list ${args.options.listId ? args.options.listId : args.options.listTitle ? args.options.listTitle : args.options.listUrl} in site ${args.options.webUrl}?`
                });
                if (result.continue) {
                    yield removeContentTypeFromList();
                }
            }
        });
    }
}
_SpoListContentTypeRemoveCommand_instances = new WeakSet(), _SpoListContentTypeRemoveCommand_initTelemetry = function _SpoListContentTypeRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoListContentTypeRemoveCommand_initOptions = function _SpoListContentTypeRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-l, --listId [listId]'
    }, {
        option: '-t, --listTitle [listTitle]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '-c, --contentTypeId <contentTypeId>'
    }, {
        option: '--confirm'
    });
}, _SpoListContentTypeRemoveCommand_initValidators = function _SpoListContentTypeRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId) {
            if (!validation_1.validation.isValidGuid(args.options.listId)) {
                return `${args.options.listId} is not a valid GUID`;
            }
        }
        return true;
    }));
}, _SpoListContentTypeRemoveCommand_initTypes = function _SpoListContentTypeRemoveCommand_initTypes() {
    this.types.string.push('contentTypeId', 'c');
}, _SpoListContentTypeRemoveCommand_initOptionSets = function _SpoListContentTypeRemoveCommand_initOptionSets() {
    this.optionSets.push(['listId', 'listTitle', 'listUrl']);
};
module.exports = new SpoListContentTypeRemoveCommand();
