// @pnp/cli-microsoft365/dist/m365/spo/commands/list/list-view-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoListViewRemoveCommand_instances, _SpoListViewRemoveCommand_initTelemetry, _SpoListViewRemoveCommand_initOptions, _SpoListViewRemoveCommand_initValidators, _SpoListViewRemoveCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoListViewRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoListViewRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoListViewRemoveCommand_instances, "m", _SpoListViewRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoListViewRemoveCommand_instances, "m", _SpoListViewRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoListViewRemoveCommand_instances, "m", _SpoListViewRemoveCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoListViewRemoveCommand_instances, "m", _SpoListViewRemoveCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.LIST_VIEW_REMOVE;
    }
    get description() {
        return 'Deletes the specified view from the list';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeViewFromList = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    const list = (args.options.listId ? args.options.listId : args.options.listTitle);
                    logger.logToStderr(`Removing view ${args.options.viewId || args.options.viewTitle} from list ${list} in site at ${args.options.webUrl}...`);
                }
                let requestUrl = '';
                const listSelector = args.options.listId ? `(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')` : `/GetByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')`;
                const viewSelector = args.options.viewId ? `(guid'${formatting_1.formatting.encodeQueryParameter(args.options.viewId)}')` : `/GetByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.viewTitle)}')`;
                requestUrl = `${args.options.webUrl}/_api/web/lists${listSelector}/views${viewSelector}`;
                const requestOptions = {
                    url: requestUrl,
                    headers: {
                        'X-HTTP-Method': 'DELETE',
                        'If-Match': '*',
                        'accept': 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.post(requestOptions);
                    // REST post call doesn't return anything
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeViewFromList();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the view ${args.options.viewId || args.options.viewTitle} from the list ${args.options.listId || args.options.listTitle} in site ${args.options.webUrl}?`
                });
                if (result.continue) {
                    yield removeViewFromList();
                }
            }
        });
    }
}
_SpoListViewRemoveCommand_instances = new WeakSet(), _SpoListViewRemoveCommand_initTelemetry = function _SpoListViewRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            viewId: typeof args.options.viewId !== 'undefined',
            viewTitle: typeof args.options.viewTitle !== 'undefined',
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoListViewRemoveCommand_initOptions = function _SpoListViewRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--listId [listId]'
    }, {
        option: '--listTitle [listTitle]'
    }, {
        option: '--viewId [viewId]'
    }, {
        option: '--viewTitle [viewTitle]'
    }, {
        option: '--confirm'
    });
}, _SpoListViewRemoveCommand_initValidators = function _SpoListViewRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.listId) {
            if (!validation_1.validation.isValidGuid(args.options.listId)) {
                return `${args.options.listId} is not a valid GUID`;
            }
        }
        if (args.options.viewId) {
            if (!validation_1.validation.isValidGuid(args.options.viewId)) {
                return `${args.options.viewId} is not a valid GUID`;
            }
        }
        return true;
    }));
}, _SpoListViewRemoveCommand_initOptionSets = function _SpoListViewRemoveCommand_initOptionSets() {
    this.optionSets.push(['listId', 'listTitle'], ['viewId', 'viewTitle']);
};
module.exports = new SpoListViewRemoveCommand();
