// @pnp/cli-microsoft365/dist/m365/flow/commands/run/run-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FlowRunListCommand_instances, _FlowRunListCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const AzmgmtItemsListCommand_1 = require("../../../base/AzmgmtItemsListCommand");
const commands_1 = require("../../commands");
class FlowRunListCommand extends AzmgmtItemsListCommand_1.AzmgmtItemsListCommand {
    constructor() {
        super();
        _FlowRunListCommand_instances.add(this);
        __classPrivateFieldGet(this, _FlowRunListCommand_instances, "m", _FlowRunListCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.RUN_LIST;
    }
    get description() {
        return 'Lists runs of the specified Microsoft Flow';
    }
    defaultProperties() {
        return ['name', 'startTime', 'status'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving list of runs for Microsoft Flow ${args.options.flow}...`);
            }
            const url = `${this.resource}providers/Microsoft.ProcessSimple/environments/${encodeURIComponent(args.options.environment)}/flows/${encodeURIComponent(args.options.flow)}/runs?api-version=2016-11-01`;
            try {
                yield this.getAllItems(url, logger, true);
                if (this.items.length > 0) {
                    this.items.forEach(i => {
                        i.startTime = i.properties.startTime;
                        i.status = i.properties.status;
                    });
                    logger.log(this.items);
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr('No runs found');
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_FlowRunListCommand_instances = new WeakSet(), _FlowRunListCommand_initOptions = function _FlowRunListCommand_initOptions() {
    this.options.unshift({
        option: '-f, --flow <flow>'
    }, {
        option: '-e, --environment <environment>'
    });
};
module.exports = new FlowRunListCommand();
