// @pnp/cli-microsoft365/dist/m365/teams/commands/membersettings/membersettings-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsMemberSettingsListCommand_instances, _TeamsMemberSettingsListCommand_initOptions, _TeamsMemberSettingsListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsMemberSettingsListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsMemberSettingsListCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsMemberSettingsListCommand_instances, "m", _TeamsMemberSettingsListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsMemberSettingsListCommand_instances, "m", _TeamsMemberSettingsListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.MEMBERSETTINGS_LIST;
    }
    get description() {
        return 'Lists member settings for a Microsoft Teams team';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}?$select=memberSettings`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res.memberSettings);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_TeamsMemberSettingsListCommand_instances = new WeakSet(), _TeamsMemberSettingsListCommand_initOptions = function _TeamsMemberSettingsListCommand_initOptions() {
    this.options.unshift({
        option: '-i, --teamId <teamId>'
    });
}, _TeamsMemberSettingsListCommand_initValidators = function _TeamsMemberSettingsListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new TeamsMemberSettingsListCommand();
