// @pnp/cli-microsoft365/dist/m365/yammer/commands/group/group-user-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YammerGroupUserAddCommand_instances, _YammerGroupUserAddCommand_initTelemetry, _YammerGroupUserAddCommand_initOptions, _YammerGroupUserAddCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const YammerCommand_1 = require("../../../base/YammerCommand");
const commands_1 = require("../../commands");
class YammerGroupUserAddCommand extends YammerCommand_1.default {
    constructor() {
        super();
        _YammerGroupUserAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _YammerGroupUserAddCommand_instances, "m", _YammerGroupUserAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _YammerGroupUserAddCommand_instances, "m", _YammerGroupUserAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _YammerGroupUserAddCommand_instances, "m", _YammerGroupUserAddCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.GROUP_USER_ADD;
    }
    get description() {
        return 'Adds a user to a Yammer Group';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${this.resource}/v1/group_memberships.json`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json;odata=nometadata'
                },
                responseType: 'json',
                data: {
                    group_id: args.options.id,
                    user_id: args.options.userId,
                    email: args.options.email
                }
            };
            try {
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_YammerGroupUserAddCommand_instances = new WeakSet(), _YammerGroupUserAddCommand_initTelemetry = function _YammerGroupUserAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            userId: typeof args.options.userId !== 'undefined',
            email: typeof args.options.email !== 'undefined'
        });
    });
}, _YammerGroupUserAddCommand_initOptions = function _YammerGroupUserAddCommand_initOptions() {
    this.options.unshift({
        option: '--id <id>'
    }, {
        option: '--userId [userId]'
    }, {
        option: '--email [email]'
    });
}, _YammerGroupUserAddCommand_initValidators = function _YammerGroupUserAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.id && typeof args.options.id !== 'number') {
            return `${args.options.id} is not a number`;
        }
        if (args.options.userId && typeof args.options.userId !== 'number') {
            return `${args.options.userId} is not a number`;
        }
        return true;
    }));
};
module.exports = new YammerGroupUserAddCommand();
