// @pnp/cli-microsoft365/dist/m365/yammer/commands/group/group-user-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YammerGroupUserRemoveCommand_instances, _YammerGroupUserRemoveCommand_initTelemetry, _YammerGroupUserRemoveCommand_initOptions, _YammerGroupUserRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const YammerCommand_1 = require("../../../base/YammerCommand");
const commands_1 = require("../../commands");
class YammerGroupUserRemoveCommand extends YammerCommand_1.default {
    constructor() {
        super();
        _YammerGroupUserRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _YammerGroupUserRemoveCommand_instances, "m", _YammerGroupUserRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _YammerGroupUserRemoveCommand_instances, "m", _YammerGroupUserRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _YammerGroupUserRemoveCommand_instances, "m", _YammerGroupUserRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.GROUP_USER_REMOVE;
    }
    get description() {
        return 'Removes a user from a Yammer group';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const executeRemoveAction = () => __awaiter(this, void 0, void 0, function* () {
                const endpoint = `${this.resource}/v1/group_memberships.json`;
                const requestOptions = {
                    url: endpoint,
                    headers: {
                        accept: 'application/json;odata.metadata=none',
                        'content-type': 'application/json;odata=nometadata'
                    },
                    responseType: 'json',
                    data: {
                        group_id: args.options.id,
                        user_id: args.options.userId
                    }
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield executeRemoveAction();
            }
            else {
                let messagePrompt = `Are you sure you want to leave group ${args.options.id}?`;
                if (args.options.userId) {
                    messagePrompt = `Are you sure you want to remove the user ${args.options.userId} from the group ${args.options.id}?`;
                }
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: messagePrompt
                });
                if (result.continue) {
                    yield executeRemoveAction();
                }
            }
        });
    }
}
_YammerGroupUserRemoveCommand_instances = new WeakSet(), _YammerGroupUserRemoveCommand_initTelemetry = function _YammerGroupUserRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            userId: args.options.userId !== undefined,
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _YammerGroupUserRemoveCommand_initOptions = function _YammerGroupUserRemoveCommand_initOptions() {
    this.options.unshift({
        option: '--id <id>'
    }, {
        option: '--userId [userId]'
    }, {
        option: '--confirm'
    });
}, _YammerGroupUserRemoveCommand_initValidators = function _YammerGroupUserRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.id && typeof args.options.id !== 'number') {
            return `${args.options.id} is not a number`;
        }
        if (args.options.userId && typeof args.options.userId !== 'number') {
            return `${args.options.userId} is not a number`;
        }
        return true;
    }));
};
module.exports = new YammerGroupUserRemoveCommand();
