// @pnp/cli-microsoft365/dist/m365/flow/commands/flow-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FlowListCommand_instances, _FlowListCommand_initTelemetry, _FlowListCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const AzmgmtItemsListCommand_1 = require("../../base/AzmgmtItemsListCommand");
const commands_1 = require("../commands");
class FlowListCommand extends AzmgmtItemsListCommand_1.AzmgmtItemsListCommand {
    constructor() {
        super();
        _FlowListCommand_instances.add(this);
        __classPrivateFieldGet(this, _FlowListCommand_instances, "m", _FlowListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _FlowListCommand_instances, "m", _FlowListCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.LIST;
    }
    get description() {
        return 'Lists Microsoft Flows in the given environment';
    }
    defaultProperties() {
        return ['name', 'displayName'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = `${this.resource}providers/Microsoft.ProcessSimple${args.options.asAdmin ? '/scopes/admin' : ''}/environments/${encodeURIComponent(args.options.environment)}/flows?api-version=2016-11-01`;
            try {
                yield this.getAllItems(url, logger, true);
                if (this.items.length > 0) {
                    this.items.forEach(i => {
                        i.displayName = i.properties.displayName;
                    });
                    logger.log(this.items);
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr('No Flows found');
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_FlowListCommand_instances = new WeakSet(), _FlowListCommand_initTelemetry = function _FlowListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            asAdmin: args.options.asAdmin === true
        });
    });
}, _FlowListCommand_initOptions = function _FlowListCommand_initOptions() {
    this.options.unshift({
        option: '-e, --environment <environment>'
    }, {
        option: '--asAdmin'
    });
};
module.exports = new FlowListCommand();
