// @pnp/cli-microsoft365/dist/m365/flow/commands/flow-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FlowGetCommand_instances, _FlowGetCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../request");
const AzmgmtCommand_1 = require("../../base/AzmgmtCommand");
const commands_1 = require("../commands");
class FlowGetCommand extends AzmgmtCommand_1.default {
    constructor() {
        super();
        _FlowGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _FlowGetCommand_instances, "m", _FlowGetCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.GET;
    }
    get description() {
        return 'Gets information about the specified Microsoft Flow';
    }
    defaultProperties() {
        return ['name', 'displayName', 'description', 'triggers', 'actions'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving information about Microsoft Flow ${args.options.name}...`);
            }
            const requestOptions = {
                url: `${this.resource}providers/Microsoft.ProcessSimple/${args.options.asAdmin ? 'scopes/admin/' : ''}environments/${encodeURIComponent(args.options.environment)}/flows/${encodeURIComponent(args.options.name)}?api-version=2016-11-01`,
                headers: {
                    accept: 'application/json'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                res.displayName = res.properties.displayName;
                res.description = res.properties.definitionSummary.description || '';
                res.triggers = res.properties.definitionSummary.triggers.map((t) => (t.type + (t.kind ? "-" + t.kind : ''))).join(', ');
                res.actions = res.properties.definitionSummary.actions.map((a) => (a.type + (a.swaggerOperationId ? "-" + a.swaggerOperationId : ''))).join(', ');
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_FlowGetCommand_instances = new WeakSet(), _FlowGetCommand_initOptions = function _FlowGetCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    }, {
        option: '-e, --environment <environment>'
    }, {
        option: '--asAdmin'
    });
};
module.exports = new FlowGetCommand();
