// @pnp/cli-microsoft365/dist/m365/flow/commands/environment/environment-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FlowEnvironmentGetCommand_instances, _FlowEnvironmentGetCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const AzmgmtCommand_1 = require("../../../base/AzmgmtCommand");
const commands_1 = require("../../commands");
class FlowEnvironmentGetCommand extends AzmgmtCommand_1.default {
    constructor() {
        super();
        _FlowEnvironmentGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _FlowEnvironmentGetCommand_instances, "m", _FlowEnvironmentGetCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.ENVIRONMENT_GET;
    }
    get description() {
        return 'Gets information about the specified Microsoft Flow environment';
    }
    defaultProperties() {
        return ['name', 'id', 'location', 'displayName', 'provisioningState', 'environmentSku', 'azureRegionHint', 'isDefault'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving information about Microsoft Flow environment ${args.options.name}...`);
            }
            const requestOptions = {
                url: `${this.resource}providers/Microsoft.ProcessSimple/environments/${encodeURIComponent(args.options.name)}?api-version=2016-11-01`,
                headers: {
                    accept: 'application/json'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                res.displayName = res.properties.displayName;
                res.provisioningState = res.properties.provisioningState;
                res.environmentSku = res.properties.environmentSku;
                res.azureRegionHint = res.properties.azureRegionHint;
                res.isDefault = res.properties.isDefault;
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_FlowEnvironmentGetCommand_instances = new WeakSet(), _FlowEnvironmentGetCommand_initOptions = function _FlowEnvironmentGetCommand_initOptions() {
    this.options.unshift({
        option: '-n, --name <name>'
    });
};
module.exports = new FlowEnvironmentGetCommand();
