// @pnp/cli-microsoft365/dist/m365/flow/commands/environment/environment-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const AzmgmtCommand_1 = require("../../../base/AzmgmtCommand");
const commands_1 = require("../../commands");
class FlowEnvironmentListCommand extends AzmgmtCommand_1.default {
    get name() {
        return commands_1.default.ENVIRONMENT_LIST;
    }
    get description() {
        return 'Lists Microsoft Flow environments in the current tenant';
    }
    defaultProperties() {
        return ['name', 'displayName'];
    }
    commandAction(logger) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving list of Microsoft Flow environments...`);
            }
            const requestOptions = {
                url: `${this.resource}providers/Microsoft.ProcessSimple/environments?api-version=2016-11-01`,
                headers: {
                    accept: 'application/json'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                if (res.value && res.value.length > 0) {
                    res.value.forEach(e => {
                        e.displayName = e.properties.displayName;
                    });
                    logger.log(res.value);
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
module.exports = new FlowEnvironmentListCommand();
