// @pnp/cli-microsoft365/dist/m365/spo/commands/group/group-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoGroupRemoveCommand_instances, _SpoGroupRemoveCommand_initTelemetry, _SpoGroupRemoveCommand_initOptions, _SpoGroupRemoveCommand_initValidators, _SpoGroupRemoveCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoGroupRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoGroupRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoGroupRemoveCommand_instances, "m", _SpoGroupRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoGroupRemoveCommand_instances, "m", _SpoGroupRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoGroupRemoveCommand_instances, "m", _SpoGroupRemoveCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoGroupRemoveCommand_instances, "m", _SpoGroupRemoveCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.GROUP_REMOVE;
    }
    get description() {
        return 'Removes group from specific web';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeGroup = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing group in web at ${args.options.webUrl}...`);
                }
                try {
                    let groupId;
                    if (args.options.name) {
                        const requestOptions = {
                            url: `${args.options.webUrl}/_api/web/sitegroups/GetByName('${args.options.name}')?$select=Id`,
                            headers: {
                                accept: 'application/json'
                            },
                            responseType: 'json'
                        };
                        const group = yield request_1.default.get(requestOptions);
                        groupId = group.Id;
                    }
                    else {
                        groupId = args.options.id;
                    }
                    const requestUrl = `${args.options.webUrl}/_api/web/sitegroups/RemoveById(${groupId})`;
                    const requestOptions = {
                        url: requestUrl,
                        method: 'POST',
                        headers: {
                            'content-length': 0,
                            'accept': 'application/json'
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.post(requestOptions);
                    // REST post call doesn't return anything
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeGroup();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the group ${args.options.id || args.options.name} from web ${args.options.webUrl}?`
                });
                if (result.continue) {
                    yield removeGroup();
                }
            }
        });
    }
}
_SpoGroupRemoveCommand_instances = new WeakSet(), _SpoGroupRemoveCommand_initTelemetry = function _SpoGroupRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: (!(!args.options.id)).toString(),
            name: (!(!args.options.name)).toString(),
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoGroupRemoveCommand_initOptions = function _SpoGroupRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--id [id]'
    }, {
        option: '--name [name]'
    }, {
        option: '--confirm'
    });
}, _SpoGroupRemoveCommand_initValidators = function _SpoGroupRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.id && typeof args.options.id !== 'number') {
            return `${args.options.id} is not a number`;
        }
        return true;
    }));
}, _SpoGroupRemoveCommand_initOptionSets = function _SpoGroupRemoveCommand_initOptionSets() {
    this.optionSets.push(['id', 'name']);
};
module.exports = new SpoGroupRemoveCommand();
