// @pnp/cli-microsoft365/dist/m365/spfx/commands/project/project-model/TsFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsFile = void 0;
const fs = require("fs");
const path = require("path");
const ts = require("typescript");
class TsFile {
    constructor(path) {
        this.path = path;
    }
    get sourceFile() {
        if (!this._sourceFile) {
            if (!this.source) {
                return undefined;
            }
            try {
                this._sourceFile = ts.createSourceFile(path.basename(this.path), this.source, ts.ScriptTarget.Latest, true);
            }
            catch (_a) { }
        }
        return this._sourceFile;
    }
    get nodes() {
        if (!this._nodes) {
            if (!this.sourceFile) {
                return undefined;
            }
            this._nodes = TsFile.getAsEnumerable(this.sourceFile, this.sourceFile);
        }
        return this._nodes;
    }
    get source() {
        if (!this._source) {
            try {
                this._source = fs.readFileSync(this.path, 'utf-8');
            }
            catch (_a) { }
        }
        return this._source;
    }
    static getAsEnumerable(file, node) {
        const nodes = [node];
        node.getChildren(file).forEach(n => {
            nodes.push(...TsFile.getAsEnumerable(file, n));
        });
        return nodes;
    }
}
exports.TsFile = TsFile;
