// @pnp/cli-microsoft365/dist/m365/spo/commands/app/app-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoAppListCommand_instances, _SpoAppListCommand_initTelemetry, _SpoAppListCommand_initOptions, _SpoAppListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const commands_1 = require("../../commands");
const SpoAppBaseCommand_1 = require("./SpoAppBaseCommand");
class SpoAppListCommand extends SpoAppBaseCommand_1.SpoAppBaseCommand {
    constructor() {
        super();
        _SpoAppListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoAppListCommand_instances, "m", _SpoAppListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoAppListCommand_instances, "m", _SpoAppListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoAppListCommand_instances, "m", _SpoAppListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.APP_LIST;
    }
    get description() {
        return 'Lists apps from the specified app catalog';
    }
    defaultProperties() {
        return [`Title`, `ID`, `Deployed`, `AppCatalogVersion`];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
            let appCatalogSiteUrl = '';
            let spoUrl = '';
            try {
                spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                appCatalogSiteUrl = yield this.getAppCatalogSiteUrl(logger, spoUrl, args);
                if (this.verbose) {
                    logger.logToStderr(`Retrieving apps...`);
                }
                const requestOptions = {
                    url: `${appCatalogSiteUrl}/_api/web/${scope}appcatalog/AvailableApps`,
                    headers: {
                        accept: 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                const apps = yield request_1.default.get(requestOptions);
                if (apps.value && apps.value.length > 0) {
                    logger.log(apps.value);
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr('No apps found');
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoAppListCommand_instances = new WeakSet(), _SpoAppListCommand_initTelemetry = function _SpoAppListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            appCatalogUrl: (!(!args.options.appCatalogUrl)).toString(),
            scope: args.options.scope || 'tenant'
        });
    });
}, _SpoAppListCommand_initOptions = function _SpoAppListCommand_initOptions() {
    this.options.unshift({
        option: '-s, --scope [scope]',
        autocomplete: ['tenant', 'sitecollection']
    }, {
        option: '-u, --appCatalogUrl [appCatalogUrl]'
    });
}, _SpoAppListCommand_initValidators = function _SpoAppListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        // verify either 'tenant' or 'sitecollection' specified if scope provided
        if (args.options.scope) {
            const testScope = args.options.scope.toLowerCase();
            if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                return `Scope must be either 'tenant' or 'sitecollection'`;
            }
            if (testScope === 'sitecollection' && !args.options.appCatalogUrl) {
                return `You must specify appCatalogUrl when the scope is sitecollection`;
            }
            if (args.options.appCatalogUrl) {
                return validation_1.validation.isValidSharePointUrl(args.options.appCatalogUrl);
            }
        }
        return true;
    }));
};
module.exports = new SpoAppListCommand();
