// @pnp/cli-microsoft365/dist/m365/spo/commands/app/app-uninstall.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoAppUninstallCommand_instances, _SpoAppUninstallCommand_initTelemetry, _SpoAppUninstallCommand_initOptions, _SpoAppUninstallCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoAppUninstallCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoAppUninstallCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoAppUninstallCommand_instances, "m", _SpoAppUninstallCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoAppUninstallCommand_instances, "m", _SpoAppUninstallCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoAppUninstallCommand_instances, "m", _SpoAppUninstallCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.APP_UNINSTALL;
    }
    get description() {
        return 'Uninstalls an app from the site';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const uninstallApp = () => __awaiter(this, void 0, void 0, function* () {
                const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
                if (this.verbose) {
                    logger.logToStderr(`Uninstalling app '${args.options.id}' from the site '${args.options.siteUrl}'...`);
                }
                const requestOptions = {
                    url: `${args.options.siteUrl}/_api/web/${scope}appcatalog/AvailableApps/GetById('${encodeURIComponent(args.options.id)}')/uninstall`,
                    headers: {
                        accept: 'application/json;odata=nometadata'
                    }
                };
                try {
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataPromise(err);
                }
            });
            if (args.options.confirm) {
                yield uninstallApp();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to uninstall the app ${args.options.id} from site ${args.options.siteUrl}?`
                });
                if (result.continue) {
                    yield uninstallApp();
                }
            }
        });
    }
}
_SpoAppUninstallCommand_instances = new WeakSet(), _SpoAppUninstallCommand_initTelemetry = function _SpoAppUninstallCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString(),
            scope: args.options.scope || 'tenant'
        });
    });
}, _SpoAppUninstallCommand_initOptions = function _SpoAppUninstallCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id <id>'
    }, {
        option: '-s, --siteUrl <siteUrl>'
    }, {
        option: '--scope [scope]',
        autocomplete: ['tenant', 'sitecollection']
    }, {
        option: '--confirm'
    });
}, _SpoAppUninstallCommand_initValidators = function _SpoAppUninstallCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.scope) {
            const testScope = args.options.scope.toLowerCase();
            if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                return `Scope must be either 'tenant' or 'sitecollection' if specified`;
            }
        }
        if (!validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return validation_1.validation.isValidSharePointUrl(args.options.siteUrl);
    }));
};
module.exports = new SpoAppUninstallCommand();
