// @pnp/cli-microsoft365/dist/m365/spo/commands/app/app-deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoAppDeployCommand_instances, _SpoAppDeployCommand_initTelemetry, _SpoAppDeployCommand_initOptions, _SpoAppDeployCommand_initValidators, _SpoAppDeployCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const commands_1 = require("../../commands");
const SpoAppBaseCommand_1 = require("./SpoAppBaseCommand");
class SpoAppDeployCommand extends SpoAppBaseCommand_1.SpoAppBaseCommand {
    constructor() {
        super();
        _SpoAppDeployCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoAppDeployCommand_instances, "m", _SpoAppDeployCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoAppDeployCommand_instances, "m", _SpoAppDeployCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoAppDeployCommand_instances, "m", _SpoAppDeployCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoAppDeployCommand_instances, "m", _SpoAppDeployCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.APP_DEPLOY;
    }
    get description() {
        return 'Deploys the specified app in the specified app catalog';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            let appId = '';
            const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
            let appCatalogUrl = '';
            try {
                const spoUrl = yield spo_1.spo.getSpoUrl(logger, this.debug);
                appCatalogUrl = yield this.getAppCatalogSiteUrl(logger, spoUrl, args);
                let res;
                if (args.options.id) {
                    if (this.verbose) {
                        logger.logToStderr(`Using the specified app id ${args.options.id}`);
                    }
                    res = { UniqueId: args.options.id };
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr(`Looking up app id for app named ${args.options.name}...`);
                    }
                    const requestOptions = {
                        url: `${appCatalogUrl}/_api/web/getfolderbyserverrelativeurl('AppCatalog')/files('${args.options.name}')?$select=UniqueId`,
                        headers: {
                            accept: 'application/json;odata=nometadata'
                        },
                        responseType: 'json'
                    };
                    res = yield request_1.default.get(requestOptions);
                }
                appId = res.UniqueId;
                if (this.verbose) {
                    logger.logToStderr(`Deploying app...`);
                }
                const requestOptions = {
                    url: `${appCatalogUrl}/_api/web/${scope}appcatalog/AvailableApps/GetById('${appId}')/deploy`,
                    headers: {
                        accept: 'application/json;odata=nometadata',
                        'content-type': 'application/json;odata=nometadata;charset=utf-8'
                    },
                    data: { 'skipFeatureDeployment': args.options.skipFeatureDeployment || false },
                    responseType: 'json'
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoAppDeployCommand_instances = new WeakSet(), _SpoAppDeployCommand_initTelemetry = function _SpoAppDeployCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: (!(!args.options.id)).toString(),
            name: (!(!args.options.name)).toString(),
            appCatalogUrl: (!(!args.options.appCatalogUrl)).toString(),
            skipFeatureDeployment: args.options.skipFeatureDeployment || false,
            scope: (!(!args.options.scope)).toString()
        });
    });
}, _SpoAppDeployCommand_initOptions = function _SpoAppDeployCommand_initOptions() {
    this.options.unshift({
        option: '-i, --id [id]'
    }, {
        option: '-n, --name [name]'
    }, {
        option: '-u, --appCatalogUrl [appCatalogUrl]'
    }, {
        option: '--skipFeatureDeployment'
    }, {
        option: '-s, --scope [scope]',
        autocomplete: ['tenant', 'sitecollection']
    });
}, _SpoAppDeployCommand_initValidators = function _SpoAppDeployCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        // verify either 'tenant' or 'sitecollection' specified if scope provided
        if (args.options.scope) {
            const testScope = args.options.scope.toLowerCase();
            if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                return `Scope must be either 'tenant' or 'sitecollection'`;
            }
            if (testScope === 'sitecollection' && !args.options.appCatalogUrl) {
                return `You must specify appCatalogUrl when the scope is sitecollection`;
            }
        }
        if (args.options.id && !validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        if (args.options.appCatalogUrl) {
            return validation_1.validation.isValidSharePointUrl(args.options.appCatalogUrl);
        }
        return true;
    }));
}, _SpoAppDeployCommand_initOptionSets = function _SpoAppDeployCommand_initOptionSets() {
    this.optionSets.push(['id', 'name']);
};
module.exports = new SpoAppDeployCommand();
