// express-msteams-host/lib/decorators/index.js
"use strict";
// Copyright (c) Wictor Wilén. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BotCallingWebhook = exports.PreventIframe = exports.MessageExtensionDeclaration = exports.OutgoingWebhookDeclaration = exports.ConnectorDeclaration = exports.BotDeclaration = void 0;
var BotDeclaration_1 = require("./BotDeclaration");
Object.defineProperty(exports, "BotDeclaration", { enumerable: true, get: function () { return BotDeclaration_1.BotDeclaration; } });
var ConnectorDeclaration_1 = require("./ConnectorDeclaration");
Object.defineProperty(exports, "ConnectorDeclaration", { enumerable: true, get: function () { return ConnectorDeclaration_1.ConnectorDeclaration; } });
var OutgoingWebhookDeclaration_1 = require("./OutgoingWebhookDeclaration");
Object.defineProperty(exports, "OutgoingWebhookDeclaration", { enumerable: true, get: function () { return OutgoingWebhookDeclaration_1.OutgoingWebhookDeclaration; } });
var MessageExtensionDeclaration_1 = require("./MessageExtensionDeclaration");
Object.defineProperty(exports, "MessageExtensionDeclaration", { enumerable: true, get: function () { return MessageExtensionDeclaration_1.MessageExtensionDeclaration; } });
var PreventIframe_1 = require("./PreventIframe");
Object.defineProperty(exports, "PreventIframe", { enumerable: true, get: function () { return PreventIframe_1.PreventIframe; } });
var BotCallingWebhook_1 = require("./BotCallingWebhook");
Object.defineProperty(exports, "BotCallingWebhook", { enumerable: true, get: function () { return BotCallingWebhook_1.BotCallingWebhook; } });
