// @pnp/cli-microsoft365/dist/m365/aad/commands/sp/sp-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadSpAddCommand_instances, _AadSpAddCommand_initTelemetry, _AadSpAddCommand_initOptions, _AadSpAddCommand_initValidators, _AadSpAddCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadSpAddCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadSpAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadSpAddCommand_instances, "m", _AadSpAddCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _AadSpAddCommand_instances, "m", _AadSpAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _AadSpAddCommand_instances, "m", _AadSpAddCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _AadSpAddCommand_instances, "m", _AadSpAddCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.SP_ADD;
    }
    get description() {
        return 'Adds a service principal to a registered Azure AD app';
    }
    getAppId(args) {
        if (args.options.appId) {
            return Promise.resolve(args.options.appId);
        }
        let spMatchQuery = '';
        if (args.options.appName) {
            spMatchQuery = `displayName eq '${encodeURIComponent(args.options.appName)}'`;
        }
        else if (args.options.objectId) {
            spMatchQuery = `id eq '${encodeURIComponent(args.options.objectId)}'`;
        }
        const appIdRequestOptions = {
            url: `${this.resource}/v1.0/applications?$filter=${spMatchQuery}`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(appIdRequestOptions)
            .then(response => {
            const spItem = response.value[0];
            if (!spItem) {
                return Promise.reject(`The specified Azure AD app doesn't exist`);
            }
            if (response.value.length > 1) {
                return Promise.reject(`Multiple Azure AD apps with name ${args.options.appName} found: ${response.value.map(x => x.appId)}`);
            }
            return Promise.resolve(spItem.appId);
        });
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const appId = yield this.getAppId(args);
                const requestOptions = {
                    url: `${this.resource}/v1.0/servicePrincipals`,
                    headers: {
                        accept: 'application/json;odata.metadata=none',
                        'content-type': 'application/json;odata=nometadata'
                    },
                    data: {
                        appId: appId
                    },
                    responseType: 'json'
                };
                const res = yield request_1.default.post(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadSpAddCommand_instances = new WeakSet(), _AadSpAddCommand_initTelemetry = function _AadSpAddCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            appId: (!(!args.options.appId)).toString(),
            appName: (!(!args.options.appName)).toString(),
            objectId: (!(!args.options.objectId)).toString()
        });
    });
}, _AadSpAddCommand_initOptions = function _AadSpAddCommand_initOptions() {
    this.options.unshift({
        option: '--appId [appId]'
    }, {
        option: '--appName [appName]'
    }, {
        option: '--objectId [objectId]'
    });
}, _AadSpAddCommand_initValidators = function _AadSpAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.appId && !validation_1.validation.isValidGuid(args.options.appId)) {
            return `${args.options.appId} is not a valid appId GUID`;
        }
        if (args.options.objectId && !validation_1.validation.isValidGuid(args.options.objectId)) {
            return `${args.options.objectId} is not a valid objectId GUID`;
        }
        return true;
    }));
}, _AadSpAddCommand_initOptionSets = function _AadSpAddCommand_initOptionSets() {
    this.optionSets.push(['appId', 'appName', 'objectId']);
};
module.exports = new AadSpAddCommand();
