// @pnp/cli-microsoft365/dist/m365/spo/commands/hubsite/hubsite-unregister.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoHubSiteUnregisterCommand_instances, _SpoHubSiteUnregisterCommand_initTelemetry, _SpoHubSiteUnregisterCommand_initOptions, _SpoHubSiteUnregisterCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoHubSiteUnregisterCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoHubSiteUnregisterCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoHubSiteUnregisterCommand_instances, "m", _SpoHubSiteUnregisterCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteUnregisterCommand_instances, "m", _SpoHubSiteUnregisterCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteUnregisterCommand_instances, "m", _SpoHubSiteUnregisterCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.HUBSITE_UNREGISTER;
    }
    get description() {
        return 'Unregisters the specified site collection as a hub site';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const unregisterHubSite = () => __awaiter(this, void 0, void 0, function* () {
                try {
                    const res = yield spo_1.spo.getRequestDigest(args.options.url);
                    const requestOptions = {
                        url: `${args.options.url}/_api/site/UnregisterHubSite`,
                        headers: {
                            'X-RequestDigest': res.FormDigestValue,
                            accept: 'application/json;odata=nometadata'
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield unregisterHubSite();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to unregister the site collection ${args.options.url} as a hub site?`
                });
                if (result.continue) {
                    yield unregisterHubSite();
                }
            }
        });
    }
}
_SpoHubSiteUnregisterCommand_instances = new WeakSet(), _SpoHubSiteUnregisterCommand_initTelemetry = function _SpoHubSiteUnregisterCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: args.options.confirm || false
        });
    });
}, _SpoHubSiteUnregisterCommand_initOptions = function _SpoHubSiteUnregisterCommand_initOptions() {
    this.options.unshift({
        option: '-u, --url <url>'
    }, {
        option: '--confirm'
    });
}, _SpoHubSiteUnregisterCommand_initValidators = function _SpoHubSiteUnregisterCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.url); }));
};
module.exports = new SpoHubSiteUnregisterCommand();
