// @pnp/cli-microsoft365/dist/m365/spo/commands/hubsite/hubsite-theme-sync.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoHubSiteThemeSyncCommand_instances, _SpoHubSiteThemeSyncCommand_initOptions, _SpoHubSiteThemeSyncCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoHubSiteThemeSyncCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoHubSiteThemeSyncCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoHubSiteThemeSyncCommand_instances, "m", _SpoHubSiteThemeSyncCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteThemeSyncCommand_instances, "m", _SpoHubSiteThemeSyncCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.HUBSITE_THEME_SYNC;
    }
    get description() {
        return 'Applies any theme updates from the parent hub site.';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (this.verbose) {
                    logger.logToStderr('Syncing hub site theme...');
                }
                const requestOptions = {
                    url: `${args.options.webUrl}/_api/web/SyncHubSiteTheme`,
                    headers: {
                        accept: 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoHubSiteThemeSyncCommand_instances = new WeakSet(), _SpoHubSiteThemeSyncCommand_initOptions = function _SpoHubSiteThemeSyncCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    });
}, _SpoHubSiteThemeSyncCommand_initValidators = function _SpoHubSiteThemeSyncCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.webUrl); }));
};
module.exports = new SpoHubSiteThemeSyncCommand();
