// @pnp/cli-microsoft365/dist/m365/spo/commands/hubsite/hubsite-register.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoHubSiteRegisterCommand_instances, _SpoHubSiteRegisterCommand_initOptions, _SpoHubSiteRegisterCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoHubSiteRegisterCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoHubSiteRegisterCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoHubSiteRegisterCommand_instances, "m", _SpoHubSiteRegisterCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteRegisterCommand_instances, "m", _SpoHubSiteRegisterCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.HUBSITE_REGISTER;
    }
    get description() {
        return 'Registers the specified site collection as a hub site';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const reqDigest = yield spo_1.spo.getRequestDigest(args.options.url);
                const requestOptions = {
                    url: `${args.options.url}/_api/site/RegisterHubSite`,
                    headers: {
                        'X-RequestDigest': reqDigest.FormDigestValue,
                        accept: 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                const res = yield request_1.default.post(requestOptions);
                logger.log(res);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoHubSiteRegisterCommand_instances = new WeakSet(), _SpoHubSiteRegisterCommand_initOptions = function _SpoHubSiteRegisterCommand_initOptions() {
    this.options.unshift({
        option: '-u, --url <url>'
    });
}, _SpoHubSiteRegisterCommand_initValidators = function _SpoHubSiteRegisterCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.url); }));
};
module.exports = new SpoHubSiteRegisterCommand();
