// @pnp/cli-microsoft365/dist/m365/spo/commands/hubsite/hubsite-data-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoHubSiteDataGetCommand_instances, _SpoHubSiteDataGetCommand_initTelemetry, _SpoHubSiteDataGetCommand_initOptions, _SpoHubSiteDataGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoHubSiteDataGetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoHubSiteDataGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoHubSiteDataGetCommand_instances, "m", _SpoHubSiteDataGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteDataGetCommand_instances, "m", _SpoHubSiteDataGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoHubSiteDataGetCommand_instances, "m", _SpoHubSiteDataGetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.HUBSITE_DATA_GET;
    }
    get description() {
        return 'Get hub site data for the specified site';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr('Retrieving hub site data...');
            }
            const forceRefresh = args.options.forceRefresh === true;
            const requestOptions = {
                url: `${args.options.webUrl}/_api/web/HubSiteData(${forceRefresh})`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                if (res['odata.null'] !== true) {
                    logger.log(JSON.parse(res.value));
                }
                else {
                    if (this.verbose) {
                        logger.logToStderr(`${args.options.webUrl} is not connected to a hub site and is not a hub site itself`);
                    }
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoHubSiteDataGetCommand_instances = new WeakSet(), _SpoHubSiteDataGetCommand_initTelemetry = function _SpoHubSiteDataGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            forceRefresh: args.options.forceRefresh === true
        });
    });
}, _SpoHubSiteDataGetCommand_initOptions = function _SpoHubSiteDataGetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-f, --forceRefresh'
    });
}, _SpoHubSiteDataGetCommand_initValidators = function _SpoHubSiteDataGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.webUrl); }));
};
module.exports = new SpoHubSiteDataGetCommand();
