// @pnp/cli-microsoft365/dist/m365/pp/commands/environment/environment-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PpEnvironmentListCommand_instances, _PpEnvironmentListCommand_initTelemetry, _PpEnvironmentListCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const PowerPlatformCommand_1 = require("../../../base/PowerPlatformCommand");
const commands_1 = require("../../commands");
class PpEnvironmentListCommand extends PowerPlatformCommand_1.default {
    constructor() {
        super();
        _PpEnvironmentListCommand_instances.add(this);
        __classPrivateFieldGet(this, _PpEnvironmentListCommand_instances, "m", _PpEnvironmentListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _PpEnvironmentListCommand_instances, "m", _PpEnvironmentListCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.ENVIRONMENT_LIST;
    }
    get description() {
        return 'Lists Microsoft Power Platform environments';
    }
    defaultProperties() {
        return ['name', 'displayName'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Retrieving list of Microsoft Power Platform environments...`);
            }
            let url = '';
            if (args.options.asAdmin) {
                url = `${this.resource}/providers/Microsoft.BusinessAppPlatform/scopes/admin/environments`;
            }
            else {
                url = `${this.resource}/providers/Microsoft.BusinessAppPlatform/environments`;
            }
            const requestOptions = {
                url: `${url}?api-version=2020-10-01`,
                headers: {
                    accept: 'application/json'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                if (res.value && res.value.length > 0) {
                    res.value.forEach(e => {
                        e.displayName = e.properties.displayName;
                    });
                    logger.log(res.value);
                }
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_PpEnvironmentListCommand_instances = new WeakSet(), _PpEnvironmentListCommand_initTelemetry = function _PpEnvironmentListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            asAdmin: !!args.options.asAdmin
        });
    });
}, _PpEnvironmentListCommand_initOptions = function _PpEnvironmentListCommand_initOptions() {
    this.options.unshift({
        option: '-a, --asAdmin'
    });
};
module.exports = new PpEnvironmentListCommand();
