// @pnp/cli-microsoft365/dist/m365/spo/commands/apppage/apppage-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoAppPageSetCommand_instances, _SpoAppPageSetCommand_initOptions, _SpoAppPageSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const urlUtil_1 = require("../../../../utils/urlUtil");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoAppPageSetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoAppPageSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoAppPageSetCommand_instances, "m", _SpoAppPageSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoAppPageSetCommand_instances, "m", _SpoAppPageSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.APPPAGE_SET;
    }
    get description() {
        return 'Updates the single-part app page';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = {
                url: `${args.options.webUrl}/_api/sitepages/Pages/UpdateFullPageApp`,
                headers: {
                    'content-type': 'application/json;odata=nometadata',
                    accept: 'application/json;odata=nometadata'
                },
                responseType: 'json',
                data: {
                    serverRelativeUrl: `${urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, '')}/SitePages/${args.options.pageName}`,
                    webPartDataAsJson: args.options.webPartData
                }
            };
            try {
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoAppPageSetCommand_instances = new WeakSet(), _SpoAppPageSetCommand_initOptions = function _SpoAppPageSetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-n, --pageName <pageName>'
    }, {
        option: '-d, --webPartData <webPartData>'
    });
}, _SpoAppPageSetCommand_initValidators = function _SpoAppPageSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        try {
            JSON.parse(args.options.webPartData);
        }
        catch (e) {
            return `Specified webPartData is not a valid JSON string. Error: ${e}`;
        }
        return true;
    }));
};
module.exports = new SpoAppPageSetCommand();
