// @pnp/cli-microsoft365/dist/m365/spo/commands/storageentity/storageentity-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoStorageEntityRemoveCommand_instances, _SpoStorageEntityRemoveCommand_initTelemetry, _SpoStorageEntityRemoveCommand_initOptions, _SpoStorageEntityRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const config_1 = require("../../../../config");
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoStorageEntityRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoStorageEntityRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoStorageEntityRemoveCommand_instances, "m", _SpoStorageEntityRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoStorageEntityRemoveCommand_instances, "m", _SpoStorageEntityRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoStorageEntityRemoveCommand_instances, "m", _SpoStorageEntityRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.STORAGEENTITY_REMOVE;
    }
    get description() {
        return 'Removes tenant property stored on the specified SharePoint Online app catalog';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeTenantProperty = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing tenant property ${args.options.key} from ${args.options.appCatalogUrl}...`);
                }
                try {
                    const spoAdminUrl = yield spo_1.spo.getSpoAdminUrl(logger, this.debug);
                    const res = yield spo_1.spo.getRequestDigest(spoAdminUrl);
                    const requestOptions = {
                        url: `${spoAdminUrl}/_vti_bin/client.svc/ProcessQuery`,
                        headers: {
                            'X-RequestDigest': res.FormDigestValue
                        },
                        data: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config_1.default.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><ObjectPath Id="31" ObjectPathId="30" /><ObjectPath Id="33" ObjectPathId="32" /><ObjectPath Id="35" ObjectPathId="34" /><Method Name="RemoveStorageEntity" Id="36" ObjectPathId="34"><Parameters><Parameter Type="String">${formatting_1.formatting.escapeXml(args.options.key)}</Parameter></Parameters></Method></Actions><ObjectPaths><Constructor Id="30" TypeId="{268004ae-ef6b-4e9b-8425-127220d84719}" /><Method Id="32" ParentId="30" Name="GetSiteByUrl"><Parameters><Parameter Type="String">${formatting_1.formatting.escapeXml(args.options.appCatalogUrl)}</Parameter></Parameters></Method><Property Id="34" ParentId="32" Name="RootWeb" /></ObjectPaths></Request>`
                    };
                    const processQuery = yield request_1.default.post(requestOptions);
                    const json = JSON.parse(processQuery);
                    const response = json[0];
                    if (response.ErrorInfo) {
                        throw response.ErrorInfo.ErrorMessage;
                    }
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeTenantProperty();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to delete the ${args.options.key} tenant property?`
                });
                if (result.continue) {
                    yield removeTenantProperty();
                }
            }
        });
    }
}
_SpoStorageEntityRemoveCommand_instances = new WeakSet(), _SpoStorageEntityRemoveCommand_initTelemetry = function _SpoStorageEntityRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoStorageEntityRemoveCommand_initOptions = function _SpoStorageEntityRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-u, --appCatalogUrl <appCatalogUrl>'
    }, {
        option: '-k, --key <key>'
    }, {
        option: '--confirm'
    });
}, _SpoStorageEntityRemoveCommand_initValidators = function _SpoStorageEntityRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.appCatalogUrl); }));
};
module.exports = new SpoStorageEntityRemoveCommand();
