// @pnp/cli-microsoft365/dist/m365/cli/commands/config/config-reset.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _CliConfigResetCommand_instances, _CliConfigResetCommand_initTelemetry, _CliConfigResetCommand_initOptions, _CliConfigResetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const settingsNames_1 = require("../../../../settingsNames");
const AnonymousCommand_1 = require("../../../base/AnonymousCommand");
const commands_1 = require("../../commands");
class CliConfigResetCommand extends AnonymousCommand_1.default {
    constructor() {
        super();
        _CliConfigResetCommand_instances.add(this);
        __classPrivateFieldGet(this, _CliConfigResetCommand_instances, "m", _CliConfigResetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _CliConfigResetCommand_instances, "m", _CliConfigResetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _CliConfigResetCommand_instances, "m", _CliConfigResetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.CONFIG_RESET;
    }
    get description() {
        return 'Resets the specified CLI configuration option to its default value';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (args.options.key) {
                Cli_1.Cli.getInstance().config.delete(args.options.key);
            }
            else {
                Cli_1.Cli.getInstance().config.clear();
            }
        });
    }
}
_CliConfigResetCommand_instances = new WeakSet(), _CliConfigResetCommand_initTelemetry = function _CliConfigResetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            key: args.options.key
        });
    });
}, _CliConfigResetCommand_initOptions = function _CliConfigResetCommand_initOptions() {
    this.options.unshift({
        option: '-k, --key [key]',
        autocomplete: CliConfigResetCommand.optionNames
    });
}, _CliConfigResetCommand_initValidators = function _CliConfigResetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.key) {
            if (CliConfigResetCommand.optionNames.indexOf(args.options.key) < 0) {
                return `${args.options.key} is not a valid setting. Allowed values: ${CliConfigResetCommand.optionNames.join(', ')}`;
            }
        }
        return true;
    }));
};
CliConfigResetCommand.optionNames = Object.getOwnPropertyNames(settingsNames_1.settingsNames);
module.exports = new CliConfigResetCommand();
