// @pnp/cli-microsoft365/dist/m365/cli/commands/config/config-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _CliConfigGetCommand_instances, _CliConfigGetCommand_initTelemetry, _CliConfigGetCommand_initOptions, _CliConfigGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const settingsNames_1 = require("../../../../settingsNames");
const AnonymousCommand_1 = require("../../../base/AnonymousCommand");
const commands_1 = require("../../commands");
class CliConfigGetCommand extends AnonymousCommand_1.default {
    constructor() {
        super();
        _CliConfigGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _CliConfigGetCommand_instances, "m", _CliConfigGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _CliConfigGetCommand_instances, "m", _CliConfigGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _CliConfigGetCommand_instances, "m", _CliConfigGetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.CONFIG_GET;
    }
    get description() {
        return 'Gets value of a CLI for Microsoft 365 configuration option';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            logger.log(Cli_1.Cli.getInstance().config.get(args.options.key));
        });
    }
}
_CliConfigGetCommand_instances = new WeakSet(), _CliConfigGetCommand_initTelemetry = function _CliConfigGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            key: args.options.key
        });
    });
}, _CliConfigGetCommand_initOptions = function _CliConfigGetCommand_initOptions() {
    this.options.unshift({
        option: '-k, --key <key>',
        autocomplete: CliConfigGetCommand.optionNames
    });
}, _CliConfigGetCommand_initValidators = function _CliConfigGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (CliConfigGetCommand.optionNames.indexOf(args.options.key) < 0) {
            return `${args.options.key} is not a valid setting. Allowed values: ${CliConfigGetCommand.optionNames.join(', ')}`;
        }
        return true;
    }));
};
CliConfigGetCommand.optionNames = Object.getOwnPropertyNames(settingsNames_1.settingsNames);
module.exports = new CliConfigGetCommand();
