// yoteams-build-core/dist/webpackServe.js
"use strict";
// Licensed under the MIT license.
// SPDX-License-Identifier: MIT
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const webpack_dev_server_1 = __importDefault(require("webpack-dev-server"));
const webpack_1 = __importDefault(require("webpack"));
const fancy_log_1 = __importDefault(require("fancy-log"));
const path_1 = __importDefault(require("path"));
(async () => {
    const webpackConfig = require(path_1.default.join(process.cwd(), "webpack.config"));
    const clientConfig = webpackConfig[1];
    const compiler = (0, webpack_1.default)(clientConfig);
    compiler.hooks.beforeCompile.tap("webpackServe", () => {
        fancy_log_1.default.info("webpack compiling");
    });
    const defaultDevServerConfig = {
        hot: false,
        host: "localhost",
        port: 9000,
        allowedHosts: "all",
        client: {
            overlay: {
                warnings: false,
                errors: true
            }
        },
        devMiddleware: {
            writeToDisk: true,
            stats: {
                all: false,
                colors: true,
                errors: true,
                warnings: true,
                timings: true,
                entrypoints: true
            }
        }
    };
    const server = new webpack_dev_server_1.default(clientConfig.devServer || defaultDevServerConfig, compiler);
    try {
        await server.start();
    }
    catch (error) {
        fancy_log_1.default.error(error);
        throw error;
    }
})();
