// @pnp/cli-microsoft365/dist/m365/spo/commands/file/file-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoFileRemoveCommand_instances, _SpoFileRemoveCommand_initTelemetry, _SpoFileRemoveCommand_initOptions, _SpoFileRemoveCommand_initValidators, _SpoFileRemoveCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoFileRemoveCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoFileRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoFileRemoveCommand_instances, "m", _SpoFileRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoFileRemoveCommand_instances, "m", _SpoFileRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoFileRemoveCommand_instances, "m", _SpoFileRemoveCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoFileRemoveCommand_instances, "m", _SpoFileRemoveCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.FILE_REMOVE;
    }
    get description() {
        return 'Removes the specified file';
    }
    getExcludedOptionsWithUrls() {
        return ['url'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeFile = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing file in site at ${args.options.webUrl}...`);
                }
                let requestUrl = '';
                if (args.options.id) {
                    requestUrl = `${args.options.webUrl}/_api/web/GetFileById(guid'${encodeURIComponent(args.options.id)}')`;
                }
                else {
                    // concatenate trailing '/' if not provided
                    // so if the provided url is for the root site, the substr bellow will get the right value
                    let serverRelativeSiteUrl = args.options.webUrl;
                    if (!serverRelativeSiteUrl.endsWith('/')) {
                        serverRelativeSiteUrl = `${serverRelativeSiteUrl}/`;
                    }
                    serverRelativeSiteUrl = serverRelativeSiteUrl.substr(serverRelativeSiteUrl.indexOf('/', 8));
                    let fileUrl = args.options.url;
                    if (!fileUrl.startsWith(serverRelativeSiteUrl)) {
                        fileUrl = `${serverRelativeSiteUrl}${fileUrl}`;
                    }
                    requestUrl = `${args.options.webUrl}/_api/web/GetFileByServerRelativeUrl('${encodeURIComponent(fileUrl)}')`;
                }
                if (args.options.recycle) {
                    requestUrl += `/recycle()`;
                }
                const requestOptions = {
                    url: requestUrl,
                    method: 'POST',
                    headers: {
                        'X-HTTP-Method': 'DELETE',
                        'If-Match': '*',
                        'accept': 'application/json;odata=nometadata'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.post(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeFile();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to ${args.options.recycle ? "recycle" : "remove"} the file ${args.options.id || args.options.url} located in site ${args.options.webUrl}?`
                });
                if (result.continue) {
                    yield removeFile();
                }
            }
        });
    }
}
_SpoFileRemoveCommand_instances = new WeakSet(), _SpoFileRemoveCommand_initTelemetry = function _SpoFileRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            id: (!(!args.options.id)).toString(),
            url: (!(!args.options.url)).toString(),
            recycle: (!(!args.options.recycle)).toString(),
            confirm: (!(!args.options.confirm)).toString()
        });
    });
}, _SpoFileRemoveCommand_initOptions = function _SpoFileRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-w, --webUrl <webUrl>'
    }, {
        option: '-i, --id [id]'
    }, {
        option: '-u, --url [url]'
    }, {
        option: '--recycle'
    }, {
        option: '--confirm'
    });
}, _SpoFileRemoveCommand_initValidators = function _SpoFileRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        if (args.options.id &&
            !validation_1.validation.isValidGuid(args.options.id)) {
            return `${args.options.id} is not a valid GUID`;
        }
        return true;
    }));
}, _SpoFileRemoveCommand_initOptionSets = function _SpoFileRemoveCommand_initOptionSets() {
    this.optionSets.push(['id', 'url']);
};
module.exports = new SpoFileRemoveCommand();
