// @pnp/cli-microsoft365/dist/m365/aad/commands/oauth2grant/oauth2grant-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadOAuth2GrantRemoveCommand_instances, _AadOAuth2GrantRemoveCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadOAuth2GrantRemoveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadOAuth2GrantRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadOAuth2GrantRemoveCommand_instances, "m", _AadOAuth2GrantRemoveCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.OAUTH2GRANT_REMOVE;
    }
    get description() {
        return 'Remove specified service principal OAuth2 permissions';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeOauth2Grant = () => __awaiter(this, void 0, void 0, function* () {
                if (this.verbose) {
                    logger.logToStderr(`Removing OAuth2 permissions...`);
                }
                try {
                    const requestOptions = {
                        url: `${this.resource}/v1.0/oauth2PermissionGrants/${encodeURIComponent(args.options.grantId)}`,
                        headers: {
                            'accept': 'application/json;odata.metadata=none'
                        },
                        responseType: 'json'
                    };
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeOauth2Grant();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the OAuth2 permissions for ${args.options.grantId}?`
                });
                if (result.continue) {
                    yield removeOauth2Grant();
                }
            }
        });
    }
}
_AadOAuth2GrantRemoveCommand_instances = new WeakSet(), _AadOAuth2GrantRemoveCommand_initOptions = function _AadOAuth2GrantRemoveCommand_initOptions() {
    this.options.unshift({
        option: '-i, --grantId <grantId>'
    }, {
        option: '--confirm'
    });
};
module.exports = new AadOAuth2GrantRemoveCommand();
