// @pnp/cli-microsoft365/dist/m365/aad/commands/oauth2grant/oauth2grant-add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadOAuth2GrantAddCommand_instances, _AadOAuth2GrantAddCommand_initOptions, _AadOAuth2GrantAddCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadOAuth2GrantAddCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadOAuth2GrantAddCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadOAuth2GrantAddCommand_instances, "m", _AadOAuth2GrantAddCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _AadOAuth2GrantAddCommand_instances, "m", _AadOAuth2GrantAddCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.OAUTH2GRANT_ADD;
    }
    get description() {
        return 'Grant the specified service principal OAuth2 permissions to the specified resource';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Granting the service principal specified permissions...`);
            }
            try {
                const requestOptions = {
                    url: `${this.resource}/v1.0/oauth2PermissionGrants`,
                    headers: {
                        'content-type': 'application/json;odata.metadata=none'
                    },
                    responseType: 'json',
                    data: {
                        "clientId": args.options.clientId,
                        "consentType": "AllPrincipals",
                        "principalId": null,
                        "resourceId": args.options.resourceId,
                        "scope": args.options.scope
                    }
                };
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadOAuth2GrantAddCommand_instances = new WeakSet(), _AadOAuth2GrantAddCommand_initOptions = function _AadOAuth2GrantAddCommand_initOptions() {
    this.options.unshift({
        option: '-i, --clientId <clientId>'
    }, {
        option: '-r, --resourceId <resourceId>'
    }, {
        option: '-s, --scope <scope>'
    });
}, _AadOAuth2GrantAddCommand_initValidators = function _AadOAuth2GrantAddCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.clientId)) {
            return `${args.options.clientId} is not a valid GUID`;
        }
        if (!validation_1.validation.isValidGuid(args.options.resourceId)) {
            return `${args.options.resourceId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new AadOAuth2GrantAddCommand();
