// @pnp/cli-microsoft365/dist/m365/aad/commands/oauth2grant/oauth2grant-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AadOAuth2GrantSetCommand_instances, _AadOAuth2GrantSetCommand_initOptions;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class AadOAuth2GrantSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _AadOAuth2GrantSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _AadOAuth2GrantSetCommand_instances, "m", _AadOAuth2GrantSetCommand_initOptions).call(this);
    }
    get name() {
        return commands_1.default.OAUTH2GRANT_SET;
    }
    get description() {
        return 'Update OAuth2 permissions for the service principal';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.verbose) {
                logger.logToStderr(`Updating OAuth2 permissions...`);
            }
            try {
                const requestOptions = {
                    url: `${this.resource}/v1.0/oauth2PermissionGrants/${encodeURIComponent(args.options.grantId)}`,
                    headers: {
                        'content-type': 'application/json'
                    },
                    responseType: 'json',
                    data: {
                        "scope": args.options.scope
                    }
                };
                yield request_1.default.patch(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_AadOAuth2GrantSetCommand_instances = new WeakSet(), _AadOAuth2GrantSetCommand_initOptions = function _AadOAuth2GrantSetCommand_initOptions() {
    this.options.unshift({
        option: '-i, --grantId <grantId>'
    }, {
        option: '-s, --scope <scope>'
    });
};
module.exports = new AadOAuth2GrantSetCommand();
