// @pnp/cli-microsoft365/dist/m365/spo/commands/site/site-inplacerecordsmanagement-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoSiteInPlaceRecordsManagementSetCommand_instances, _SpoSiteInPlaceRecordsManagementSetCommand_initTelemetry, _SpoSiteInPlaceRecordsManagementSetCommand_initOptions, _SpoSiteInPlaceRecordsManagementSetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoSiteInPlaceRecordsManagementSetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoSiteInPlaceRecordsManagementSetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoSiteInPlaceRecordsManagementSetCommand_instances, "m", _SpoSiteInPlaceRecordsManagementSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoSiteInPlaceRecordsManagementSetCommand_instances, "m", _SpoSiteInPlaceRecordsManagementSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoSiteInPlaceRecordsManagementSetCommand_instances, "m", _SpoSiteInPlaceRecordsManagementSetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.SITE_INPLACERECORDSMANAGEMENT_SET;
    }
    get description() {
        return 'Activates or deactivates in-place records management for a site collection';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const enabled = args.options.enabled.toLocaleLowerCase() === 'true';
            const requestOptions = {
                url: `${args.options.siteUrl}/_api/site/features/${enabled ? 'add' : 'remove'}`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                data: {
                    featureId: 'da2e115b-07e4-49d9-bb2c-35e93bb9fca9',
                    force: true
                },
                responseType: 'json'
            };
            if (this.verbose) {
                logger.logToStderr(`${enabled ? 'Activating' : 'Deactivating'} in-place records management for site ${args.options.siteUrl}`);
            }
            try {
                yield request_1.default.post(requestOptions);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoSiteInPlaceRecordsManagementSetCommand_instances = new WeakSet(), _SpoSiteInPlaceRecordsManagementSetCommand_initTelemetry = function _SpoSiteInPlaceRecordsManagementSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            enabled: args.options.enabled
        });
    });
}, _SpoSiteInPlaceRecordsManagementSetCommand_initOptions = function _SpoSiteInPlaceRecordsManagementSetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --siteUrl <siteUrl>'
    }, {
        option: '--enabled <enabled>'
    });
}, _SpoSiteInPlaceRecordsManagementSetCommand_initValidators = function _SpoSiteInPlaceRecordsManagementSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidBoolean(args.options.enabled)) {
            return 'Invalid "enabled" option value. Specify "true" or "false"';
        }
        return validation_1.validation.isValidSharePointUrl(args.options.siteUrl);
    }));
};
module.exports = new SpoSiteInPlaceRecordsManagementSetCommand();
