// @pnp/cli-microsoft365/dist/m365/spo/commands/eventreceiver/eventreceiver-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoEventreceiverGetCommand_instances, _SpoEventreceiverGetCommand_initTelemetry, _SpoEventreceiverGetCommand_initOptions, _SpoEventreceiverGetCommand_initValidators, _SpoEventreceiverGetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const formatting_1 = require("../../../../utils/formatting");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoEventreceiverGetCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoEventreceiverGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoEventreceiverGetCommand_instances, "m", _SpoEventreceiverGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoEventreceiverGetCommand_instances, "m", _SpoEventreceiverGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoEventreceiverGetCommand_instances, "m", _SpoEventreceiverGetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _SpoEventreceiverGetCommand_instances, "m", _SpoEventreceiverGetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.EVENTRECEIVER_GET;
    }
    get description() {
        return 'Gets a specific event receiver attached to the web, site or list (if any of the list options are filled in) by receiver name of id';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            let requestUrl = `${args.options.webUrl}/_api/`;
            let listUrl = '';
            let filter = '?$filter=';
            if (args.options.listId) {
                listUrl = `lists(guid'${formatting_1.formatting.encodeQueryParameter(args.options.listId)}')/`;
            }
            else if (args.options.listTitle) {
                listUrl = `lists/getByTitle('${formatting_1.formatting.encodeQueryParameter(args.options.listTitle)}')/`;
            }
            else if (args.options.listUrl) {
                const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                listUrl = `GetList('${formatting_1.formatting.encodeQueryParameter(listServerRelativeUrl)}')/`;
            }
            if (!args.options.scope || args.options.scope === 'web') {
                requestUrl += `web/${listUrl}eventreceivers`;
            }
            else {
                requestUrl += 'site/eventreceivers';
            }
            if (args.options.id) {
                filter += `receiverid eq (guid'${args.options.id}')`;
            }
            else {
                filter += `receivername eq '${args.options.name}'`;
            }
            const requestOptions = {
                url: requestUrl + filter,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res.value);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoEventreceiverGetCommand_instances = new WeakSet(), _SpoEventreceiverGetCommand_initTelemetry = function _SpoEventreceiverGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            scope: typeof args.options.scope !== 'undefined',
            id: typeof args.options.id !== 'undefined',
            name: typeof args.options.name !== 'undefined'
        });
    });
}, _SpoEventreceiverGetCommand_initOptions = function _SpoEventreceiverGetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--listTitle [listTitle]'
    }, {
        option: '--listId  [listId]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '-n, --name [name]'
    }, {
        option: '-i, --id [id]'
    }, {
        option: '-s, --scope [scope]',
        autocomplete: ['web', 'site']
    });
}, _SpoEventreceiverGetCommand_initValidators = function _SpoEventreceiverGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        const listOptions = [args.options.listId, args.options.listTitle, args.options.listUrl];
        if (listOptions.some(item => item !== undefined) && listOptions.filter(item => item !== undefined).length > 1) {
            return `Specify either list id or title or list url`;
        }
        if (args.options.listId && !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} is not a valid GUID`;
        }
        if (args.options.scope && ['web', 'site'].indexOf(args.options.scope) === -1) {
            return `${args.options.scope} is not a valid type value. Allowed values web|site.`;
        }
        if (args.options.scope && args.options.scope === 'site' && (args.options.listId || args.options.listUrl || args.options.listTitle)) {
            return 'Scope cannot be set to site when retrieving list event receivers.';
        }
        return true;
    }));
}, _SpoEventreceiverGetCommand_initOptionSets = function _SpoEventreceiverGetCommand_initOptionSets() {
    this.optionSets.push(['name', 'id']);
};
module.exports = new SpoEventreceiverGetCommand();
