// @pnp/cli-microsoft365/dist/m365/spo/commands/eventreceiver/eventreceiver-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoEventreceiverListCommand_instances, _SpoEventreceiverListCommand_initTelemetry, _SpoEventreceiverListCommand_initOptions, _SpoEventreceiverListCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const urlUtil_1 = require("../../../../utils/urlUtil");
const validation_1 = require("../../../../utils/validation");
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
class SpoEventreceiverListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        _SpoEventreceiverListCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoEventreceiverListCommand_instances, "m", _SpoEventreceiverListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoEventreceiverListCommand_instances, "m", _SpoEventreceiverListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoEventreceiverListCommand_instances, "m", _SpoEventreceiverListCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.EVENTRECEIVER_LIST;
    }
    get description() {
        return 'Retrieves event receivers for the specified web, site or list.';
    }
    defaultProperties() {
        return ['ReceiverId', 'ReceiverName'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            let requestUrl = `${args.options.webUrl}/_api/`;
            let listUrl = '';
            if (args.options.listId) {
                listUrl = `lists(guid'${encodeURIComponent(args.options.listId)}')/`;
            }
            else if (args.options.listTitle) {
                listUrl = `lists/getByTitle('${encodeURIComponent(args.options.listTitle)}')/`;
            }
            else if (args.options.listUrl) {
                const listServerRelativeUrl = urlUtil_1.urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
                listUrl = `GetList('${encodeURIComponent(listServerRelativeUrl)}')/`;
            }
            if (!args.options.scope || args.options.scope === 'web') {
                requestUrl += `web/${listUrl}eventreceivers`;
            }
            else {
                requestUrl += 'site/eventreceivers';
            }
            const requestOptions = {
                url: requestUrl,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            try {
                const res = yield request_1.default.get(requestOptions);
                logger.log(res.value);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
}
_SpoEventreceiverListCommand_instances = new WeakSet(), _SpoEventreceiverListCommand_initTelemetry = function _SpoEventreceiverListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            listId: typeof args.options.listId !== 'undefined',
            listTitle: typeof args.options.listTitle !== 'undefined',
            listUrl: typeof args.options.listUrl !== 'undefined',
            scope: typeof args.options.scope !== 'undefined'
        });
    });
}, _SpoEventreceiverListCommand_initOptions = function _SpoEventreceiverListCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '--listTitle [listTitle]'
    }, {
        option: '--listId  [listId]'
    }, {
        option: '--listUrl [listUrl]'
    }, {
        option: '-s, --scope [scope]',
        autocomplete: ['web', 'site']
    });
}, _SpoEventreceiverListCommand_initValidators = function _SpoEventreceiverListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        const isValidSharePointUrl = validation_1.validation.isValidSharePointUrl(args.options.webUrl);
        if (isValidSharePointUrl !== true) {
            return isValidSharePointUrl;
        }
        const listOptions = [args.options.listId, args.options.listTitle, args.options.listUrl];
        if (listOptions.some(item => item !== undefined) && listOptions.filter(item => item !== undefined).length > 1) {
            return `Specify either list id or title or list url`;
        }
        if (args.options.listId && !validation_1.validation.isValidGuid(args.options.listId)) {
            return `${args.options.listId} is not a valid GUID`;
        }
        if (args.options.scope && ['web', 'site'].indexOf(args.options.scope) === -1) {
            return `${args.options.scope} is not a valid type value. Allowed values web|site.`;
        }
        if (args.options.scope && args.options.scope === 'site' && (args.options.listId || args.options.listUrl || args.options.listTitle)) {
            return 'Scope cannot be set to site when retrieving list event receivers.';
        }
        return true;
    }));
};
module.exports = new SpoEventreceiverListCommand();
