// @pnp/cli-microsoft365/dist/m365/spo/commands/propertybag/propertybag-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpoPropertyBagGetCommand_instances, _SpoPropertyBagGetCommand_initTelemetry, _SpoPropertyBagGetCommand_initOptions, _SpoPropertyBagGetCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const spo_1 = require("../../../../utils/spo");
const validation_1 = require("../../../../utils/validation");
const commands_1 = require("../../commands");
const propertybag_base_1 = require("./propertybag-base");
class SpoPropertyBagGetCommand extends propertybag_base_1.SpoPropertyBagBaseCommand {
    constructor() {
        super();
        _SpoPropertyBagGetCommand_instances.add(this);
        __classPrivateFieldGet(this, _SpoPropertyBagGetCommand_instances, "m", _SpoPropertyBagGetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _SpoPropertyBagGetCommand_instances, "m", _SpoPropertyBagGetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _SpoPropertyBagGetCommand_instances, "m", _SpoPropertyBagGetCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.PROPERTYBAG_GET;
    }
    get description() {
        return 'Gets the value of the specified property from the property bag';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const contextResponse = yield spo_1.spo.getRequestDigest(args.options.webUrl);
                this.formDigestValue = contextResponse.FormDigestValue;
                const identityResp = yield spo_1.spo.getCurrentWebIdentity(args.options.webUrl, this.formDigestValue);
                let propertyBagData;
                const opts = args.options;
                if (opts.folder) {
                    propertyBagData = yield this.getFolderPropertyBag(identityResp, opts.webUrl, opts.folder, logger);
                }
                else {
                    propertyBagData = yield this.getWebPropertyBag(identityResp, opts.webUrl, logger);
                }
                const property = this.filterByKey(propertyBagData, args.options.key);
                if (property) {
                    logger.log(property.value);
                }
                else if (this.verbose) {
                    logger.logToStderr('Property not found.');
                }
            }
            catch (err) {
                this.handleRejectedPromise(err);
            }
        });
    }
    filterByKey(propertyBag, key) {
        const keys = Object.keys(propertyBag);
        for (let i = 0; i < keys.length; i++) {
            // we have to normalize the keys and values before we can filter
            // since they carry extra information
            // ex. : 'vti_level$  Int32' instead of 'vti_level'
            const formattedProperty = this.formatProperty(keys[i], propertyBag[keys[i]]);
            if (formattedProperty.key === key) {
                return formattedProperty;
            }
        }
        return null;
    }
}
_SpoPropertyBagGetCommand_instances = new WeakSet(), _SpoPropertyBagGetCommand_initTelemetry = function _SpoPropertyBagGetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            folder: typeof args.options.folder !== 'undefined'
        });
    });
}, _SpoPropertyBagGetCommand_initOptions = function _SpoPropertyBagGetCommand_initOptions() {
    this.options.unshift({
        option: '-u, --webUrl <webUrl>'
    }, {
        option: '-k, --key <key>'
    }, {
        option: '-f, --folder [folder]'
    });
}, _SpoPropertyBagGetCommand_initValidators = function _SpoPropertyBagGetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () { return validation_1.validation.isValidSharePointUrl(args.options.webUrl); }));
};
module.exports = new SpoPropertyBagGetCommand();
