// @pnp/cli-microsoft365/dist/m365/teams/commands/channel/channel-member-set.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsChannelMemberSetCommand_instances, _TeamsChannelMemberSetCommand_initTelemetry, _TeamsChannelMemberSetCommand_initOptions, _TeamsChannelMemberSetCommand_initValidators, _TeamsChannelMemberSetCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsChannelMemberSetCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsChannelMemberSetCommand_instances.add(this);
        this.teamId = '';
        this.channelId = '';
        __classPrivateFieldGet(this, _TeamsChannelMemberSetCommand_instances, "m", _TeamsChannelMemberSetCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberSetCommand_instances, "m", _TeamsChannelMemberSetCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberSetCommand_instances, "m", _TeamsChannelMemberSetCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberSetCommand_instances, "m", _TeamsChannelMemberSetCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.CHANNEL_MEMBER_SET;
    }
    get description() {
        return 'Updates the role of the specified member in the specified Microsoft Teams private or shared team channel';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.teamId = yield this.getTeamId(args);
                this.channelId = yield this.getChannelId(args);
                const memberId = yield this.getMemberId(args);
                const requestOptions = {
                    url: `${this.resource}/v1.0/teams/${this.teamId}/channels/${this.channelId}/members/${memberId}`,
                    headers: {
                        'accept': 'application/json;odata.metadata=none',
                        'Prefer': 'return=representation'
                    },
                    responseType: 'json',
                    data: {
                        '@odata.type': '#microsoft.graph.aadUserConversationMember',
                        roles: [args.options.role]
                    }
                };
                const member = yield request_1.default.patch(requestOptions);
                logger.log(member);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getTeamId(args) {
        if (args.options.teamId) {
            return Promise.resolve(args.options.teamId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.teamName)
            .then(group => {
            if (group.resourceProvisioningOptions.indexOf('Team') === -1) {
                return Promise.reject(`The specified team does not exist in the Microsoft Teams`);
            }
            return group.id;
        });
    }
    getChannelId(args) {
        if (args.options.channelId) {
            return Promise.resolve(args.options.channelId);
        }
        const requestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(this.teamId)}/channels?$filter=displayName eq '${encodeURIComponent(args.options.channelName)}'`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(requestOptions)
            .then(response => {
            const channelItem = response.value[0];
            if (!channelItem) {
                return Promise.reject(`The specified channel does not exist in the Microsoft Teams team`);
            }
            if (channelItem.membershipType !== "private") {
                return Promise.reject(`The specified channel is not a private channel`);
            }
            return Promise.resolve(channelItem.id);
        });
    }
    getMemberId(args) {
        if (args.options.id) {
            return Promise.resolve(args.options.id);
        }
        const requestOptions = {
            url: `${this.resource}/v1.0/teams/${this.teamId}/channels/${this.channelId}/members`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(requestOptions)
            .then(response => {
            const conversationMembers = response.value.filter(x => {
                var _a, _b;
                return args.options.userId && ((_a = x.userId) === null || _a === void 0 ? void 0 : _a.toLocaleLowerCase()) === args.options.userId.toLocaleLowerCase() ||
                    args.options.userName && ((_b = x.email) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase()) === args.options.userName.toLocaleLowerCase();
            });
            const conversationMember = conversationMembers[0];
            if (!conversationMember) {
                return Promise.reject(`The specified member does not exist in the Microsoft Teams channel`);
            }
            if (conversationMembers.length > 1) {
                return Promise.reject(`Multiple Microsoft Teams channel members with name ${args.options.userName} found: ${response.value.map(x => x.userId)}`);
            }
            return Promise.resolve(conversationMember.id);
        });
    }
}
_TeamsChannelMemberSetCommand_instances = new WeakSet(), _TeamsChannelMemberSetCommand_initTelemetry = function _TeamsChannelMemberSetCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            teamId: typeof args.options.teamId !== 'undefined',
            teamName: typeof args.options.teamName !== 'undefined',
            channelId: typeof args.options.channelId !== 'undefined',
            channelName: typeof args.options.channelName !== 'undefined',
            userName: typeof args.options.userName !== 'undefined',
            userId: typeof args.options.userId !== 'undefined',
            id: typeof args.options.id !== 'undefined'
        });
    });
}, _TeamsChannelMemberSetCommand_initOptions = function _TeamsChannelMemberSetCommand_initOptions() {
    this.options.unshift({
        option: '--teamId [teamId]'
    }, {
        option: '--teamName [teamName]'
    }, {
        option: '--channelId [channelId]'
    }, {
        option: '--channelName [channelName]'
    }, {
        option: '--userName [userName]'
    }, {
        option: '--userId [userId]'
    }, {
        option: '--id [id]'
    }, {
        option: '-r, --role <role>',
        autocomplete: ['owner', 'member']
    });
}, _TeamsChannelMemberSetCommand_initValidators = function _TeamsChannelMemberSetCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.teamId && !validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (args.options.channelId && !validation_1.validation.isValidTeamsChannelId(args.options.channelId)) {
            return `${args.options.channelId} is not a valid Teams Channel ID`;
        }
        if (args.options.userId && !validation_1.validation.isValidGuid(args.options.userId)) {
            return `${args.options.userId} is not a valid GUID`;
        }
        if (['owner', 'member'].indexOf(args.options.role) === -1) {
            return `${args.options.role} is not a valid role value. Allowed values owner|member`;
        }
        return true;
    }));
}, _TeamsChannelMemberSetCommand_initOptionSets = function _TeamsChannelMemberSetCommand_initOptionSets() {
    this.optionSets.push(['teamId', 'teamName'], ['channelId', 'channelName'], ['userName', 'userId', 'id']);
};
module.exports = new TeamsChannelMemberSetCommand();
