// @pnp/cli-microsoft365/dist/m365/teams/commands/channel/channel-member-list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsChannelMemberListCommand_instances, _TeamsChannelMemberListCommand_initTelemetry, _TeamsChannelMemberListCommand_initOptions, _TeamsChannelMemberListCommand_initValidators, _TeamsChannelMemberListCommand_initOptionSets;
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const odata_1 = require("../../../../utils/odata");
const validation_1 = require("../../../../utils/validation");
const aadGroup_1 = require("../../../../utils/aadGroup");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsChannelMemberListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsChannelMemberListCommand_instances.add(this);
        this.teamId = '';
        __classPrivateFieldGet(this, _TeamsChannelMemberListCommand_instances, "m", _TeamsChannelMemberListCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberListCommand_instances, "m", _TeamsChannelMemberListCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberListCommand_instances, "m", _TeamsChannelMemberListCommand_initValidators).call(this);
        __classPrivateFieldGet(this, _TeamsChannelMemberListCommand_instances, "m", _TeamsChannelMemberListCommand_initOptionSets).call(this);
    }
    get name() {
        return commands_1.default.CHANNEL_MEMBER_LIST;
    }
    get description() {
        return 'Lists members of the specified Microsoft Teams team channel';
    }
    alias() {
        return [commands_1.default.CONVERSATIONMEMBER_LIST];
    }
    defaultProperties() {
        return ['id', 'roles', 'displayName', 'userId', 'email'];
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            this.showDeprecationWarning(logger, commands_1.default.CONVERSATIONMEMBER_LIST, commands_1.default.CHANNEL_MEMBER_LIST);
            try {
                this.teamId = yield this.getTeamId(args);
                const channelId = yield this.getChannelId(args);
                const endpoint = `${this.resource}/v1.0/teams/${this.teamId}/channels/${channelId}/members`;
                let memberships = yield odata_1.odata.getAllItems(endpoint);
                if (args.options.role) {
                    if (args.options.role === 'member') {
                        // Members have no role value
                        memberships = memberships.filter(i => i.roles.length === 0);
                    }
                    else {
                        memberships = memberships.filter(i => i.roles.indexOf(args.options.role) !== -1);
                    }
                }
                logger.log(memberships);
            }
            catch (err) {
                this.handleRejectedODataJsonPromise(err);
            }
        });
    }
    getTeamId(args) {
        if (args.options.teamId) {
            return Promise.resolve(args.options.teamId);
        }
        return aadGroup_1.aadGroup
            .getGroupByDisplayName(args.options.teamName)
            .then(group => {
            if (group.resourceProvisioningOptions.indexOf('Team') === -1) {
                return Promise.reject(`The specified team does not exist in the Microsoft Teams`);
            }
            return group.id;
        });
    }
    getChannelId(args) {
        if (args.options.channelId) {
            return Promise.resolve(args.options.channelId);
        }
        const channelRequestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(this.teamId)}/channels?$filter=displayName eq '${encodeURIComponent(args.options.channelName)}'`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            responseType: 'json'
        };
        return request_1.default
            .get(channelRequestOptions)
            .then(response => {
            const channelItem = response.value[0];
            if (!channelItem) {
                return Promise.reject(`The specified channel does not exist in the Microsoft Teams team`);
            }
            return Promise.resolve(channelItem.id);
        });
    }
}
_TeamsChannelMemberListCommand_instances = new WeakSet(), _TeamsChannelMemberListCommand_initTelemetry = function _TeamsChannelMemberListCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            teamId: typeof args.options.teamId !== 'undefined',
            teamName: typeof args.options.teamName !== 'undefined',
            channelId: typeof args.options.channelId !== 'undefined',
            channelName: typeof args.options.channelName !== 'undefined',
            role: typeof args.options.role
        });
    });
}, _TeamsChannelMemberListCommand_initOptions = function _TeamsChannelMemberListCommand_initOptions() {
    this.options.unshift({
        option: '-i, --teamId [teamId]'
    }, {
        option: '--teamName [teamName]'
    }, {
        option: '-c, --channelId [channelId]'
    }, {
        option: '--channelName [channelName]'
    }, {
        option: '-r, --role [role]',
        autocomplete: ['owner', 'member', 'guest']
    });
}, _TeamsChannelMemberListCommand_initValidators = function _TeamsChannelMemberListCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (args.options.teamId && !validation_1.validation.isValidGuid(args.options.teamId)) {
            return `${args.options.teamId} is not a valid GUID`;
        }
        if (args.options.channelId && !validation_1.validation.isValidTeamsChannelId(args.options.channelId)) {
            return `${args.options.channelId} is not a valid Teams Channel ID`;
        }
        if (args.options.role) {
            if (['owner', 'member', 'guest'].indexOf(args.options.role) === -1) {
                return `${args.options.role} is not a valid role value. Allowed values owner|member|guest`;
            }
        }
        return true;
    }));
}, _TeamsChannelMemberListCommand_initOptionSets = function _TeamsChannelMemberListCommand_initOptionSets() {
    this.optionSets.push(['teamId', 'teamName'], ['channelId', 'channelName']);
};
module.exports = new TeamsChannelMemberListCommand();
