// @pnp/cli-microsoft365/dist/m365/teams/commands/user/user-app-remove.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TeamsUserAppRemoveCommand_instances, _TeamsUserAppRemoveCommand_initTelemetry, _TeamsUserAppRemoveCommand_initOptions, _TeamsUserAppRemoveCommand_initValidators;
Object.defineProperty(exports, "__esModule", { value: true });
const Cli_1 = require("../../../../cli/Cli");
const request_1 = require("../../../../request");
const validation_1 = require("../../../../utils/validation");
const GraphCommand_1 = require("../../../base/GraphCommand");
const commands_1 = require("../../commands");
class TeamsUserAppRemoveCommand extends GraphCommand_1.default {
    constructor() {
        super();
        _TeamsUserAppRemoveCommand_instances.add(this);
        __classPrivateFieldGet(this, _TeamsUserAppRemoveCommand_instances, "m", _TeamsUserAppRemoveCommand_initTelemetry).call(this);
        __classPrivateFieldGet(this, _TeamsUserAppRemoveCommand_instances, "m", _TeamsUserAppRemoveCommand_initOptions).call(this);
        __classPrivateFieldGet(this, _TeamsUserAppRemoveCommand_instances, "m", _TeamsUserAppRemoveCommand_initValidators).call(this);
    }
    get name() {
        return commands_1.default.USER_APP_REMOVE;
    }
    get description() {
        return 'Uninstall an app from the personal scope of the specified user.';
    }
    commandAction(logger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            const removeApp = () => __awaiter(this, void 0, void 0, function* () {
                const endpoint = `${this.resource}/v1.0`;
                const requestOptions = {
                    url: `${endpoint}/users/${args.options.userId}/teamwork/installedApps/${args.options.appId}`,
                    headers: {
                        'accept': 'application/json;odata.metadata=none'
                    },
                    responseType: 'json'
                };
                try {
                    yield request_1.default.delete(requestOptions);
                }
                catch (err) {
                    this.handleRejectedODataJsonPromise(err);
                }
            });
            if (args.options.confirm) {
                yield removeApp();
            }
            else {
                const result = yield Cli_1.Cli.prompt({
                    type: 'confirm',
                    name: 'continue',
                    default: false,
                    message: `Are you sure you want to remove the app with id ${args.options.appId} for user ${args.options.userId}?`
                });
                if (result.continue) {
                    yield removeApp();
                }
            }
        });
    }
}
_TeamsUserAppRemoveCommand_instances = new WeakSet(), _TeamsUserAppRemoveCommand_initTelemetry = function _TeamsUserAppRemoveCommand_initTelemetry() {
    this.telemetry.push((args) => {
        Object.assign(this.telemetryProperties, {
            confirm: (!!args.options.confirm).toString()
        });
    });
}, _TeamsUserAppRemoveCommand_initOptions = function _TeamsUserAppRemoveCommand_initOptions() {
    this.options.unshift({
        option: '--appId <appId>'
    }, {
        option: '--userId <userId>'
    }, {
        option: '--confirm'
    });
}, _TeamsUserAppRemoveCommand_initValidators = function _TeamsUserAppRemoveCommand_initValidators() {
    this.validators.push((args) => __awaiter(this, void 0, void 0, function* () {
        if (!validation_1.validation.isValidGuid(args.options.userId)) {
            return `${args.options.userId} is not a valid GUID`;
        }
        return true;
    }));
};
module.exports = new TeamsUserAppRemoveCommand();
